// @buidlrrr/rain-sdk/dist/socket/RainSocket.js
import { io } from 'socket.io-client';
import { ALLOWED_ENVIRONMENTS, ENV_CONFIG } from '../config/environments.js';
export class RainSocket {
    socket;
    constructor(config = {}) {
        const { environment = 'development' } = config;
        function isValidEnvironment(env) {
            return ALLOWED_ENVIRONMENTS.includes(env);
        }
        if (!isValidEnvironment(environment)) {
            throw new Error(`Invalid environment "${environment}". Allowed values: ${ALLOWED_ENVIRONMENTS.join(', ')}`);
        }
        const { apiUrl } = ENV_CONFIG[environment];
        this.socket = io(apiUrl, { transports: ['websocket', 'polling'] });
    }
    get connected() {
        return this.socket.connected;
    }
    onConnect(callback) {
        this.socket.on('connect', callback);
    }
    onDisconnect(callback) {
        this.socket.on('disconnect', callback);
    }
    /**
     * Subscribe to the enter-option event for a specific market.
     * Fires when a user buys into a market option.
     * Returns an unsubscribe function — call it to stop listening.
     */
    onEnterOption(marketId, callback) {
        const event = `enter-option/${marketId}`;
        const handler = (raw) => {
            const investments = (raw?.enterOption?.investment ?? []).map((inv) => ({
                choiceIndex: inv?.choiceIndex ?? inv?.option ?? 0,
                amount: String(inv?.amount ?? inv?.investment ?? '0'),
                optionName: inv?.optionName,
            }));
            callback({
                poolId: raw?.pool?._id ?? marketId,
                investments,
                totalInvestmentWei: String(raw?.enterOption?.totalInvestment ?? '0'),
                tokenDecimals: raw?.pool?.token?.tokenDecimals ?? 6,
            });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the order-created event for a specific market.
     * Fires when a new limit order is placed.
     */
    onOrderCreated(marketId, callback) {
        const event = `order-created/${marketId}`;
        const handler = (raw) => {
            callback({
                poolId: raw?.pool?._id ?? marketId,
                order: raw?.order ?? {},
                tokenDecimals: raw?.pool?.token?.tokenDecimals ?? 6,
            });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the order-cancelled event for a specific market.
     * Fires when an open order is cancelled.
     */
    onOrderCancelled(marketId, callback) {
        const event = `order-cancelled/${marketId}`;
        const handler = (raw) => {
            callback({
                poolId: raw?.pool?._id ?? marketId,
                order: raw?.order ?? {},
                tokenDecimals: raw?.pool?.token?.tokenDecimals ?? 6,
            });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the order-filled event for a specific market.
     * Fires when an order is partially or fully filled.
     */
    onOrderFilled(marketId, callback) {
        const event = `order-filled/${marketId}`;
        const handler = (raw) => {
            callback({ poolId: raw?.pool?._id ?? marketId, raw });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the dispute-opened event for a specific market.
     */
    onDisputeOpened(marketId, callback) {
        const event = `dispute-opened/${marketId}`;
        const handler = (raw) => {
            callback({
                poolId: raw?.pool?._id ?? marketId,
                isDisputed: raw?.pool?.isDisputed ?? false,
                isAiResolver: raw?.pool?.isAiResolver ?? false,
                status: raw?.pool?.status ?? '',
                disputeData: raw?.pool?.disputeData ?? null,
                winnerDecidedBy: raw?.pool?.winnerOption?.decidedBy ?? '',
            });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the appeal-opened event for a specific market.
     */
    onAppealOpened(marketId, callback) {
        const event = `appeal-opened/${marketId}`;
        const handler = (raw) => {
            callback({
                poolId: raw?.pool?._id ?? marketId,
                eventType: raw?.eventType ?? '',
                isAppealed: raw?.pool?.isAppealed ?? false,
            });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the dispute-time-extented event for a specific market.
     * Note: event name matches the backend spelling ("extented").
     */
    onDisputeTimeExtended(marketId, callback) {
        const event = `dispute-time-extented/${marketId}`;
        const handler = (raw) => {
            callback({
                poolId: raw?.pool?._id ?? marketId,
                eventType: raw?.eventType ?? '',
                newEndTime: raw?.newEndTime ?? '',
            });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the dispute-winner event for a specific market.
     * Fires when a dispute winner is decided.
     */
    onDisputeWinner(marketId, callback) {
        const event = `dispute-winner/${marketId}`;
        const handler = (raw) => {
            callback({
                poolId: raw?.pool?._id ?? marketId,
                status: raw?.pool?.status ?? '',
                winnerOption: {
                    decidedBy: raw?.pool?.winnerOption?.decidedBy ?? '',
                    choiceIndex: raw?.pool?.winnerOption?.choiceIndex ?? 0,
                    createdAt: raw?.pool?.winnerOption?.createdAt ?? '',
                },
            });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the appeal-winner event for a specific market.
     * Fires when an appeal winner is finalized.
     */
    onAppealWinner(marketId, callback) {
        const event = `appeal-winner/${marketId}`;
        const handler = (raw) => {
            callback({
                poolId: raw?.pool?._id ?? marketId,
                eventType: raw?.eventType ?? '',
                winnerFinalized: raw?.winnerFinalized ?? null,
                winnerOption: raw?.pool?.winnerOption ?? null,
            });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    /**
     * Subscribe to the claim-reward event for a specific market and user.
     * Fires when a user's reward claim is confirmed.
     */
    onClaimReward(marketId, userId, callback) {
        const event = `claim-reward/${marketId}/${userId}`;
        const handler = (_raw) => {
            callback({ poolId: marketId, userId });
        };
        this.socket.on(event, handler);
        return () => this.socket.off(event, handler);
    }
    disconnect() {
        this.socket.disconnect();
    }
}
