// @opinion-labs/opinion-clob-sdk/dist/builder/builder_client.js
/**
 * Builder SDK for Opinion CLOB
 *
 * This module provides the BuilderClient class for builder mode operations:
 * - Creating and managing user sub-account API keys
 * - Building order structures for user signing
 * - Placing orders on behalf of users with their signatures
 * - Building and submitting Safe transactions for chain operations
 */
import { createPublicClient, http, getAddress, getContract, encodeFunctionData, hashTypedData, } from 'viem';
import { bsc } from 'viem/chains';
import { privateKeyToAccount } from 'viem/accounts';
import { fetch as undiciFetch, ProxyAgent } from 'undici';
import { BuilderInvalidParamError, BuilderApiError } from './errors.js';
// Import generated API client and types for user operations
import { postOrderCancel as apiCancelOrder, getOrder as apiGetOrders, } from '@opinion-labs/opinion-api';
import { createClient, createConfig } from '@opinion-labs/opinion-api/client';
import { OrderSide, OrderType } from '../models/enums.js';
import { ZERO_ADDRESS } from '../models/order.js';
import { DEFAULT_CONTRACT_ADDRESSES, SUPPORTED_CHAIN_IDS, EIP712_DOMAIN, } from '../config.js';
import { safeAmountToWei, calculateOrderAmounts, generateSalt, } from '../chain/order_utils/utils.js';
import { encodeMultiSendCall } from '../chain/safe/multisend.js';
import { MultiSendOperation, NULL_HASH } from '../chain/safe/constants.js';
import { erc20Abi } from '../chain/contracts/erc20.js';
import { conditionalTokensAbi } from '../chain/contracts/conditional_tokens.js';
// Constants
const EOA_SIGNATURE_TYPE = 0;
const POLY_GNOSIS_SAFE_SIGNATURE_TYPE = 2;
/**
 * EIP-712 Order type definition
 */
const ORDER_TYPES = {
    Order: [
        { name: 'salt', type: 'uint256' },
        { name: 'maker', type: 'address' },
        { name: 'signer', type: 'address' },
        { name: 'taker', type: 'address' },
        { name: 'tokenId', type: 'uint256' },
        { name: 'makerAmount', type: 'uint256' },
        { name: 'takerAmount', type: 'uint256' },
        { name: 'expiration', type: 'uint256' },
        { name: 'nonce', type: 'uint256' },
        { name: 'feeRateBps', type: 'uint256' },
        { name: 'side', type: 'uint8' },
        { name: 'signatureType', type: 'uint8' },
    ],
};
/**
 * Safe EIP-712 types
 */
const SAFE_TX_TYPES = {
    SafeTx: [
        { name: 'to', type: 'address' },
        { name: 'value', type: 'uint256' },
        { name: 'data', type: 'bytes' },
        { name: 'operation', type: 'uint8' },
        { name: 'safeTxGas', type: 'uint256' },
        { name: 'baseGas', type: 'uint256' },
        { name: 'gasPrice', type: 'uint256' },
        { name: 'gasToken', type: 'address' },
        { name: 'refundReceiver', type: 'address' },
        { name: 'nonce', type: 'uint256' },
    ],
};
/**
 * Builder client for managing user API keys and placing orders on behalf of users.
 *
 * Builder mode allows:
 * 1. Creating sub-account API keys for users
 * 2. Building EIP-712 order structures for users to sign
 * 3. Submitting orders on behalf of users using their signatures
 *
 * @example
 * ```typescript
 * const client = new BuilderClient({
 *   host: 'https://api.example.com',
 *   builderApiKey: 'your-builder-apikey',
 *   chainId: 56,
 *   rpcUrl: 'https://bsc-dataseed.binance.org/'
 * });
 *
 * // Create API key for user
 * const result = await client.createUser('0x1234...');
 *
 * // Build order for user to sign
 * const orderData = await client.buildOrderForSigning({
 *   marketId: 123,
 *   tokenId: 'token_id',
 *   userWalletAddress: '0x1234...',
 *   side: OrderSide.BUY,
 *   orderType: OrderType.LIMIT_ORDER,
 *   amount: 100.0,
 *   price: '0.5'
 * });
 *
 * // User signs the order (off-chain)
 * const signature = '0x...'; // User's signature
 *
 * // Place order for user
 * const result = await client.placeOrderForUser({
 *   order: orderData.order,
 *   signature: signature,
 *   userWalletAddress: '0x1234...'
 * });
 * ```
 */
export class BuilderClient {
    host;
    builderApiKey;
    chainId;
    rpcUrl;
    useBeta;
    contractAddresses;
    proxyUrl;
    // Cache settings
    quoteTokensCacheTtl;
    marketCacheTtl;
    quoteTokensCache = null;
    marketCache = new Map();
    // Web3 instance (lazy initialized)
    publicClient = null;
    constructor(config) {
        if (!SUPPORTED_CHAIN_IDS.includes(config.chainId)) {
            throw new BuilderInvalidParamError(`chainId must be one of ${SUPPORTED_CHAIN_IDS.join(', ')}`);
        }
        this.chainId = config.chainId;
        this.host = config.host;
        this.builderApiKey = config.builderApiKey;
        this.rpcUrl = config.rpcUrl || '';
        this.useBeta = config.useBeta || false;
        this.proxyUrl = config.proxyUrl;
        this.quoteTokensCacheTtl = config.quoteTokensCacheTtl ?? 3600;
        this.marketCacheTtl = config.marketCacheTtl ?? 300;
        // Get base addresses from mainnet defaults
        const baseAddresses = {
            conditionalTokens: DEFAULT_CONTRACT_ADDRESSES[config.chainId].conditionalTokens,
            multisend: DEFAULT_CONTRACT_ADDRESSES[config.chainId].multisend,
            feeManager: DEFAULT_CONTRACT_ADDRESSES[config.chainId].feeManager,
        };
        // For beta/testnet, contract_addresses must be provided
        if (this.useBeta && !config.contractAddresses) {
            throw new BuilderInvalidParamError('contractAddresses is required for beta/testnet environments. ' +
                'Please provide addresses for: conditionalTokens, multisend, feeManager, weth');
        }
        // Apply custom overrides
        if (config.contractAddresses) {
            if (config.contractAddresses.conditionalTokens) {
                baseAddresses.conditionalTokens = config.contractAddresses.conditionalTokens;
            }
            if (config.contractAddresses.multisend) {
                baseAddresses.multisend = config.contractAddresses.multisend;
            }
            if (config.contractAddresses.feeManager) {
                baseAddresses.feeManager = config.contractAddresses.feeManager;
            }
            if (config.contractAddresses.weth) {
                baseAddresses.weth = config.contractAddresses.weth;
            }
            if (config.contractAddresses.ctfExchange) {
                baseAddresses.ctfExchange = config.contractAddresses.ctfExchange;
            }
        }
        this.contractAddresses = baseAddresses;
    }
    // ============================================================================
    // HTTP Request Helpers
    // ============================================================================
    /**
     * Make an API request with builder-apikey header.
     */
    async makeBuilderRequest(method, path, data, useApiKeyHeader = false) {
        const url = `${this.host}${path}`;
        // Use 'apikey' header for public endpoints, 'builder-apikey' for builder-specific endpoints
        const headers = useApiKeyHeader
            ? { apikey: this.builderApiKey, 'Content-Type': 'application/json' }
            : { 'builder-apikey': this.builderApiKey, 'Content-Type': 'application/json' };
        try {
            let response;
            if (this.proxyUrl) {
                const proxyDispatcher = new ProxyAgent(this.proxyUrl);
                if (method === 'GET') {
                    const queryParams = data
                        ? '?' + new URLSearchParams(data).toString()
                        : '';
                    const resp = await undiciFetch(url + queryParams, {
                        method: 'GET',
                        headers,
                        dispatcher: proxyDispatcher,
                    });
                    response = resp;
                }
                else {
                    const resp = await undiciFetch(url, {
                        method: 'POST',
                        headers,
                        body: JSON.stringify(data),
                        dispatcher: proxyDispatcher,
                    });
                    response = resp;
                }
            }
            else {
                if (method === 'GET') {
                    const queryParams = data
                        ? '?' + new URLSearchParams(data).toString()
                        : '';
                    response = await fetch(url + queryParams, {
                        method: 'GET',
                        headers,
                    });
                }
                else {
                    response = await fetch(url, {
                        method: 'POST',
                        headers,
                        body: JSON.stringify(data),
                    });
                }
            }
            if (!response.ok) {
                const errorText = await response.text();
                throw new BuilderApiError(`API request failed: ${response.status} ${errorText}`);
            }
            return (await response.json());
        }
        catch (error) {
            if (error instanceof BuilderApiError) {
                throw error;
            }
            throw new BuilderApiError(`API request failed: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    /**
     * Validate API response
     */
    validateResponse(response, operation) {
        if (response.errno !== 0) {
            throw new BuilderApiError(`Failed to ${operation}: ${response.errmsg || 'Unknown error'}`);
        }
        if (response.result === undefined) {
            throw new BuilderApiError(`Failed to ${operation}: No result in response`);
        }
        return response.result;
    }
    /**
     * Parse list response
     */
    parseListResponse(response, operationName) {
        if (response.errno !== 0) {
            throw new BuilderApiError(`Failed to ${operationName}: ${response.errmsg || 'Unknown error'}`);
        }
        if (!response.result?.list) {
            throw new BuilderApiError(`Invalid list response format for ${operationName}`);
        }
        return response.result.list;
    }
    /**
     * Validate market response
     */
    validateMarketResponse(response, operationName) {
        if (response.errno !== 0) {
            throw new BuilderApiError(`Failed to ${operationName}: ${response.errmsg || 'Unknown error'}`);
        }
        if (!response.result?.data) {
            throw new BuilderApiError(`Invalid response format for ${operationName}`);
        }
        return response.result.data;
    }
    // ============================================================================
    // User Management
    // ============================================================================
    /**
     * Create a user sub-account under this builder.
     *
     * This method:
     * 1. Registers the user in the system if not exists
     * 2. Initiates Gnosis Safe multi-sig wallet creation (platform pays gas)
     * 3. Generates an API key for the user
     *
     * IMPORTANT: The API key is returned ONLY ONCE during creation.
     * Store it securely! If lost, use regenerateUserApiKey() to get a new one.
     *
     * @param walletAddress - User's login wallet address (EOA, 0x...)
     * @returns User info including apikey
     *
     * @example
     * ```typescript
     * // Create user and SAVE the apikey
     * const result = await builder.createUser('0x1234...');
     * const userApikey = result.apikey; // SAVE THIS SECURELY!
     * ```
     */
    async createUser(walletAddress) {
        if (!walletAddress || !walletAddress.startsWith('0x')) {
            throw new BuilderInvalidParamError('Invalid wallet address format');
        }
        const normalizedAddress = walletAddress.toLowerCase();
        const response = await this.makeBuilderRequest('POST', `/builder/user/${normalizedAddress}`);
        const result = this.validateResponse(response, 'create user');
        const apikey = result.apiKey || '';
        if (apikey) {
            console.warn('IMPORTANT: Store this API key securely! It will NOT be returned again. ' +
                'If lost, use regenerateUserApiKey() to generate a new one.');
        }
        return {
            apikey,
            walletAddress: result.walletAddress || normalizedAddress,
            builderName: result.builderName || '',
            multiSigWallet: result.multiSigWallet || '',
            walletCreationTxHash: result.walletCreationTxHash || '',
            enableTrading: result.enableTrading || false,
        };
    }
    /**
     * Get user information for a builder's sub-account by login wallet address.
     *
     * NOTE: API key is NOT returned for security reasons.
     * If you need the apikey, you should have saved it when createUser() was called.
     * If lost, use regenerateUserApiKey() to generate a new one.
     *
     * @param walletAddress - User's login wallet address (EOA, 0x...)
     * @returns User info (without apikey)
     */
    async getUser(walletAddress) {
        if (!walletAddress || !walletAddress.startsWith('0x')) {
            throw new BuilderInvalidParamError('Invalid wallet address format');
        }
        const normalizedAddress = walletAddress.toLowerCase();
        const response = await this.makeBuilderRequest('GET', `/builder/user/${normalizedAddress}`);
        const result = this.validateResponse(response, 'get user');
        return {
            walletAddress: result.walletAddress || normalizedAddress,
            builderName: result.builderName || '',
            multiSigWallet: result.multiSigWallet || '',
            enableTrading: result.enableTrading || false,
        };
    }
    /**
     * Regenerate API key for a user. The old key is invalidated immediately.
     *
     * IMPORTANT: The new API key is returned ONLY ONCE. Store it securely!
     *
     * @param walletAddress - User's login wallet address (EOA, 0x...)
     * @returns User info including new apikey
     */
    async regenerateUserApiKey(walletAddress) {
        if (!walletAddress || !walletAddress.startsWith('0x')) {
            throw new BuilderInvalidParamError('Invalid wallet address format');
        }
        const normalizedAddress = walletAddress.toLowerCase();
        const response = await this.makeBuilderRequest('POST', `/builder/user/${normalizedAddress}/regenerate-apikey`);
        const result = this.validateResponse(response, 'regenerate user apikey');
        const apikey = result.apiKey || '';
        if (apikey) {
            console.warn('IMPORTANT: Store this new API key securely! It will NOT be returned again.');
        }
        return {
            apikey,
            walletAddress: result.walletAddress || normalizedAddress,
            builderName: result.builderName || '',
            multiSigWallet: result.multiSigWallet || '',
            enableTrading: result.enableTrading || false,
        };
    }
    // ============================================================================
    // Market Data (Read-only)
    // ============================================================================
    /**
     * Get list of supported quote tokens
     */
    async getQuoteTokens(useCache = true) {
        const currentTime = Date.now();
        if (useCache && this.quoteTokensCacheTtl > 0 && this.quoteTokensCache) {
            const cacheAge = currentTime - this.quoteTokensCache.timestamp;
            if (cacheAge < this.quoteTokensCacheTtl * 1000) {
                return this.quoteTokensCache.data;
            }
        }
        // Use apikey header for public market data endpoints
        const response = await this.makeBuilderRequest('GET', '/quoteToken', { chainId: String(this.chainId) }, true // useApiKeyHeader
        );
        const list = this.parseListResponse(response, 'get quote tokens');
        if (this.quoteTokensCacheTtl > 0) {
            this.quoteTokensCache = { data: list, timestamp: currentTime };
        }
        return list;
    }
    /**
     * Get detailed information about a specific market
     */
    async getMarket(marketId, useCache = true) {
        if (!marketId || marketId <= 0) {
            throw new BuilderInvalidParamError('marketId must be a positive integer');
        }
        const currentTime = Date.now();
        if (useCache && this.marketCacheTtl > 0) {
            const cached = this.marketCache.get(marketId);
            if (cached) {
                const cacheAge = currentTime - cached.timestamp;
                if (cacheAge < this.marketCacheTtl * 1000) {
                    return cached.data;
                }
            }
        }
        // Use apikey header for public market data endpoints
        const response = await this.makeBuilderRequest('GET', `/market/${marketId}`, undefined, true // useApiKeyHeader
        );
        const market = this.validateMarketResponse(response, 'get market');
        if (this.marketCacheTtl > 0) {
            this.marketCache.set(marketId, { data: market, timestamp: currentTime });
        }
        return market;
    }
    /**
     * Get orderbook for a specific token
     */
    async getOrderbook(tokenId) {
        if (!tokenId) {
            throw new BuilderInvalidParamError('tokenId is required');
        }
        // Use apikey header for public market data endpoints
        const response = await this.makeBuilderRequest('GET', '/token/orderbook', { token_id: tokenId }, true // useApiKeyHeader
        );
        return response;
    }
    // ============================================================================
    // Order Building
    // ============================================================================
    /**
     * Build an order structure for a user to sign.
     *
     * @param options - Order options
     * @returns Order data for signing
     */
    async buildOrderForSigning(options) {
        const { marketId, tokenId, userWalletAddress, side, orderType, amount, price: priceInput, amountType = 'quote', signerAddress: signerAddressInput, } = options;
        if (!marketId || marketId <= 0) {
            throw new BuilderInvalidParamError('marketId must be a positive integer');
        }
        if (!tokenId) {
            throw new BuilderInvalidParamError('tokenId is required');
        }
        if (!userWalletAddress || !userWalletAddress.startsWith('0x')) {
            throw new BuilderInvalidParamError('Invalid userWalletAddress format');
        }
        // Validate price requirement for LIMIT_ORDER
        if (orderType === OrderType.LIMIT_ORDER && !priceInput) {
            throw new BuilderInvalidParamError('price is required for LIMIT_ORDER');
        }
        // Market order: price is ignored and set to "0"
        // Also validate amount_type restrictions for market orders
        let price = priceInput || '0';
        if (orderType === OrderType.MARKET_ORDER) {
            price = '0';
            // Market BUY: only quote amount allowed
            if (side === OrderSide.BUY && amountType === 'base') {
                throw new BuilderInvalidParamError("amountType='base' is not allowed for market BUY orders");
            }
            // Market SELL: only base amount allowed
            if (side === OrderSide.SELL && amountType === 'quote') {
                throw new BuilderInvalidParamError("amountType='quote' is not allowed for market SELL orders");
            }
        }
        // maker is always the userWalletAddress (Safe address or EOA)
        const makerAddress = getAddress(userWalletAddress);
        // signer is either explicitly provided (Safe owner) or same as maker (EOA mode)
        let signerAddr;
        if (signerAddressInput) {
            if (!signerAddressInput.startsWith('0x')) {
                throw new BuilderInvalidParamError('Invalid signerAddress format');
            }
            signerAddr = getAddress(signerAddressInput);
        }
        else {
            signerAddr = makerAddress;
        }
        // Get market and quote token info
        const market = await this.getMarket(marketId);
        const marketData = market;
        if (parseInt(marketData.chainId ?? '') !== this.chainId) {
            throw new BuilderApiError('Cannot build order for different chain');
        }
        const quoteTokenAddr = marketData.quoteToken ?? '';
        // Get quote token details
        const quoteTokenList = await this.getQuoteTokens();
        const quoteToken = quoteTokenList.find((item) => {
            const qt = item;
            return qt.quoteTokenAddress?.toLowerCase() === quoteTokenAddr?.toLowerCase();
        });
        if (!quoteToken) {
            throw new BuilderApiError('Quote token not found for this market');
        }
        const qt = quoteToken;
        const exchangeAddr = getAddress(qt.ctfExchangeAddress ?? '');
        const currencyDecimal = quoteToken.decimal;
        // Calculate maker and taker amounts
        const priceDecimal = parseFloat(price);
        let makerAmount;
        let takerAmount;
        if (orderType === OrderType.MARKET_ORDER) {
            // Market order: takerAmount is 0, makerAmount is direct conversion
            takerAmount = 0n;
            makerAmount = safeAmountToWei(amount, currencyDecimal);
        }
        else {
            // LIMIT_ORDER
            let makerAmountWei;
            if (amountType === 'quote') {
                makerAmountWei = safeAmountToWei(amount, currencyDecimal);
            }
            else {
                // Convert base to quote for buy, keep base for sell
                if (side === OrderSide.BUY) {
                    makerAmountWei = safeAmountToWei(amount * priceDecimal, currencyDecimal);
                }
                else {
                    makerAmountWei = safeAmountToWei(amount, currencyDecimal);
                }
            }
            const amounts = calculateOrderAmounts(priceDecimal, makerAmountWei, side, currencyDecimal);
            makerAmount = amounts.recalculatedMakerAmount;
            takerAmount = amounts.takerAmount;
        }
        // Generate order fields
        const salt = generateSalt();
        const expiration = Math.floor(Date.now() / 1000) + 86400 * 30; // 30 days from now
        const nonce = 0;
        const feeRateBps = '0';
        // Determine signature type:
        // - If maker == signer: EOA mode (signatureType = 0)
        // - If maker != signer: Safe mode (signatureType = 2, POLY_GNOSIS_SAFE)
        const isSafeMode = makerAddress.toLowerCase() !== signerAddr.toLowerCase();
        const signatureType = isSafeMode ? POLY_GNOSIS_SAFE_SIGNATURE_TYPE : EOA_SIGNATURE_TYPE;
        // Build order structure
        const order = {
            salt: salt.toString(),
            maker: makerAddress,
            signer: signerAddr,
            taker: ZERO_ADDRESS,
            tokenId: tokenId,
            makerAmount: makerAmount.toString(),
            takerAmount: takerAmount.toString(),
            expiration: expiration.toString(),
            nonce: nonce.toString(),
            feeRateBps: feeRateBps,
            side: side.toString(),
            signatureType: signatureType.toString(),
        };
        // Build EIP-712 domain
        const domain = {
            name: EIP712_DOMAIN.name,
            version: EIP712_DOMAIN.version,
            chainId: this.chainId,
            verifyingContract: exchangeAddr,
        };
        // Prepare message for signing
        const message = {
            salt: BigInt(order.salt),
            maker: order.maker,
            signer: order.signer,
            taker: order.taker,
            tokenId: BigInt(order.tokenId),
            makerAmount: BigInt(order.makerAmount),
            takerAmount: BigInt(order.takerAmount),
            expiration: BigInt(order.expiration),
            nonce: BigInt(order.nonce),
            feeRateBps: BigInt(order.feeRateBps),
            side: parseInt(order.side),
            signatureType: parseInt(order.signatureType),
        };
        // Calculate struct hash using viem
        const structHash = hashTypedData({
            domain: {
                ...domain,
                chainId: BigInt(domain.chainId),
            },
            types: ORDER_TYPES,
            primaryType: 'Order',
            message,
        });
        // Build typedData for reference
        // IMPORTANT: Use BigInt for large numeric fields (salt, tokenId, amounts) to avoid precision loss
        // JavaScript's parseInt() can only handle numbers up to 2^53-1, but these fields are uint256
        const typedData = {
            types: {
                EIP712Domain: [
                    { name: 'name', type: 'string' },
                    { name: 'version', type: 'string' },
                    { name: 'chainId', type: 'uint256' },
                    { name: 'verifyingContract', type: 'address' },
                ],
                ...ORDER_TYPES,
            },
            primaryType: 'Order',
            domain: {
                name: domain.name,
                version: domain.version,
                chainId: domain.chainId,
                verifyingContract: domain.verifyingContract,
            },
            message: {
                salt: BigInt(order.salt),
                maker: order.maker,
                signer: order.signer,
                taker: order.taker,
                tokenId: BigInt(order.tokenId),
                makerAmount: BigInt(order.makerAmount),
                takerAmount: BigInt(order.takerAmount),
                expiration: BigInt(order.expiration),
                nonce: BigInt(order.nonce),
                feeRateBps: BigInt(order.feeRateBps),
                side: parseInt(order.side),
                signatureType: parseInt(order.signatureType),
            },
        };
        return {
            order,
            domain,
            structHash,
            typedData,
            exchangeAddress: exchangeAddr,
            currencyAddress: quoteTokenAddr,
            currencyDecimal: currencyDecimal,
            marketId,
            orderType,
            price,
        };
    }
    // ============================================================================
    // Order Placement
    // ============================================================================
    /**
     * Place an order on behalf of a user using their signature.
     *
     * @param options - Order placement options
     * @returns API response with order details
     */
    async placeOrderForUser(options) {
        const { order, signature, userWalletAddress, marketId, orderType, price, currencyAddress } = options;
        if (!signature || !signature.startsWith('0x')) {
            throw new BuilderInvalidParamError('signature must be a hex string starting with 0x');
        }
        if (!userWalletAddress || !userWalletAddress.startsWith('0x')) {
            throw new BuilderInvalidParamError('Invalid userWalletAddress format');
        }
        const normalizedAddress = userWalletAddress.toLowerCase();
        // Validate that order maker matches userWalletAddress
        if (order.maker?.toLowerCase() !== normalizedAddress) {
            throw new BuilderInvalidParamError('Order maker must match userWalletAddress');
        }
        // Validate signature type
        const sigType = order.signatureType;
        if (sigType !== String(EOA_SIGNATURE_TYPE) &&
            sigType !== String(POLY_GNOSIS_SAFE_SIGNATURE_TYPE)) {
            throw new BuilderInvalidParamError('signatureType must be 0 (EOA) or 2 (POLY_GNOSIS_SAFE) for builder orders');
        }
        // Generate unique order sign
        const orderSign = signature.length > 66 ? signature.slice(2, 66) : signature.slice(2);
        // Build request payload
        const requestData = {
            order: {
                salt: order.salt || '',
                maker: order.maker || '',
                signer: order.signer || '',
                taker: order.taker || ZERO_ADDRESS,
                tokenId: order.tokenId || '',
                makerAmount: order.makerAmount || '',
                takerAmount: order.takerAmount || '',
                expiration: order.expiration || '',
                nonce: order.nonce || '0',
                feeRateBps: order.feeRateBps || '0',
                side: order.side || '0',
                signatureType: order.signatureType || '0',
                topicId: marketId || 0,
                price: price || '0',
                tradingMethod: orderType || OrderType.LIMIT_ORDER,
                currencyAddress: currencyAddress || '',
                sign: orderSign,
                timestamp: Math.floor(Date.now() / 1000),
                safeRate: '0',
                orderExpTime: order.expiration || '0',
            },
            signature,
            maker: normalizedAddress,
        };
        const response = await this.makeBuilderRequest('POST', '/builder/order', requestData);
        return this.validateResponse(response, 'place order for user');
    }
    /**
     * Place an order using the result from buildOrderForSigning.
     * This is a convenience method that extracts all necessary fields.
     */
    async placeOrderForUserFromBuildResult(buildResult, signature, userWalletAddress) {
        return this.placeOrderForUser({
            order: buildResult.order,
            signature,
            userWalletAddress,
            marketId: buildResult.marketId,
            orderType: buildResult.orderType,
            price: buildResult.price,
            currencyAddress: buildResult.currencyAddress,
        });
    }
    // ============================================================================
    // Utility Methods
    // ============================================================================
    /**
     * Sign an order with a private key (for testing purposes).
     *
     * In production, the user should sign the order themselves using their wallet.
     *
     * @param typedData - EIP-712 typed data from buildOrderForSigning
     * @param privateKey - Private key (hex string)
     * @returns Signature as hex string
     */
    static async signOrderWithPrivateKey(typedData, privateKey) {
        try {
            const account = privateKeyToAccount(privateKey);
            const signature = await account.signTypedData({
                domain: {
                    name: typedData.domain.name,
                    version: typedData.domain.version,
                    chainId: BigInt(typedData.domain.chainId),
                    verifyingContract: typedData.domain.verifyingContract,
                },
                types: typedData.types,
                primaryType: typedData.primaryType,
                message: typedData.message,
            });
            return signature;
        }
        catch (e) {
            throw new BuilderApiError(`Failed to sign order: ${e instanceof Error ? e.message : String(e)}`);
        }
    }
    // ============================================================================
    // Safe Transaction Operations
    // ============================================================================
    /**
     * Get or create PublicClient instance
     */
    getPublicClient() {
        if (!this.publicClient) {
            if (!this.rpcUrl) {
                throw new BuilderInvalidParamError('rpcUrl is required for Safe operations');
            }
            let fetchOptions;
            if (this.proxyUrl) {
                const proxyDispatcher = new ProxyAgent(this.proxyUrl);
                fetchOptions = {
                    fetchOptions: {
                        dispatcher: proxyDispatcher,
                    },
                };
            }
            this.publicClient = createPublicClient({
                chain: bsc,
                transport: http(this.rpcUrl, fetchOptions),
            });
        }
        return this.publicClient;
    }
    /**
     * Get Safe nonce for a given address
     */
    async getSafeNonce(safeAddress) {
        const publicClient = this.getPublicClient();
        const safeContract = getContract({
            address: safeAddress,
            abi: [
                {
                    inputs: [],
                    name: 'nonce',
                    outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
                    stateMutability: 'view',
                    type: 'function',
                },
            ],
            client: publicClient,
        });
        return await safeContract.read.nonce();
    }
    /**
     * Build an Enable Trading transaction for user to sign.
     *
     * Creates approval transactions for:
     * 1. ERC20 tokens to CTF Exchange
     * 2. ERC20 tokens to ConditionalTokens contract
     * 3. ConditionalTokens to CTF Exchange (setApprovalForAll)
     *
     * @param safeAddress - User's Safe wallet address
     * @param quoteTokens - Dict of {tokenAddress: ctfExchangeAddress}
     * @returns Safe transaction data for signing
     */
    async buildEnableTradingTx(safeAddress, quoteTokens) {
        const normalizedSafeAddress = getAddress(safeAddress);
        // Get quote tokens if not provided
        const tokenMap = quoteTokens ?? {};
        if (!quoteTokens) {
            const quoteTokenList = await this.getQuoteTokens();
            for (const token of quoteTokenList) {
                const qt = token;
                const tokenChainId = qt.chainId;
                const tokenAddress = qt.quoteTokenAddress;
                const exchangeAddress = qt.ctfExchangeAddress;
                if (tokenChainId === String(this.chainId) && tokenAddress && exchangeAddress) {
                    tokenMap[tokenAddress] = exchangeAddress;
                }
            }
        }
        const conditionalTokensAddr = this.contractAddresses.conditionalTokens;
        if (!conditionalTokensAddr) {
            throw new BuilderInvalidParamError(`conditionalTokens address not found for chain ${this.chainId}`);
        }
        const multiSendTxs = [];
        const maxApproval = 2n ** 256n - 1n;
        for (const [erc20Address, ctfExchangeAddress] of Object.entries(tokenMap)) {
            const erc20Addr = getAddress(erc20Address);
            const ctfExchangeAddr = getAddress(ctfExchangeAddress);
            // Approve CTF Exchange
            const approveExchangeData = encodeFunctionData({
                abi: erc20Abi,
                functionName: 'approve',
                args: [ctfExchangeAddr, maxApproval],
            });
            multiSendTxs.push({
                operation: MultiSendOperation.CALL,
                to: erc20Addr,
                value: 0n,
                data: approveExchangeData,
            });
            // Approve ConditionalTokens
            const approveCTData = encodeFunctionData({
                abi: erc20Abi,
                functionName: 'approve',
                args: [conditionalTokensAddr, maxApproval],
            });
            multiSendTxs.push({
                operation: MultiSendOperation.CALL,
                to: erc20Addr,
                value: 0n,
                data: approveCTData,
            });
            // setApprovalForAll for ConditionalTokens to CTF Exchange
            const setApprovalData = encodeFunctionData({
                abi: conditionalTokensAbi,
                functionName: 'setApprovalForAll',
                args: [ctfExchangeAddr, true],
            });
            multiSendTxs.push({
                operation: MultiSendOperation.CALL,
                to: conditionalTokensAddr,
                value: 0n,
                data: setApprovalData,
            });
        }
        if (multiSendTxs.length === 0) {
            throw new BuilderInvalidParamError('No transactions to build for enable trading');
        }
        return this.buildSafeTxResult(normalizedSafeAddress, multiSendTxs, 300000n);
    }
    /**
     * Build a Withdraw (ERC20 transfer) transaction for user to sign.
     */
    async buildWithdrawTx(safeAddress, tokenAddress, amount, toAddress) {
        const normalizedSafeAddress = getAddress(safeAddress);
        const tokenAddr = getAddress(tokenAddress);
        const toAddr = getAddress(toAddress);
        const transferData = encodeFunctionData({
            abi: erc20Abi,
            functionName: 'transfer',
            args: [toAddr, amount],
        });
        const txs = [
            {
                operation: MultiSendOperation.CALL,
                to: tokenAddr,
                value: 0n,
                data: transferData,
            },
        ];
        return this.buildSafeTxResult(normalizedSafeAddress, txs, 100000n);
    }
    /**
     * Build a Split Position transaction for user to sign.
     */
    async buildSplitTx(safeAddress, collateralToken, conditionId, amount, partition = [1, 2]) {
        const normalizedSafeAddress = getAddress(safeAddress);
        const collateralAddr = getAddress(collateralToken);
        const conditionIdBytes = conditionId.startsWith('0x') ? conditionId : `0x${conditionId}`;
        const conditionalTokensAddr = this.contractAddresses.conditionalTokens;
        const splitData = encodeFunctionData({
            abi: conditionalTokensAbi,
            functionName: 'splitPosition',
            args: [
                collateralAddr,
                NULL_HASH,
                conditionIdBytes,
                partition.map((p) => BigInt(p)),
                amount,
            ],
        });
        const txs = [
            {
                operation: MultiSendOperation.CALL,
                to: conditionalTokensAddr,
                value: 0n,
                data: splitData,
            },
        ];
        return this.buildSafeTxResult(normalizedSafeAddress, txs, 300000n);
    }
    /**
     * Build a Merge Position transaction for user to sign.
     */
    async buildMergeTx(safeAddress, collateralToken, conditionId, amount, partition = [1, 2]) {
        const normalizedSafeAddress = getAddress(safeAddress);
        const collateralAddr = getAddress(collateralToken);
        const conditionIdBytes = conditionId.startsWith('0x') ? conditionId : `0x${conditionId}`;
        const conditionalTokensAddr = this.contractAddresses.conditionalTokens;
        const mergeData = encodeFunctionData({
            abi: conditionalTokensAbi,
            functionName: 'mergePositions',
            args: [
                collateralAddr,
                NULL_HASH,
                conditionIdBytes,
                partition.map((p) => BigInt(p)),
                amount,
            ],
        });
        const txs = [
            {
                operation: MultiSendOperation.CALL,
                to: conditionalTokensAddr,
                value: 0n,
                data: mergeData,
            },
        ];
        return this.buildSafeTxResult(normalizedSafeAddress, txs, 300000n);
    }
    /**
     * Build a Redeem Position transaction for user to sign.
     */
    async buildRedeemTx(safeAddress, collateralToken, conditionId, partition = [1, 2]) {
        const normalizedSafeAddress = getAddress(safeAddress);
        const collateralAddr = getAddress(collateralToken);
        const conditionIdBytes = conditionId.startsWith('0x') ? conditionId : `0x${conditionId}`;
        const conditionalTokensAddr = this.contractAddresses.conditionalTokens;
        const redeemData = encodeFunctionData({
            abi: conditionalTokensAbi,
            functionName: 'redeemPositions',
            args: [collateralAddr, NULL_HASH, conditionIdBytes, partition.map((p) => BigInt(p))],
        });
        const txs = [
            {
                operation: MultiSendOperation.CALL,
                to: conditionalTokensAddr,
                value: 0n,
                data: redeemData,
            },
        ];
        return this.buildSafeTxResult(normalizedSafeAddress, txs, 300000n);
    }
    /**
     * Build the Safe transaction result
     */
    async buildSafeTxResult(safeAddress, txs, safeTxGas) {
        const multisendAddr = this.contractAddresses.multisend;
        const multiSendData = encodeMultiSendCall(txs);
        const nonce = await this.getSafeNonce(safeAddress);
        const safeTx = {
            to: multisendAddr,
            value: 0n,
            data: multiSendData,
            operation: 1, // DELEGATE_CALL
            safeTxGas,
            baseGas: 0n,
            gasPrice: 0n,
            gasToken: ZERO_ADDRESS,
            refundReceiver: ZERO_ADDRESS,
            nonce,
        };
        // Build EIP-712 data for signing
        const domain = {
            chainId: BigInt(this.chainId),
            verifyingContract: safeAddress,
        };
        const message = {
            to: safeTx.to,
            value: safeTx.value,
            data: safeTx.data,
            operation: safeTx.operation,
            safeTxGas: safeTx.safeTxGas,
            baseGas: safeTx.baseGas,
            gasPrice: safeTx.gasPrice,
            gasToken: safeTx.gasToken,
            refundReceiver: safeTx.refundReceiver,
            nonce: safeTx.nonce,
        };
        const safeTxHash = hashTypedData({
            domain,
            types: SAFE_TX_TYPES,
            primaryType: 'SafeTx',
            message,
        });
        const eip712Data = {
            types: {
                EIP712Domain: [
                    { name: 'chainId', type: 'uint256' },
                    { name: 'verifyingContract', type: 'address' },
                ],
                ...SAFE_TX_TYPES,
            },
            primaryType: 'SafeTx',
            domain: {
                chainId: this.chainId,
                verifyingContract: safeAddress,
            },
            message: {
                to: safeTx.to,
                value: safeTx.value.toString(),
                data: safeTx.data,
                operation: safeTx.operation,
                safeTxGas: safeTx.safeTxGas.toString(),
                baseGas: safeTx.baseGas.toString(),
                gasPrice: safeTx.gasPrice.toString(),
                gasToken: safeTx.gasToken,
                refundReceiver: safeTx.refundReceiver,
                nonce: safeTx.nonce.toString(),
            },
        };
        return {
            safeTx,
            eip712Data,
            safeTxHash,
            submissionDataTemplate: this.toSubmissionData(safeTx, safeAddress, ''),
        };
    }
    /**
     * Convert SafeTx to submission data format
     */
    toSubmissionData(safeTx, safeAddress, signature) {
        // Remove 0x prefix from signature if present
        const sig = signature.startsWith('0x') ? signature.slice(2) : signature;
        // Remove 0x prefix from data if present
        const data = safeTx.data.startsWith('0x') ? safeTx.data.slice(2) : safeTx.data;
        return {
            toAddress: safeTx.to,
            value: safeTx.value.toString(),
            data: data,
            operation: safeTx.operation,
            safeTxGas: safeTx.safeTxGas.toString(),
            baseGas: safeTx.baseGas.toString(),
            gasPrice: safeTx.gasPrice.toString(),
            gasToken: safeTx.gasToken === ZERO_ADDRESS ? '' : safeTx.gasToken,
            refundReceiver: safeTx.refundReceiver === ZERO_ADDRESS ? '' : safeTx.refundReceiver,
            nonce: safeTx.nonce.toString(),
            signatures: sig,
            safe: safeAddress,
        };
    }
    /**
     * Submit a signed Safe transaction to the backend.
     *
     * @param walletAddress - User's wallet address (Safe owner)
     * @param safeTxResult - Result from build_*_tx methods
     * @param signature - User's signature on the Safe transaction (hex string)
     * @returns API response containing safeTxHash
     */
    async submitSafeTx(walletAddress, safeTxResult, signature) {
        if (!walletAddress || !walletAddress.startsWith('0x')) {
            throw new BuilderInvalidParamError('Invalid walletAddress format');
        }
        if (!signature) {
            throw new BuilderInvalidParamError('signature is required');
        }
        const normalizedAddress = walletAddress.toLowerCase();
        const safeAddress = getAddress(walletAddress);
        const submissionData = this.toSubmissionData(safeTxResult.safeTx, safeAddress, signature);
        const response = await this.makeBuilderRequest('POST', `/builder/safe/${normalizedAddress}/tx`, submissionData);
        return this.validateResponse(response, 'submit safe transaction');
    }
    /**
     * Sign a Safe transaction with a private key (for testing purposes).
     *
     * In production, the user should sign the transaction themselves using their wallet.
     *
     * @param eip712Data - EIP-712 data from build_*_tx methods
     * @param privateKey - Private key (hex string)
     * @returns Signature as hex string
     */
    static async signSafeTxWithPrivateKey(eip712Data, privateKey) {
        try {
            const account = privateKeyToAccount(privateKey);
            const signature = await account.signTypedData({
                domain: {
                    chainId: BigInt(eip712Data.domain.chainId),
                    verifyingContract: eip712Data.domain.verifyingContract,
                },
                types: eip712Data.types,
                primaryType: eip712Data.primaryType,
                message: eip712Data.message,
            });
            return signature;
        }
        catch (e) {
            throw new BuilderApiError(`Failed to sign Safe transaction: ${e instanceof Error ? e.message : String(e)}`);
        }
    }
    // ============================================================================
    // Order Cancellation (using user's apikey)
    // ============================================================================
    /**
     * Create an API client configured with user's apikey.
     *
     * @param userApiKey - User's API key
     * @returns API client instance for user operations
     */
    createUserApiClient(userApiKey) {
        let customFetch;
        if (this.proxyUrl) {
            const proxyDispatcher = new ProxyAgent(this.proxyUrl);
            customFetch = async (input, init) => {
                const request = new Request(input, init);
                const response = await undiciFetch(request.url, {
                    method: request.method,
                    headers: Object.fromEntries(request.headers.entries()),
                    body: request.body ? await request.text() : undefined,
                    dispatcher: proxyDispatcher,
                });
                return response;
            };
        }
        return createClient(createConfig({
            baseUrl: this.host,
            headers: {
                apikey: userApiKey,
            },
            fetch: customFetch,
        }));
    }
    /**
     * Cancel an order for a user using their API key.
     *
     * @param userApiKey - User's API key (obtained via getUser or createUser)
     * @param orderId - Order ID to cancel
     * @returns API response with cancellation result
     */
    async cancelOrderForUser(userApiKey, orderId) {
        if (!userApiKey) {
            throw new BuilderInvalidParamError('userApiKey is required');
        }
        if (!orderId || typeof orderId !== 'string') {
            throw new BuilderInvalidParamError('orderId must be a non-empty string');
        }
        const userApiClient = this.createUserApiClient(userApiKey);
        const response = await apiCancelOrder({
            client: userApiClient,
            headers: { apikey: userApiKey },
            body: { orderId },
        });
        if (response.error) {
            const errorData = response.error;
            throw new BuilderApiError(`Cancel order failed: ${errorData?.errmsg || errorData?.message || 'Unknown error'}`);
        }
        return { result: true };
    }
    /**
     * Cancel multiple orders for a user in batch.
     *
     * @param userApiKey - User's API key
     * @param orderIds - List of order IDs to cancel
     * @returns List of results
     */
    async cancelOrdersBatchForUser(userApiKey, orderIds) {
        if (!userApiKey) {
            throw new BuilderInvalidParamError('userApiKey is required');
        }
        if (!orderIds || !Array.isArray(orderIds) || orderIds.length === 0) {
            throw new BuilderInvalidParamError('orderIds must be a non-empty array');
        }
        const results = [];
        for (let i = 0; i < orderIds.length; i++) {
            const orderId = orderIds[i];
            try {
                const result = await this.cancelOrderForUser(userApiKey, orderId);
                results.push({
                    index: i,
                    success: true,
                    result,
                    orderId: orderId,
                });
            }
            catch (error) {
                const errorMsg = error instanceof Error ? error.message : String(error);
                console.error(`Failed to cancel order ${orderId}: ${errorMsg}`);
                results.push({
                    index: i,
                    success: false,
                    error: errorMsg,
                    orderId: orderId,
                });
            }
        }
        const errors = results.filter((r) => !r.success);
        if (errors.length > 0) {
            console.warn(`Batch order cancellation completed with ${errors.length} errors out of ${orderIds.length} orders`);
        }
        return results;
    }
    /**
     * Get orders for a user using their API key.
     *
     * @param userApiKey - User's API key
     * @param options - Query options
     * @returns API response containing order list
     */
    async getUserOrders(userApiKey, options = {}) {
        if (!userApiKey) {
            throw new BuilderInvalidParamError('userApiKey is required');
        }
        const { marketId = 0, status = '', limit = 10, page = 1 } = options;
        const userApiClient = this.createUserApiClient(userApiKey);
        const response = await apiGetOrders({
            client: userApiClient,
            headers: { apikey: userApiKey },
            query: {
                chainId: String(this.chainId),
                page,
                limit: Math.min(limit, 20),
                marketId: marketId > 0 ? marketId : undefined,
                status: status || undefined,
            },
        });
        if (response.error) {
            const errorData = response.error;
            throw new BuilderApiError(`Get orders failed: ${errorData?.errmsg || errorData?.message || 'Unknown error'}`);
        }
        return response.data;
    }
    /**
     * Cancel all open orders for a user.
     *
     * @param userApiKey - User's API key
     * @param options - Filter options
     * @returns Cancellation summary
     */
    async cancelAllOrdersForUser(userApiKey, options = {}) {
        if (!userApiKey) {
            throw new BuilderInvalidParamError('userApiKey is required');
        }
        const { marketId, side } = options;
        // Collect all open orders using pagination
        const allOrders = [];
        let currentPage = 1;
        const limit = 20;
        const maxPages = 100;
        while (currentPage <= maxPages) {
            const ordersResponse = await this.getUserOrders(userApiKey, {
                marketId: marketId || 0,
                status: '1', // 1 = pending/open orders
                limit,
                page: currentPage,
            });
            const result = ordersResponse.result;
            const orderList = result?.list || [];
            if (orderList.length === 0) {
                break;
            }
            for (const order of orderList) {
                const orderId = order.orderId;
                if (orderId) {
                    allOrders.push({ orderId, side: order.side ?? 0 });
                }
            }
            if (orderList.length < limit) {
                break;
            }
            currentPage++;
        }
        if (currentPage > maxPages) {
            console.warn(`Reached maximum page limit (${maxPages}), there may be more orders`);
        }
        if (allOrders.length === 0) {
            console.info('No open orders to cancel');
            return {
                totalOrders: 0,
                cancelled: 0,
                failed: 0,
                results: [],
            };
        }
        // Filter by side if specified
        let filteredOrders = allOrders;
        if (side !== undefined) {
            filteredOrders = allOrders.filter((order) => order.side === side);
        }
        // Extract order IDs
        const orderIds = filteredOrders.map((order) => order.orderId);
        if (orderIds.length === 0) {
            console.info('No orders match the filter criteria');
            return {
                totalOrders: 0,
                cancelled: 0,
                failed: 0,
                results: [],
            };
        }
        console.info(`Found ${orderIds.length} orders to cancel`);
        // Cancel all orders
        const results = await this.cancelOrdersBatchForUser(userApiKey, orderIds);
        // Count successes and failures
        const cancelled = results.filter((r) => r.success).length;
        const failed = results.filter((r) => !r.success).length;
        console.info(`Cancelled ${cancelled} orders, ${failed} failed out of ${orderIds.length} total`);
        return {
            totalOrders: orderIds.length,
            cancelled,
            failed,
            results,
        };
    }
}
