// @opinion-labs/opinion-clob-sdk/dist/builder/user_client.js
/**
 * UserClient - Simulates a real user's wallet client
 *
 * Used for testing Builder mode. UserClient holds the user's private key and can sign orders.
 * In production, signing is done by the user's wallet (e.g., MetaMask).
 */
import { serializeSignature } from 'viem';
import { privateKeyToAccount, generatePrivateKey, sign } from 'viem/accounts';
import * as fs from 'fs';
/**
 * UserClient simulates a real user's wallet client.
 *
 * Holds user's private key and can sign EIP-712 typed data.
 * Used for testing the complete Builder order flow.
 *
 * @example
 * ```typescript
 * // Create a random user
 * const user = UserClient.createRandom();
 *
 * // Or load from file
 * const user = UserClient.loadFromFile('.test_user.json');
 *
 * // Sign an order
 * const signature = await user.signTypedData(typedData);
 * ```
 */
export class UserClient {
    account;
    _privateKey;
    /**
     * The user's wallet address
     */
    address;
    /**
     * Initialize UserClient.
     *
     * @param privateKey - User's private key (hex string, with or without 0x prefix)
     */
    constructor(privateKey) {
        let pk = privateKey;
        if (!pk.startsWith('0x')) {
            pk = '0x' + pk;
        }
        this._privateKey = pk;
        this.account = privateKeyToAccount(this._privateKey);
        this.address = this.account.address;
    }
    /**
     * Generate a random test user.
     *
     * @returns New UserClient instance
     */
    static createRandom() {
        const privateKey = generatePrivateKey();
        return new UserClient(privateKey);
    }
    /**
     * Load user from JSON file.
     *
     * @param path - JSON file path
     * @returns UserClient instance
     * @throws Error if file does not exist or has invalid format
     */
    static loadFromFile(path) {
        const content = fs.readFileSync(path, 'utf-8');
        const data = JSON.parse(content);
        if (!data.privateKey) {
            throw new Error("JSON file must contain 'privateKey' field");
        }
        return new UserClient(data.privateKey);
    }
    /**
     * Load user from file, or create new user and save if file doesn't exist.
     *
     * @param path - JSON file path
     * @returns UserClient instance
     */
    static loadOrCreate(path) {
        if (fs.existsSync(path)) {
            return UserClient.loadFromFile(path);
        }
        else {
            const user = UserClient.createRandom();
            user.saveToFile(path);
            return user;
        }
    }
    /**
     * Save user to JSON file.
     *
     * @param path - JSON file path
     */
    saveToFile(path) {
        const data = {
            address: this.address,
            privateKey: this._privateKey,
        };
        fs.writeFileSync(path, JSON.stringify(data, null, 2));
        console.log(`User saved to ${path}`);
        console.log(`   Address: ${this.address}`);
    }
    /**
     * Sign EIP-712 typed data.
     *
     * This is the core method for user signing Safe transactions in Builder mode.
     * In production, this signing is done by the user's wallet.
     *
     * @param typedData - EIP-712 typed data containing domain, types, message
     * @returns Signature (hex string with 0x prefix)
     */
    async signTypedData(typedData) {
        const signature = await this.account.signTypedData({
            domain: typedData.domain,
            types: typedData.types,
            primaryType: typedData.primaryType,
            message: typedData.message,
        });
        return signature;
    }
    /**
     * Sign a hash directly (for order signing).
     *
     * This is the core method for user signing orders in Builder mode.
     * Uses the same signing method as the original SDK.
     *
     * @param hashToSign - Hash to sign (hex string, with or without 0x prefix)
     * @returns Signature (hex string with 0x prefix)
     */
    async signHash(hashToSign) {
        let hash = hashToSign;
        if (!hash.startsWith('0x')) {
            hash = '0x' + hash;
        }
        const signResult = await sign({
            hash: hash,
            privateKey: this._privateKey,
        });
        return serializeSignature(signResult);
    }
    /**
     * Get the underlying account
     */
    getAccount() {
        return this.account;
    }
    toString() {
        return `UserClient(address=${this.address})`;
    }
}
