// @opinion-labs/opinion-clob-sdk/dist/models/enums.js
/**
 * Market/Topic Status
 */
export var TopicStatus;
(function (TopicStatus) {
    TopicStatus[TopicStatus["CREATED"] = 1] = "CREATED";
    TopicStatus[TopicStatus["ACTIVATED"] = 2] = "ACTIVATED";
    TopicStatus[TopicStatus["RESOLVING"] = 3] = "RESOLVING";
    TopicStatus[TopicStatus["RESOLVED"] = 4] = "RESOLVED";
    TopicStatus[TopicStatus["FAILED"] = 5] = "FAILED";
    TopicStatus[TopicStatus["DELETED"] = 6] = "DELETED";
})(TopicStatus || (TopicStatus = {}));
/**
 * Market/Topic Type
 */
export var TopicType;
(function (TopicType) {
    TopicType[TopicType["BINARY"] = 0] = "BINARY";
    TopicType[TopicType["CATEGORICAL"] = 1] = "CATEGORICAL";
    TopicType[TopicType["ALL"] = 2] = "ALL";
})(TopicType || (TopicType = {}));
/**
 * Status filter for querying markets
 */
export var TopicStatusFilter;
(function (TopicStatusFilter) {
    TopicStatusFilter["ALL"] = "";
    TopicStatusFilter["ACTIVATED"] = "activated";
    TopicStatusFilter["RESOLVED"] = "resolved";
})(TopicStatusFilter || (TopicStatusFilter = {}));
/**
 * Sort type for market listing
 */
export var TopicSortType;
(function (TopicSortType) {
    TopicSortType[TopicSortType["BY_TIME_DESC"] = 1] = "BY_TIME_DESC";
    TopicSortType[TopicSortType["BY_CUTOFF_TIME_ASC"] = 2] = "BY_CUTOFF_TIME_ASC";
    TopicSortType[TopicSortType["BY_VOLUME_DESC"] = 3] = "BY_VOLUME_DESC";
    TopicSortType[TopicSortType["BY_VOLUME_ASC"] = 4] = "BY_VOLUME_ASC";
    TopicSortType[TopicSortType["BY_VOLUME_24H_DESC"] = 5] = "BY_VOLUME_24H_DESC";
    TopicSortType[TopicSortType["BY_VOLUME_24H_ASC"] = 6] = "BY_VOLUME_24H_ASC";
    TopicSortType[TopicSortType["BY_VOLUME_7D_DESC"] = 7] = "BY_VOLUME_7D_DESC";
    TopicSortType[TopicSortType["BY_VOLUME_7D_ASC"] = 8] = "BY_VOLUME_7D_ASC";
})(TopicSortType || (TopicSortType = {}));
/**
 * Order side
 */
export var OrderSide;
(function (OrderSide) {
    OrderSide[OrderSide["BUY"] = 0] = "BUY";
    OrderSide[OrderSide["SELL"] = 1] = "SELL";
})(OrderSide || (OrderSide = {}));
/**
 * Order type
 */
export var OrderType;
(function (OrderType) {
    OrderType[OrderType["MARKET_ORDER"] = 1] = "MARKET_ORDER";
    OrderType[OrderType["LIMIT_ORDER"] = 2] = "LIMIT_ORDER";
})(OrderType || (OrderType = {}));
/**
 * Signature type for order signing
 */
export var SignatureType;
(function (SignatureType) {
    SignatureType[SignatureType["EOA"] = 0] = "EOA";
    SignatureType[SignatureType["POLY_GNOSIS_SAFE"] = 2] = "POLY_GNOSIS_SAFE";
    SignatureType[SignatureType["POLY_PROXY"] = 1] = "POLY_PROXY";
})(SignatureType || (SignatureType = {}));
