// @opinion-labs/opinion-clob-sdk/src/websocket/client.ts
/**
 * WebSocket client for Opinion CLOB SDK
 *
 * This module provides WebSocket connectivity for real-time market data and updates.
 */

import { WebSocket as IsowsWebSocket } from 'isows';
import WS from 'ws';
import { HttpsProxyAgent } from 'https-proxy-agent';
import {
  type MarketDepthDiffMessage,
  type MarketLastPriceMessage,
  type MarketLastTradeMessage,
  type TradeOrderUpdateMessage,
  type TradeRecordNewMessage,
  WebSocketAction,
  WebSocketChannel,
  parseMarketDepthDiffMessage,
  parseMarketLastPriceMessage,
  parseMarketLastTradeMessage,
  parseTradeOrderUpdateMessage,
  parseTradeRecordNewMessage,
} from './models.js';

// Type for WebSocket that works with both isows and ws
type WebSocketType = InstanceType<typeof IsowsWebSocket> | WS;

/**
 * Subscription key type: tuple of (channel, marketId)
 */
type SubscriptionKey = `${string}:${number}`;

/**
 * Callback types for each channel
 */
export type MarketDepthDiffCallback = (message: MarketDepthDiffMessage) => void;
export type MarketLastPriceCallback = (message: MarketLastPriceMessage) => void;
export type MarketLastTradeCallback = (message: MarketLastTradeMessage) => void;
export type TradeOrderUpdateCallback = (message: TradeOrderUpdateMessage) => void;
export type TradeRecordNewCallback = (message: TradeRecordNewMessage) => void;

/**
 * Union type for all subscription callbacks
 */
export type SubscriptionCallback =
  | MarketDepthDiffCallback
  | MarketLastPriceCallback
  | MarketLastTradeCallback
  | TradeOrderUpdateCallback
  | TradeRecordNewCallback;

/**
 * WebSocket client configuration options
 */
export interface WebSocketClientConfig {
  /** API key for authentication */
  apiKey: string;
  /** Wallet address for heartbeat messages */
  walletAddress: string;
  /** WebSocket server URL (default: wss://ws.opinion.trade) */
  wsUrl?: string;
  /** Interval in seconds between application-level heartbeat messages (default: 30) */
  heartbeatInterval?: number;
  /** HTTP proxy URL for WebSocket connection (optional, e.g., 'http://127.0.0.1:7890') */
  proxyUrl?: string;
  /** Callback function called when a message is received */
  onMessage?: (message: Record<string, unknown>) => void;
  /** Callback function called when connection is opened */
  onOpen?: () => void;
  /** Callback function called when connection is closed */
  onClose?: () => void;
  /** Callback function called when an error occurs */
  onError?: (error: Error) => void;
}

/**
 * WebSocket client for Opinion CLOB SDK with heartbeat support and event callbacks.
 *
 * This client maintains a persistent WebSocket connection to the Opinion WebSocket server,
 * automatically sending heartbeat messages to keep the connection alive.
 *
 * @example
 * ```typescript
 * import { Client, WebSocketClient } from '@opinion-labs/opinion-clob-sdk';
 *
 * const client = new Client({...});
 * const wsClient = client.createWebSocketClient({
 *   onMessage: (msg) => console.log('Received:', msg),
 *   onOpen: () => console.log('Connected'),
 *   onClose: () => console.log('Disconnected'),
 *   onError: (error) => console.log('Error:', error),
 * });
 *
 * // Connect
 * await wsClient.connect();
 *
 * // Subscribe to market depth
 * wsClient.subscribeMarketDepthDiff(1274, (msg) => {
 *   console.log(`Depth update: ${msg.side} ${msg.price} @ ${msg.size}`);
 * });
 *
 * // Later, close the connection
 * wsClient.close();
 * ```
 */
export class WebSocketClient {
  private apiKey: string;
  private wsUrl: string;
  private heartbeatInterval: number;
  private proxyUrl?: string;

  // Callbacks
  private onMessageCallback?: (message: Record<string, unknown>) => void;
  private onOpenCallback?: () => void;
  private onCloseCallback?: () => void;
  private onErrorCallback?: (error: Error) => void;

  // Subscription management
  private subscriptions: Map<SubscriptionKey, SubscriptionCallback> = new Map();
  private activeSubscriptions: Set<SubscriptionKey> = new Set();

  // Internal state
  private ws: WebSocketType | null = null;
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  private _isConnected = false;

  constructor(config: WebSocketClientConfig) {
    if (!config.apiKey) {
      throw new Error('apiKey is required');
    }
    if (!config.walletAddress) {
      throw new Error('walletAddress is required');
    }

    this.apiKey = config.apiKey;
    this.wsUrl = config.wsUrl ?? 'wss://ws.opinion.trade';
    this.heartbeatInterval = config.heartbeatInterval ?? 30;
    this.proxyUrl = config.proxyUrl;

    this.onMessageCallback = config.onMessage;
    this.onOpenCallback = config.onOpen;
    this.onCloseCallback = config.onClose;
    this.onErrorCallback = config.onError;
  }

  /**
   * Check if the WebSocket is connected
   */
  get isConnected(): boolean {
    return this._isConnected;
  }

  /**
   * Build WebSocket URL with apikey query parameter
   */
  private buildWsUrl(): string {
    const params = new URLSearchParams({ apikey: this.apiKey });
    const separator = this.wsUrl.includes('?') ? '&' : '?';
    return `${this.wsUrl}${separator}${params.toString()}`;
  }

  /**
   * Create subscription key from channel and market ID
   */
  private createSubscriptionKey(channel: string, marketId: number): SubscriptionKey {
    return `${channel}:${marketId}`;
  }

  /**
   * Internal handler for WebSocket open event
   */
  private handleOpen = (): void => {
    this._isConnected = true;

    // Start heartbeat
    if (this.heartbeatInterval > 0) {
      this.startHeartbeat();
    }

    // Call user callback
    if (this.onOpenCallback) {
      try {
        this.onOpenCallback();
      } catch (error) {
        console.error('Error in onOpen callback:', error);
      }
    }
  };

  /**
   * Internal handler for WebSocket message event
   */
  private handleMessage = (event: MessageEvent): void => {
    try {
      const data = JSON.parse(event.data as string) as Record<string, unknown>;

      // Route to subscription-specific callback
      const msgType = data.msgType as string | undefined;
      const marketId = data.marketId as number | undefined;
      const rootMarketId = data.rootMarketId as number | undefined;

      if (msgType === WebSocketChannel.MARKET_DEPTH_DIFF && marketId != null) {
        const key = this.createSubscriptionKey(WebSocketChannel.MARKET_DEPTH_DIFF, marketId);
        const callback = this.subscriptions.get(key) as MarketDepthDiffCallback | undefined;
        if (callback) {
          try {
            const parsedMsg = parseMarketDepthDiffMessage(data);
            callback(parsedMsg);
          } catch (error) {
            console.error('Error in subscription callback:', error);
          }
        }
      } else if (msgType === WebSocketChannel.MARKET_LAST_PRICE) {
        // Use rootMarketId if available, otherwise use marketId
        const keyId = rootMarketId ?? marketId;
        if (keyId != null) {
          const key = this.createSubscriptionKey(WebSocketChannel.MARKET_LAST_PRICE, keyId);
          const callback = this.subscriptions.get(key) as MarketLastPriceCallback | undefined;
          if (callback) {
            try {
              const parsedMsg = parseMarketLastPriceMessage(data);
              callback(parsedMsg);
            } catch (error) {
              console.error('Error in subscription callback:', error);
            }
          }
        }
      } else if (msgType === WebSocketChannel.MARKET_LAST_TRADE) {
        const keyId = rootMarketId ?? marketId;
        if (keyId != null) {
          const key = this.createSubscriptionKey(WebSocketChannel.MARKET_LAST_TRADE, keyId);
          const callback = this.subscriptions.get(key) as MarketLastTradeCallback | undefined;
          if (callback) {
            try {
              const parsedMsg = parseMarketLastTradeMessage(data);
              callback(parsedMsg);
            } catch (error) {
              console.error('Error in subscription callback:', error);
            }
          }
        }
      } else if (msgType === WebSocketChannel.TRADE_ORDER_UPDATE) {
        const keyId = rootMarketId && rootMarketId !== 0 ? rootMarketId : marketId;
        if (keyId != null) {
          const key = this.createSubscriptionKey(WebSocketChannel.TRADE_ORDER_UPDATE, keyId);
          const callback = this.subscriptions.get(key) as TradeOrderUpdateCallback | undefined;
          if (callback) {
            try {
              const parsedMsg = parseTradeOrderUpdateMessage(data);
              callback(parsedMsg);
            } catch (error) {
              console.error('Error in subscription callback:', error);
            }
          }
        }
      } else if (msgType === WebSocketChannel.TRADE_RECORD_NEW) {
        const keyId = rootMarketId && rootMarketId !== 0 ? rootMarketId : marketId;
        if (keyId != null) {
          const key = this.createSubscriptionKey(WebSocketChannel.TRADE_RECORD_NEW, keyId);
          const callback = this.subscriptions.get(key) as TradeRecordNewCallback | undefined;
          if (callback) {
            try {
              const parsedMsg = parseTradeRecordNewMessage(data);
              callback(parsedMsg);
            } catch (error) {
              console.error('Error in subscription callback:', error);
            }
          }
        }
      }

      // Also call general message callback if provided
      if (this.onMessageCallback) {
        try {
          this.onMessageCallback(data);
        } catch (error) {
          console.error('Error in onMessage callback:', error);
        }
      }
    } catch (error) {
      console.error('Failed to parse message:', event.data, error);
    }
  };

  /**
   * Internal handler for WebSocket error event
   */
  private handleError = (_event: Event): void => {
    this._isConnected = false;
    const error = new Error('WebSocket error occurred');

    if (this.onErrorCallback) {
      try {
        this.onErrorCallback(error);
      } catch (e) {
        console.error('Error in onError callback:', e);
      }
    }
  };

  /**
   * Internal handler for WebSocket close event
   */
  private handleClose = (): void => {
    this._isConnected = false;
    this.stopHeartbeat();

    if (this.onCloseCallback) {
      try {
        this.onCloseCallback();
      } catch (error) {
        console.error('Error in onClose callback:', error);
      }
    }
  };

  /**
   * Send heartbeat message to keep connection alive
   */
  private sendHeartbeat(): void {
    if (!this._isConnected || !this.ws) {
      return;
    }

    try {
      const heartbeatMsg = { action: WebSocketAction.HEARTBEAT };
      this.ws.send(JSON.stringify(heartbeatMsg));
    } catch (error) {
      console.error('Failed to send heartbeat:', error);
    }
  }

  /**
   * Start heartbeat timer
   */
  private startHeartbeat(): void {
    if (this.heartbeatTimer) {
      return;
    }

    this.heartbeatTimer = setInterval(() => {
      if (this._isConnected) {
        this.sendHeartbeat();
      }
    }, this.heartbeatInterval * 1000);
  }

  /**
   * Stop heartbeat timer
   */
  private stopHeartbeat(): void {
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
  }

  /**
   * Create WebSocket instance (with proxy support for Node.js)
   */
  private createWebSocket(url: string): WebSocketType {
    if (this.proxyUrl) {
      // Use ws library with proxy agent for Node.js
      const agent = new HttpsProxyAgent(this.proxyUrl);
      return new WS(url, { agent });
    } else {
      // Use isows for cross-platform compatibility
      return new IsowsWebSocket(url);
    }
  }

  /**
   * Connect to WebSocket server.
   *
   * @returns Promise that resolves when connection is established
   */
  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this.ws) {
        this.close();
      }

      const fullWsUrl = this.buildWsUrl();
      this.ws = this.createWebSocket(fullWsUrl);

      const onOpen = (): void => {
        this.handleOpen();
        resolve();
      };

      const onError = (event: Event | Error): void => {
        this.handleError(event as Event);
        reject(new Error('WebSocket connection failed'));
      };

      // Handle different event APIs between ws and isows
      if (this.ws instanceof WS) {
        // ws library uses on/once pattern
        this.ws.once('open', onOpen);
        this.ws.once('error', onError);
        this.ws.on('message', (data: WS.RawData) => {
          this.handleMessage({ data: String(data) } as MessageEvent);
        });
        this.ws.on('close', () => this.handleClose());
        this.ws.on('error', (err: Error) => this.handleError(err as unknown as Event));
      } else {
        // isows uses addEventListener pattern
        this.ws.addEventListener('open', onOpen, { once: true });
        this.ws.addEventListener('error', onError as (event: Event) => void, { once: true });
        this.ws.addEventListener('message', this.handleMessage);
        this.ws.addEventListener('close', this.handleClose);
        this.ws.addEventListener('error', this.handleError);
      }
    });
  }

  /**
   * Send a message to the WebSocket server.
   *
   * @param message - Dictionary to send as JSON message
   * @throws Error if WebSocket is not connected
   */
  send(message: Record<string, unknown>): void {
    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected');
    }

    try {
      const jsonMessage = JSON.stringify(message);
      this.ws.send(jsonMessage);
    } catch (error) {
      console.error('Failed to send message:', error);
      throw error;
    }
  }

  /**
   * Subscribe to market depth diff updates for a specific market.
   *
   * @param marketId - Market ID to subscribe to
   * @param callback - Callback function that receives MarketDepthDiffMessage objects
   * @throws Error if WebSocket is not connected or marketId is invalid
   *
   * @example
   * ```typescript
   * wsClient.subscribeMarketDepthDiff(1274, (msg) => {
   *   console.log(`Market ${msg.marketId}: ${msg.side} ${msg.price} @ ${msg.size}`);
   * });
   * ```
   */
  subscribeMarketDepthDiff(marketId: number, callback: MarketDepthDiffCallback): void {
    if (!Number.isInteger(marketId) || marketId <= 0) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected. Call connect() first.');
    }

    const channel = WebSocketChannel.MARKET_DEPTH_DIFF;
    const key = this.createSubscriptionKey(channel, marketId);

    // Store callback
    this.subscriptions.set(key, callback);

    // Send subscription message
    const subscribeMsg = {
      action: WebSocketAction.SUBSCRIBE,
      channel,
      marketId,
    };
    this.send(subscribeMsg);

    // Track active subscription
    this.activeSubscriptions.add(key);
  }

  /**
   * Unsubscribe from market depth diff updates for a specific market.
   *
   * @param marketId - Market ID to unsubscribe from
   * @throws Error if WebSocket is not connected or marketId is invalid
   */
  unsubscribeMarketDepthDiff(marketId: number): void {
    if (!Number.isInteger(marketId) || marketId <= 0) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected');
    }

    const channel = WebSocketChannel.MARKET_DEPTH_DIFF;
    const key = this.createSubscriptionKey(channel, marketId);

    // Send unsubscribe message
    const unsubscribeMsg = {
      action: WebSocketAction.UNSUBSCRIBE,
      channel,
      marketId,
    };
    this.send(unsubscribeMsg);

    // Remove subscription
    this.subscriptions.delete(key);
    this.activeSubscriptions.delete(key);
  }

  /**
   * Subscribe to market last price updates.
   *
   * @param options - Either marketId (for binary markets) or rootMarketId (for categorical markets)
   * @param callback - Callback function that receives MarketLastPriceMessage objects
   * @throws Error if WebSocket is not connected or parameters are invalid
   *
   * @example
   * ```typescript
   * // For binary market
   * wsClient.subscribeMarketLastPrice({ marketId: 1274 }, (msg) => {
   *   console.log(`Market ${msg.marketId}: Last price ${msg.price}`);
   * });
   *
   * // For categorical market
   * wsClient.subscribeMarketLastPrice({ rootMarketId: 61 }, (msg) => {
   *   console.log(`Market ${msg.marketId}: Last price ${msg.price}`);
   * });
   * ```
   */
  subscribeMarketLastPrice(
    options: { marketId?: number; rootMarketId?: number },
    callback: MarketLastPriceCallback
  ): void {
    const { marketId, rootMarketId } = options;

    if (marketId != null && rootMarketId != null) {
      throw new Error('Cannot provide both marketId and rootMarketId. Provide only one of them.');
    }

    if (marketId == null && rootMarketId == null) {
      throw new Error('Either marketId or rootMarketId must be provided.');
    }

    if (marketId != null && (!Number.isInteger(marketId) || marketId <= 0)) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (rootMarketId != null && (!Number.isInteger(rootMarketId) || rootMarketId <= 0)) {
      throw new Error(`rootMarketId must be a positive integer, got: ${rootMarketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected. Call connect() first.');
    }

    const channel = WebSocketChannel.MARKET_LAST_PRICE;
    const keyId = rootMarketId ?? marketId!;
    const key = this.createSubscriptionKey(channel, keyId);

    // Store callback
    this.subscriptions.set(key, callback);

    // Send subscription message
    const subscribeMsg: Record<string, unknown> = {
      action: WebSocketAction.SUBSCRIBE,
      channel,
    };

    if (rootMarketId != null) {
      subscribeMsg.rootMarketId = rootMarketId;
    } else {
      subscribeMsg.marketId = marketId;
    }

    this.send(subscribeMsg);
    this.activeSubscriptions.add(key);
  }

  /**
   * Unsubscribe from market last price updates.
   *
   * @param options - Either marketId (for binary markets) or rootMarketId (for categorical markets)
   * @throws Error if WebSocket is not connected or parameters are invalid
   */
  unsubscribeMarketLastPrice(options: { marketId?: number; rootMarketId?: number }): void {
    const { marketId, rootMarketId } = options;

    if (marketId != null && rootMarketId != null) {
      throw new Error('Cannot provide both marketId and rootMarketId. Provide only one of them.');
    }

    if (marketId == null && rootMarketId == null) {
      throw new Error('Either marketId or rootMarketId must be provided.');
    }

    if (marketId != null && (!Number.isInteger(marketId) || marketId <= 0)) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (rootMarketId != null && (!Number.isInteger(rootMarketId) || rootMarketId <= 0)) {
      throw new Error(`rootMarketId must be a positive integer, got: ${rootMarketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected');
    }

    const channel = WebSocketChannel.MARKET_LAST_PRICE;
    const keyId = rootMarketId ?? marketId!;
    const key = this.createSubscriptionKey(channel, keyId);

    // Send unsubscribe message
    const unsubscribeMsg: Record<string, unknown> = {
      action: WebSocketAction.UNSUBSCRIBE,
      channel,
    };

    if (rootMarketId != null) {
      unsubscribeMsg.rootMarketId = rootMarketId;
    } else {
      unsubscribeMsg.marketId = marketId;
    }

    this.send(unsubscribeMsg);
    this.subscriptions.delete(key);
    this.activeSubscriptions.delete(key);
  }

  /**
   * Subscribe to market last trade updates.
   *
   * @param options - Either marketId (for binary markets) or rootMarketId (for categorical markets)
   * @param callback - Callback function that receives MarketLastTradeMessage objects
   * @throws Error if WebSocket is not connected or parameters are invalid
   *
   * @example
   * ```typescript
   * wsClient.subscribeMarketLastTrade({ marketId: 1274 }, (msg) => {
   *   console.log(`Market ${msg.marketId}: ${msg.side} ${msg.shares} shares @ ${msg.price}`);
   * });
   * ```
   */
  subscribeMarketLastTrade(
    options: { marketId?: number; rootMarketId?: number },
    callback: MarketLastTradeCallback
  ): void {
    const { marketId, rootMarketId } = options;

    if (marketId != null && rootMarketId != null) {
      throw new Error('Cannot provide both marketId and rootMarketId. Provide only one of them.');
    }

    if (marketId == null && rootMarketId == null) {
      throw new Error('Either marketId or rootMarketId must be provided.');
    }

    if (marketId != null && (!Number.isInteger(marketId) || marketId <= 0)) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (rootMarketId != null && (!Number.isInteger(rootMarketId) || rootMarketId <= 0)) {
      throw new Error(`rootMarketId must be a positive integer, got: ${rootMarketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected. Call connect() first.');
    }

    const channel = WebSocketChannel.MARKET_LAST_TRADE;
    const keyId = rootMarketId ?? marketId!;
    const key = this.createSubscriptionKey(channel, keyId);

    // Store callback
    this.subscriptions.set(key, callback);

    // Send subscription message
    const subscribeMsg: Record<string, unknown> = {
      action: WebSocketAction.SUBSCRIBE,
      channel,
    };

    if (rootMarketId != null) {
      subscribeMsg.rootMarketId = rootMarketId;
    } else {
      subscribeMsg.marketId = marketId;
    }

    this.send(subscribeMsg);
    this.activeSubscriptions.add(key);
  }

  /**
   * Unsubscribe from market last trade updates.
   *
   * @param options - Either marketId (for binary markets) or rootMarketId (for categorical markets)
   * @throws Error if WebSocket is not connected or parameters are invalid
   */
  unsubscribeMarketLastTrade(options: { marketId?: number; rootMarketId?: number }): void {
    const { marketId, rootMarketId } = options;

    if (marketId != null && rootMarketId != null) {
      throw new Error('Cannot provide both marketId and rootMarketId. Provide only one of them.');
    }

    if (marketId == null && rootMarketId == null) {
      throw new Error('Either marketId or rootMarketId must be provided.');
    }

    if (marketId != null && (!Number.isInteger(marketId) || marketId <= 0)) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (rootMarketId != null && (!Number.isInteger(rootMarketId) || rootMarketId <= 0)) {
      throw new Error(`rootMarketId must be a positive integer, got: ${rootMarketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected');
    }

    const channel = WebSocketChannel.MARKET_LAST_TRADE;
    const keyId = rootMarketId ?? marketId!;
    const key = this.createSubscriptionKey(channel, keyId);

    // Send unsubscribe message
    const unsubscribeMsg: Record<string, unknown> = {
      action: WebSocketAction.UNSUBSCRIBE,
      channel,
    };

    if (rootMarketId != null) {
      unsubscribeMsg.rootMarketId = rootMarketId;
    } else {
      unsubscribeMsg.marketId = marketId;
    }

    this.send(unsubscribeMsg);
    this.subscriptions.delete(key);
    this.activeSubscriptions.delete(key);
  }

  /**
   * Subscribe to trade order update messages.
   *
   * @param options - Either marketId (for binary markets) or rootMarketId (for categorical markets)
   * @param callback - Callback function that receives TradeOrderUpdateMessage objects
   * @throws Error if WebSocket is not connected or parameters are invalid
   *
   * @example
   * ```typescript
   * wsClient.subscribeTradeOrderUpdate({ marketId: 1274 }, (msg) => {
   *   console.log(`Order ${msg.orderId}: ${msg.orderUpdateType}, Status: ${msg.status}`);
   * });
   * ```
   */
  subscribeTradeOrderUpdate(
    options: { marketId?: number; rootMarketId?: number },
    callback: TradeOrderUpdateCallback
  ): void {
    const { marketId, rootMarketId } = options;

    if (marketId != null && rootMarketId != null) {
      throw new Error('Cannot provide both marketId and rootMarketId. Provide only one of them.');
    }

    if (marketId == null && rootMarketId == null) {
      throw new Error('Either marketId or rootMarketId must be provided.');
    }

    if (marketId != null && (!Number.isInteger(marketId) || marketId <= 0)) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (rootMarketId != null && (!Number.isInteger(rootMarketId) || rootMarketId <= 0)) {
      throw new Error(`rootMarketId must be a positive integer, got: ${rootMarketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected. Call connect() first.');
    }

    const channel = WebSocketChannel.TRADE_ORDER_UPDATE;
    const keyId = rootMarketId ?? marketId!;
    const key = this.createSubscriptionKey(channel, keyId);

    // Store callback
    this.subscriptions.set(key, callback);

    // Send subscription message
    const subscribeMsg: Record<string, unknown> = {
      action: WebSocketAction.SUBSCRIBE,
      channel,
    };

    if (rootMarketId != null) {
      subscribeMsg.rootMarketId = rootMarketId;
    } else {
      subscribeMsg.marketId = marketId;
    }

    this.send(subscribeMsg);
    this.activeSubscriptions.add(key);
  }

  /**
   * Unsubscribe from trade order update messages.
   *
   * @param options - Either marketId (for binary markets) or rootMarketId (for categorical markets)
   * @throws Error if WebSocket is not connected or parameters are invalid
   */
  unsubscribeTradeOrderUpdate(options: { marketId?: number; rootMarketId?: number }): void {
    const { marketId, rootMarketId } = options;

    if (marketId != null && rootMarketId != null) {
      throw new Error('Cannot provide both marketId and rootMarketId. Provide only one of them.');
    }

    if (marketId == null && rootMarketId == null) {
      throw new Error('Either marketId or rootMarketId must be provided.');
    }

    if (marketId != null && (!Number.isInteger(marketId) || marketId <= 0)) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (rootMarketId != null && (!Number.isInteger(rootMarketId) || rootMarketId <= 0)) {
      throw new Error(`rootMarketId must be a positive integer, got: ${rootMarketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected');
    }

    const channel = WebSocketChannel.TRADE_ORDER_UPDATE;
    const keyId = rootMarketId ?? marketId!;
    const key = this.createSubscriptionKey(channel, keyId);

    // Send unsubscribe message
    const unsubscribeMsg: Record<string, unknown> = {
      action: WebSocketAction.UNSUBSCRIBE,
      channel,
    };

    if (rootMarketId != null) {
      unsubscribeMsg.rootMarketId = rootMarketId;
    } else {
      unsubscribeMsg.marketId = marketId;
    }

    this.send(unsubscribeMsg);
    this.subscriptions.delete(key);
    this.activeSubscriptions.delete(key);
  }

  /**
   * Subscribe to trade record new messages.
   *
   * @param options - Either marketId (for binary markets) or rootMarketId (for categorical markets)
   * @param callback - Callback function that receives TradeRecordNewMessage objects
   * @throws Error if WebSocket is not connected or parameters are invalid
   *
   * @example
   * ```typescript
   * wsClient.subscribeTradeRecordNew({ marketId: 1274 }, (msg) => {
   *   console.log(`Trade ${msg.tradeNo}: ${msg.side} ${msg.shares} shares @ ${msg.price}`);
   * });
   * ```
   */
  subscribeTradeRecordNew(
    options: { marketId?: number; rootMarketId?: number },
    callback: TradeRecordNewCallback
  ): void {
    const { marketId, rootMarketId } = options;

    if (marketId != null && rootMarketId != null) {
      throw new Error('Cannot provide both marketId and rootMarketId. Provide only one of them.');
    }

    if (marketId == null && rootMarketId == null) {
      throw new Error('Either marketId or rootMarketId must be provided.');
    }

    if (marketId != null && (!Number.isInteger(marketId) || marketId <= 0)) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (rootMarketId != null && (!Number.isInteger(rootMarketId) || rootMarketId <= 0)) {
      throw new Error(`rootMarketId must be a positive integer, got: ${rootMarketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected. Call connect() first.');
    }

    const channel = WebSocketChannel.TRADE_RECORD_NEW;
    const keyId = rootMarketId ?? marketId!;
    const key = this.createSubscriptionKey(channel, keyId);

    // Store callback
    this.subscriptions.set(key, callback);

    // Send subscription message
    const subscribeMsg: Record<string, unknown> = {
      action: WebSocketAction.SUBSCRIBE,
      channel,
    };

    if (rootMarketId != null) {
      subscribeMsg.rootMarketId = rootMarketId;
    } else {
      subscribeMsg.marketId = marketId;
    }

    this.send(subscribeMsg);
    this.activeSubscriptions.add(key);
  }

  /**
   * Unsubscribe from trade record new messages.
   *
   * @param options - Either marketId (for binary markets) or rootMarketId (for categorical markets)
   * @throws Error if WebSocket is not connected or parameters are invalid
   */
  unsubscribeTradeRecordNew(options: { marketId?: number; rootMarketId?: number }): void {
    const { marketId, rootMarketId } = options;

    if (marketId != null && rootMarketId != null) {
      throw new Error('Cannot provide both marketId and rootMarketId. Provide only one of them.');
    }

    if (marketId == null && rootMarketId == null) {
      throw new Error('Either marketId or rootMarketId must be provided.');
    }

    if (marketId != null && (!Number.isInteger(marketId) || marketId <= 0)) {
      throw new Error(`marketId must be a positive integer, got: ${marketId}`);
    }

    if (rootMarketId != null && (!Number.isInteger(rootMarketId) || rootMarketId <= 0)) {
      throw new Error(`rootMarketId must be a positive integer, got: ${rootMarketId}`);
    }

    if (!this._isConnected || !this.ws) {
      throw new Error('WebSocket is not connected');
    }

    const channel = WebSocketChannel.TRADE_RECORD_NEW;
    const keyId = rootMarketId ?? marketId!;
    const key = this.createSubscriptionKey(channel, keyId);

    // Send unsubscribe message
    const unsubscribeMsg: Record<string, unknown> = {
      action: WebSocketAction.UNSUBSCRIBE,
      channel,
    };

    if (rootMarketId != null) {
      unsubscribeMsg.rootMarketId = rootMarketId;
    } else {
      unsubscribeMsg.marketId = marketId;
    }

    this.send(unsubscribeMsg);
    this.subscriptions.delete(key);
    this.activeSubscriptions.delete(key);
  }

  /**
   * Close WebSocket connection.
   *
   * This method stops the heartbeat timer and closes the WebSocket connection.
   */
  close(): void {
    this.stopHeartbeat();

    if (this.ws) {
      try {
        this.ws.close();
      } catch (error) {
        console.error('Error closing WebSocket:', error);
      } finally {
        this.ws = null;
      }
    }

    this._isConnected = false;
  }
}
