// @opinion-labs/opinion-clob-sdk/src/websocket/index.ts
/**
 * WebSocket module for Opinion CLOB SDK
 *
 * This module provides WebSocket connectivity for real-time market data and updates.
 */

// Export WebSocket client
export { WebSocketClient, type WebSocketClientConfig } from './client.js';

// Export callback types
export type {
  MarketDepthDiffCallback,
  MarketLastPriceCallback,
  MarketLastTradeCallback,
  TradeOrderUpdateCallback,
  TradeRecordNewCallback,
  SubscriptionCallback,
} from './client.js';

// Export message types
export type {
  BaseWebSocketMessage,
  MarketDepthDiffMessage,
  MarketLastPriceMessage,
  MarketLastTradeMessage,
  TradeOrderUpdateMessage,
  TradeRecordNewMessage,
  SubscribeMessageBinary,
  SubscribeMessageCategorical,
  HeartbeatMessage,
  SubscribeMessage,
  WebSocketMessage,
} from './models.js';

// Export enums
export { WebSocketChannel, WebSocketAction } from './models.js';

// Export parser functions
export {
  parseMarketDepthDiffMessage,
  parseMarketLastPriceMessage,
  parseMarketLastTradeMessage,
  parseTradeOrderUpdateMessage,
  parseTradeRecordNewMessage,
} from './models.js';
