// @opinion-labs/opinion-api/dist/sdk.gen.js
// This file is auto-generated by @hey-api/openapi-ts
import { client } from './client.gen.js';
/**
 * Get Market List
 *
 * Get Market List(All Markets)
 */
export const getMarket = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/market',
        ...options
    });
};
/**
 * Get Detail of a Categorical Market
 *
 * Get Detail of a Categorical Market
 */
export const getMarketCategoricalByMarketId = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/market/categorical/{marketId}',
        ...options
    });
};
/**
 * Get Detail of a Binary Market
 *
 * Get Detail of a Binary Market
 */
export const getMarketByMarketId = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/market/{marketId}',
        ...options
    });
};
/**
 * Get Market Detail by Slug
 *
 * Get market detail by slug, supports both binary and categorical markets
 */
export const getMarketSlugBySlug = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/market/slug/{slug}',
        ...options
    });
};
/**
 * Get orders
 *
 * Get user's orders with simplified query parameters
 */
export const getOrder = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/order',
        ...options
    });
};
/**
 * Create an order
 *
 * Create an Order
 */
export const postOrder = (options) => {
    return (options.client ?? client).post({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/order',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Cancel an order
 *
 * Cancel an Order
 */
export const postOrderCancel = (options) => {
    return (options.client ?? client).post({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/order/cancel',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Get order detail
 *
 * Get order detail with related trades
 */
export const getOrderByOrderId = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/order/{orderId}',
        ...options
    });
};
/**
 * Get user positions
 *
 * Get user's current positions with simplified query parameters
 */
export const getPositions = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/positions',
        ...options
    });
};
/**
 * Get another user's positions
 *
 * Get positions of a specific user by wallet address
 */
export const getPositionsUserByWalletAddress = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/positions/user/{walletAddress}',
        ...options
    });
};
/**
 * Get quote token list
 *
 * Retrieve a list of quote tokens (trading pair quote currencies)
 */
export const getQuoteToken = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/quoteToken',
        ...options
    });
};
/**
 * Get fee rates for a token
 *
 * Get the taker and maker fee rates for a specific token
 */
export const getTokenFeeRates = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/token/fee-rates',
        ...options
    });
};
/**
 * Get latest price of a token
 *
 * Get the latest trade price and details for a specific token
 */
export const getTokenLatestPrice = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/token/latest-price',
        ...options
    });
};
/**
 * Get orderbook of a token
 *
 * Get orderbook (market depth) of a specific token
 */
export const getTokenOrderbook = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/token/orderbook',
        ...options
    });
};
/**
 * Get price history of a token
 *
 * Get historical price data for a specific token (Polymarket-compatible format)
 */
export const getTokenPriceHistory = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/token/price-history',
        ...options
    });
};
/**
 * Get Topic List
 *
 * Get Topic List(All Topics)
 */
export const getTopic = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/topic',
        ...options
    });
};
/**
 * Get Detail of a Categorical Topic
 *
 * Get Detail of a Categorical Topic
 */
export const getTopicMultiByTopicId = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/topic/multi/{topicId}',
        ...options
    });
};
/**
 * Get Detail of a Binary Topic
 *
 * Get Detail of a Binary Topic
 */
export const getTopicByTopicId = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/topic/{topicId}',
        ...options
    });
};
/**
 * Get user's trades
 *
 * Get user's trades with simplified query parameters
 */
export const getTrade = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/trade',
        ...options
    });
};
/**
 * Get another user's trades
 *
 * Get trades of a specific user by wallet address. Only returns filled (successful) trades.
 */
export const getTradeUserByWalletAddress = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/trade/user/{walletAddress}',
        ...options
    });
};
/**
 * Auth
 *
 * Authentication of apikey and returns user info
 */
export const getUserAuth = (options) => {
    return (options?.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/user/auth',
        ...options
    });
};
/**
 * Get user's quote token balances
 *
 * Get user's quote token (USDC, etc.) balances including available and frozen amounts
 */
export const getUserBalance = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/user/balance',
        ...options
    });
};
