// @opinion-labs/opinion-clob-sdk/dist/chain/contract_caller.js
import { getContract, encodeFunctionData, formatUnits, } from 'viem';
import { Safe } from './safe/safe.js';
import { MultiSendOperation } from './safe/index.js';
import { NULL_HASH } from './safe/constants.js';
import { conditionalTokensAbi } from './contracts/conditional_tokens.js';
import { erc20Abi } from './contracts/erc20.js';
import { feeManagerAbi } from './contracts/fee_manager.js';
import { Signer } from './order_utils/signer.js';
import { BalanceNotEnough, NoPositionsToRedeem, InsufficientGasBalance } from '../errors.js';
/**
 * ContractCaller handles all blockchain interactions
 */
export class ContractCaller {
    safe;
    signer;
    multiSigAddr;
    conditionalTokensAddr;
    feeManagerAddr;
    enableTradingCheckInterval;
    enableTradingLastTime = null;
    tokenDecimalsCache = new Map();
    constructor(rpcUrl, privateKey, multiSigAddr, conditionalTokensAddr, multiSendAddr, feeManagerAddr, chainId = 56, enableTradingCheckInterval = 3600, proxyUrl) {
        this.signer = new Signer(privateKey);
        this.multiSigAddr = multiSigAddr;
        this.conditionalTokensAddr = conditionalTokensAddr;
        this.feeManagerAddr = feeManagerAddr;
        this.enableTradingCheckInterval = enableTradingCheckInterval;
        this.safe = new Safe(rpcUrl, privateKey, multiSigAddr, multiSendAddr, chainId, proxyUrl);
    }
    /**
     * Get signer instance
     */
    getSigner() {
        return this.signer;
    }
    /**
     * Get multi-sig address
     */
    getMultiSigAddress() {
        return this.multiSigAddr;
    }
    /**
     * Get public client from Safe
     */
    getPublicClient() {
        return this.safe.getPublicClient();
    }
    /**
     * Get ERC20 contract instance
     */
    getErc20Contract(address) {
        return getContract({
            address,
            abi: erc20Abi,
            client: this.getPublicClient(),
        });
    }
    /**
     * Get ConditionalTokens contract instance
     */
    getConditionalTokensContract() {
        return getContract({
            address: this.conditionalTokensAddr,
            abi: conditionalTokensAbi,
            client: this.getPublicClient(),
        });
    }
    /**
     * Get FeeManager contract instance
     */
    getFeeManagerContract() {
        return getContract({
            address: this.feeManagerAddr,
            abi: feeManagerAbi,
            client: this.getPublicClient(),
        });
    }
    /**
     * Get token decimals with caching
     */
    async getTokenDecimals(tokenAddress) {
        const key = tokenAddress.toLowerCase();
        if (!this.tokenDecimalsCache.has(key)) {
            try {
                const contract = this.getErc20Contract(tokenAddress);
                const decimals = await contract.read.decimals();
                this.tokenDecimalsCache.set(key, decimals);
            }
            catch (error) {
                console.warn(`Failed to get decimals for ${tokenAddress}, defaulting to 18`);
                this.tokenDecimalsCache.set(key, 18);
            }
        }
        return this.tokenDecimalsCache.get(key);
    }
    /**
     * Check gas balance for signer
     */
    async checkGasBalance(estimatedGas = 500000n) {
        const signerAddress = this.signer.address();
        const publicClient = this.getPublicClient();
        const gasBalance = await publicClient.getBalance({ address: signerAddress });
        // Get current gas price
        const gasPrice = await publicClient.getGasPrice();
        // Add 20% safety margin
        const estimatedGasWithMargin = (estimatedGas * 120n) / 100n;
        const requiredGas = estimatedGasWithMargin * gasPrice;
        if (gasBalance < requiredGas) {
            const gasBalanceEth = formatUnits(gasBalance, 18);
            const requiredEth = formatUnits(requiredGas, 18);
            throw new InsufficientGasBalance(`Insufficient gas balance. Signer ${signerAddress} has ${gasBalanceEth} BNB, ` +
                `but needs approximately ${requiredEth} BNB for gas`);
        }
    }
    /**
     * Get position ID for outcome token
     */
    async getPositionId(conditionId, indexSet, collateralToken) {
        const contract = this.getConditionalTokensContract();
        const collectionId = await contract.read.getCollectionId([
            NULL_HASH,
            conditionId,
            BigInt(indexSet),
        ]);
        return await contract.read.getPositionId([collateralToken, collectionId]);
    }
    /**
     * Split collateral into outcome tokens
     */
    async split(collateralToken, conditionId, amount, partition = [1, 2]) {
        await this.checkGasBalance(300000n);
        // Check collateral balance
        const erc20 = this.getErc20Contract(collateralToken);
        const balance = await erc20.read.balanceOf([this.multiSigAddr]);
        if (balance < amount) {
            throw new BalanceNotEnough('Insufficient collateral balance for split');
        }
        const data = encodeFunctionData({
            abi: conditionalTokensAbi,
            functionName: 'splitPosition',
            args: [collateralToken, NULL_HASH, conditionId, partition.map((p) => BigInt(p)), amount],
        });
        const txs = [
            {
                operation: MultiSendOperation.CALL,
                to: this.conditionalTokensAddr,
                value: 0n,
                data,
            },
        ];
        const result = await this.safe.executeMultiSend(txs);
        // Wait for receipt and validate
        const publicClient = this.getPublicClient();
        const receipt = await publicClient.waitForTransactionReceipt({
            hash: result.txHash,
            timeout: 120_000,
        });
        if (receipt.status !== 'success') {
            throw new Error(`Split transaction failed. Transaction hash: ${result.txHash}`);
        }
        return {
            txHash: result.txHash,
            safeTxHash: result.safeTxHash,
            returnValue: result.returnValue,
        };
    }
    /**
     * Merge outcome tokens back to collateral
     */
    async merge(collateralToken, conditionId, amount, partition = [1, 2]) {
        await this.checkGasBalance(300000n);
        // Check balance of positions
        for (const indexSet of partition) {
            const positionId = await this.getPositionId(conditionId, indexSet, collateralToken);
            const ct = this.getConditionalTokensContract();
            const balance = await ct.read.balanceOf([this.multiSigAddr, positionId]);
            if (balance < amount) {
                throw new BalanceNotEnough('Insufficient outcome token balance for merge');
            }
        }
        const data = encodeFunctionData({
            abi: conditionalTokensAbi,
            functionName: 'mergePositions',
            args: [collateralToken, NULL_HASH, conditionId, partition.map((p) => BigInt(p)), amount],
        });
        const txs = [
            {
                operation: MultiSendOperation.CALL,
                to: this.conditionalTokensAddr,
                value: 0n,
                data,
            },
        ];
        const result = await this.safe.executeMultiSend(txs);
        const publicClient = this.getPublicClient();
        const receipt = await publicClient.waitForTransactionReceipt({
            hash: result.txHash,
            timeout: 120_000,
        });
        if (receipt.status !== 'success') {
            throw new Error(`Merge transaction failed. Transaction hash: ${result.txHash}`);
        }
        return {
            txHash: result.txHash,
            safeTxHash: result.safeTxHash,
            returnValue: result.returnValue,
        };
    }
    /**
     * Redeem outcome tokens after market resolution
     */
    async redeem(collateralToken, conditionId, partition = [1, 2]) {
        await this.checkGasBalance(300000n);
        // Check if user has any positions to redeem
        let hasPositions = false;
        for (const indexSet of partition) {
            const positionId = await this.getPositionId(conditionId, indexSet, collateralToken);
            const ct = this.getConditionalTokensContract();
            const balance = await ct.read.balanceOf([this.multiSigAddr, positionId]);
            if (balance > 0n) {
                hasPositions = true;
                break;
            }
        }
        if (!hasPositions) {
            throw new NoPositionsToRedeem();
        }
        const data = encodeFunctionData({
            abi: conditionalTokensAbi,
            functionName: 'redeemPositions',
            args: [collateralToken, NULL_HASH, conditionId, partition.map((p) => BigInt(p))],
        });
        const txs = [
            {
                operation: MultiSendOperation.CALL,
                to: this.conditionalTokensAddr,
                value: 0n,
                data,
            },
        ];
        const result = await this.safe.executeMultiSend(txs);
        const publicClient = this.getPublicClient();
        const receipt = await publicClient.waitForTransactionReceipt({
            hash: result.txHash,
            timeout: 120_000,
        });
        if (receipt.status !== 'success') {
            throw new Error(`Redeem transaction failed. Transaction hash: ${result.txHash}`);
        }
        return {
            txHash: result.txHash,
            safeTxHash: result.safeTxHash,
            returnValue: result.returnValue,
        };
    }
    /**
     * Enable trading by approving necessary tokens
     */
    async enableTrading(supportedQuoteTokens) {
        // Check if we recently enabled trading
        if (this.enableTradingLastTime !== null &&
            Date.now() - this.enableTradingLastTime < this.enableTradingCheckInterval * 1000) {
            return { txHash: null, safeTxHash: null, returnValue: null };
        }
        this.enableTradingLastTime = Date.now();
        await this.checkGasBalance(500000n);
        const txs = [];
        const MAX_UINT256 = 2n ** 256n - 1n;
        for (const [erc20Address, ctfExchangeAddress] of supportedQuoteTokens) {
            const erc20 = this.getErc20Contract(erc20Address);
            const decimals = await this.getTokenDecimals(erc20Address);
            const minThreshold = BigInt(1000000000) * BigInt(10 ** decimals);
            // Check and approve for CTF Exchange
            let allowance = await erc20.read.allowance([this.multiSigAddr, ctfExchangeAddress]);
            if (allowance < minThreshold) {
                // Reset to 0 first if needed (for tokens like USDT)
                if (allowance > 0n) {
                    const resetData = encodeFunctionData({
                        abi: erc20Abi,
                        functionName: 'approve',
                        args: [ctfExchangeAddress, 0n],
                    });
                    txs.push({
                        operation: MultiSendOperation.CALL,
                        to: erc20Address,
                        value: 0n,
                        data: resetData,
                    });
                }
                // Approve unlimited
                const approveData = encodeFunctionData({
                    abi: erc20Abi,
                    functionName: 'approve',
                    args: [ctfExchangeAddress, MAX_UINT256],
                });
                txs.push({
                    operation: MultiSendOperation.CALL,
                    to: erc20Address,
                    value: 0n,
                    data: approveData,
                });
            }
            // Check and approve for ConditionalTokens (splitting)
            allowance = await erc20.read.allowance([this.multiSigAddr, this.conditionalTokensAddr]);
            if (allowance < minThreshold) {
                if (allowance > 0n) {
                    const resetData = encodeFunctionData({
                        abi: erc20Abi,
                        functionName: 'approve',
                        args: [this.conditionalTokensAddr, 0n],
                    });
                    txs.push({
                        operation: MultiSendOperation.CALL,
                        to: erc20Address,
                        value: 0n,
                        data: resetData,
                    });
                }
                const approveData = encodeFunctionData({
                    abi: erc20Abi,
                    functionName: 'approve',
                    args: [this.conditionalTokensAddr, MAX_UINT256],
                });
                txs.push({
                    operation: MultiSendOperation.CALL,
                    to: erc20Address,
                    value: 0n,
                    data: approveData,
                });
            }
            // Check and approve ConditionalTokens for CTF Exchange
            const ct = this.getConditionalTokensContract();
            const isApproved = await ct.read.isApprovedForAll([this.multiSigAddr, ctfExchangeAddress]);
            if (!isApproved) {
                const approveAllData = encodeFunctionData({
                    abi: conditionalTokensAbi,
                    functionName: 'setApprovalForAll',
                    args: [ctfExchangeAddress, true],
                });
                txs.push({
                    operation: MultiSendOperation.CALL,
                    to: this.conditionalTokensAddr,
                    value: 0n,
                    data: approveAllData,
                });
            }
        }
        if (txs.length > 0) {
            const result = await this.safe.executeMultiSend(txs);
            const publicClient = this.getPublicClient();
            const receipt = await publicClient.waitForTransactionReceipt({
                hash: result.txHash,
                timeout: 120_000,
            });
            if (receipt.status !== 'success') {
                throw new Error(`Enable trading failed. Transaction hash: ${result.txHash}`);
            }
            return {
                txHash: result.txHash,
                safeTxHash: result.safeTxHash,
                returnValue: result.returnValue,
            };
        }
        return { txHash: null, safeTxHash: null, returnValue: null };
    }
    /**
     * Get fee rate settings from FeeManager contract
     */
    async getFeeRateSettings(tokenId) {
        const contract = this.getFeeManagerContract();
        const [makerFeeRateBps, takerFeeRateBps, enabled] = await contract.read.getFeeRateSettings([
            tokenId,
        ]);
        // Convert basis points to max fee rate percentage
        // Formula: fee_rate_bps * 0.25 / 10000
        const makerMaxFeeRate = (Number(makerFeeRateBps) * 0.25) / 10000;
        const takerMaxFeeRate = (Number(takerFeeRateBps) * 0.25) / 10000;
        return {
            makerMaxFeeRate,
            takerMaxFeeRate,
            enabled,
        };
    }
}
