// @opinion-labs/opinion-clob-sdk/dist/client.js
import { getAddress } from 'viem';
import { fetch as undiciFetch, ProxyAgent } from 'undici';
import { ContractCaller, } from './chain/contract_caller.js';
import { OrderBuilder } from './chain/order_utils/order_builder.js';
import { calculateOrderAmounts, safeAmountToWei } from './chain/order_utils/utils.js';
import { ZERO_ADDRESS, } from './models/order.js';
import { TopicStatus, TopicStatusFilter, TopicType, TopicSortType, OrderSide, OrderType, SignatureType, } from './models/enums.js';
import { DEFAULT_CONTRACT_ADDRESSES, SUPPORTED_CHAIN_IDS, } from './config.js';
import { InvalidParamError, OpenApiError } from './errors.js';
import { WebSocketClient } from './websocket/index.js';
// Import generated API client and types
import { getMarket as getMarketList, getMarketByMarketId as getBinaryMarketDetail, getMarketCategoricalByMarketId as getCategoricalMarketDetail, getMarketSlugBySlug as getMarketDetailBySlug, getQuoteToken as getQuoteTokenList, getTokenOrderbook as apiGetOrderbook, getTokenLatestPrice as apiGetLatestPrice, getTokenPriceHistory as apiGetPriceHistory, getPositions, getTrade as getTrades, getOrder as getOrders, getOrderByOrderId as apiGetOrderById, getUserBalance as getUserBalances, getUserAuth as apiGetUserAuth, postOrder as createOrder, postOrderCancel as apiCancelOrder, } from '@opinion-labs/opinion-api';
import { createClient, createConfig } from '@opinion-labs/opinion-api/client';
/**
 * Main SDK Client for Opinion CLOB operations
 */
export class Client {
    apiClient;
    chainId;
    contractCaller;
    apiKey;
    ctfExchangeAddress;
    proxyUrl;
    // Cache settings
    quoteTokensCacheTtl;
    marketCacheTtl;
    quoteTokensCache = null;
    marketCache = new Map();
    constructor(config) {
        // Validate chain ID
        if (!SUPPORTED_CHAIN_IDS.includes(config.chainId)) {
            throw new InvalidParamError(`chainId must be one of ${SUPPORTED_CHAIN_IDS.join(', ')}`);
        }
        this.chainId = config.chainId;
        this.apiKey = config.apiKey;
        this.quoteTokensCacheTtl = config.quoteTokensCacheTtl ?? 3600;
        this.marketCacheTtl = config.marketCacheTtl ?? 300;
        this.ctfExchangeAddress = config.ctfExchangeAddress;
        this.proxyUrl = config.proxyUrl;
        // Create custom fetch function with proxy support
        let customFetch;
        if (config.proxyUrl) {
            const proxyDispatcher = new ProxyAgent(config.proxyUrl);
            customFetch = async (input, init) => {
                const request = new Request(input, init);
                const response = await undiciFetch(request.url, {
                    method: request.method,
                    headers: Object.fromEntries(request.headers.entries()),
                    body: request.body ? await request.text() : undefined,
                    dispatcher: proxyDispatcher,
                });
                return response;
            };
        }
        // Initialize API client with configuration
        this.apiClient = createClient(createConfig({
            baseUrl: config.host,
            headers: {
                apikey: config.apiKey,
            },
            fetch: customFetch,
        }));
        // Get default contract addresses
        const defaults = DEFAULT_CONTRACT_ADDRESSES[config.chainId];
        this.contractCaller = new ContractCaller(config.rpcUrl, config.privateKey, getAddress(config.multiSigAddress), config.conditionalTokensAddress ?? defaults.conditionalTokens, config.multiSendAddress ?? defaults.multisend, config.feeManagerAddress ?? defaults.feeManager, config.chainId, config.enableTradingCheckInterval ?? 3600, config.proxyUrl);
    }
    /**
     * Check if cache is valid
     */
    isCacheValid(cache, ttl) {
        if (!cache || ttl <= 0)
            return false;
        return Date.now() - cache.timestamp < ttl * 1000;
    }
    /**
     * Transform market data - filter to known fields and compute isIncentivized.
     */
    transformMarketData(market) {
        const incentiveFactor = market['incentiveFactor'];
        const isIncentivized = incentiveFactor !== null && incentiveFactor !== undefined;
        const transformed = {
            chainId: market['chainId'],
            conditionId: market['conditionId'],
            createdAt: market['createdAt'],
            cutoffAt: market['cutoffAt'],
            isIncentivized,
            marketId: market['marketId'],
            marketTitle: market['marketTitle'],
            noLabel: market['noLabel'],
            noTokenId: market['noTokenId'],
            questionId: market['questionId'],
            quoteToken: market['quoteToken'],
            resolvedAt: market['resolvedAt'],
            resultTokenId: market['resultTokenId'],
            rules: market['rules'],
            slug: market['slug'],
            status: market['status'],
            statusEnum: market['statusEnum'],
            volume: market['volume'],
            yesLabel: market['yesLabel'],
            yesTokenId: market['yesTokenId'],
        };
        if (market['collection'] !== undefined) {
            transformed['collection'] = market['collection'];
        }
        if (market['childMarkets'] && Array.isArray(market['childMarkets'])) {
            transformed['childMarkets'] = market['childMarkets'].map((child) => this.transformChildMarketData(child));
        }
        return transformed;
    }
    /**
     * Transform child market data - filter to known fields.
     */
    transformChildMarketData(child) {
        return {
            chainId: child['chainId'],
            conditionId: child['conditionId'],
            createdAt: child['createdAt'],
            cutoffAt: child['cutoffAt'],
            marketId: child['marketId'],
            marketTitle: child['marketTitle'],
            noLabel: child['noLabel'],
            noTokenId: child['noTokenId'],
            questionId: child['questionId'],
            quoteToken: child['quoteToken'],
            resolvedAt: child['resolvedAt'],
            resultTokenId: child['resultTokenId'],
            rules: child['rules'],
            slug: child['slug'],
            status: child['status'],
            statusEnum: child['statusEnum'],
            volume: child['volume'],
            yesLabel: child['yesLabel'],
            yesTokenId: child['yesTokenId'],
        };
    }
    /**
     * Handle API response - returns response as-is to match Python SDK behavior.
     * Does not throw on non-zero errno, caller must check errno if needed.
     */
    handleApiResponse(response, defaultErrorMsg) {
        if (response.error) {
            const errorData = response.error;
            throw new OpenApiError(errorData?.errmsg || defaultErrorMsg);
        }
        const data = response.data;
        if (!data) {
            throw new OpenApiError(defaultErrorMsg);
        }
        // Return the full response including errno/errmsg to match Python SDK behavior.
        // Python SDK does not throw on non-zero errno, it returns the response as-is.
        return {
            errno: data.errno ?? 0,
            errmsg: data.errmsg ?? '',
            result: data.result,
        };
    }
    // ==================== Market Data APIs ====================
    /**
     * Get list of supported quote tokens
     */
    async getQuoteTokens(useCache = true) {
        if (useCache && this.isCacheValid(this.quoteTokensCache, this.quoteTokensCacheTtl)) {
            return { errno: 0, errmsg: '', result: this.quoteTokensCache.data };
        }
        const response = await getQuoteTokenList({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: {
                chainId: String(this.chainId),
            },
        });
        const result = this.handleApiResponse(response, 'Failed to get quote tokens');
        const quoteTokenResult = {
            total: result.result?.total ?? 0,
            list: result.result?.list ?? [],
        };
        if (this.quoteTokensCacheTtl > 0) {
            this.quoteTokensCache = { data: quoteTokenResult, timestamp: Date.now() };
        }
        return { errno: 0, errmsg: '', result: quoteTokenResult };
    }
    /**
     * Get markets with pagination
     */
    async getMarkets(options = {}) {
        const { topicType = TopicType.ALL, page = 1, limit = 20, status, sortBy = TopicSortType.BY_TIME_DESC, } = options;
        if (page < 1)
            throw new InvalidParamError('page must be >= 1');
        if (limit < 1 || limit > 20)
            throw new InvalidParamError('limit must be between 1 and 20');
        const queryParams = {
            chainId: String(this.chainId),
            marketType: topicType,
            page,
            limit,
            sortBy,
        };
        if (status && status !== TopicStatusFilter.ALL) {
            queryParams.status = status;
        }
        const response = await getMarketList({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: queryParams,
        });
        const result = this.handleApiResponse(response, 'Failed to get markets');
        // Transform market data to match Python SDK's model format
        const rawList = (result.result?.list ?? []);
        const transformedList = rawList.map((market) => this.transformMarketData(market));
        return {
            errno: result.errno,
            errmsg: result.errmsg,
            result: {
                total: result.result?.total ?? 0,
                list: transformedList,
            },
        };
    }
    /**
     * Get market details by ID
     */
    async getMarket(marketId, useCache = true) {
        if (!marketId)
            throw new InvalidParamError('marketId is required');
        if (useCache) {
            const cached = this.marketCache.get(marketId);
            if (this.isCacheValid(cached ?? null, this.marketCacheTtl)) {
                return { errno: 0, errmsg: '', result: cached.data };
            }
        }
        const response = await getBinaryMarketDetail({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            path: { marketId },
        });
        const result = this.handleApiResponse(response, 'Failed to get market details');
        // Return the response with errno/errmsg preserved (match Python SDK)
        // Transform market data if present
        let transformedData;
        if (result.result?.data) {
            transformedData = this.transformMarketData(result.result.data);
        }
        const marketResult = {
            data: transformedData,
        };
        if (result.errno === 0 && this.marketCacheTtl > 0) {
            this.marketCache.set(marketId, { data: marketResult, timestamp: Date.now() });
        }
        return { errno: result.errno, errmsg: result.errmsg, result: marketResult };
    }
    /**
     * Get categorical market details
     */
    async getCategoricalMarket(marketId) {
        if (!marketId)
            throw new InvalidParamError('marketId is required');
        const response = await getCategoricalMarketDetail({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            path: { marketId },
        });
        const result = this.handleApiResponse(response, 'Failed to get categorical market details');
        // Transform market data if present
        let transformedData;
        if (result.result?.data) {
            transformedData = this.transformMarketData(result.result.data);
        }
        return {
            errno: result.errno,
            errmsg: result.errmsg,
            result: {
                data: transformedData,
            },
        };
    }
    /**
     * Get market details by slug (supports both binary and categorical markets)
     */
    async getMarketBySlug(slug) {
        if (!slug || !slug.trim())
            throw new InvalidParamError('slug is required');
        const trimmedSlug = slug.trim();
        const response = await getMarketDetailBySlug({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            path: { slug: trimmedSlug },
        });
        const result = this.handleApiResponse(response, 'Failed to get market details by slug');
        let transformedData;
        if (result.result?.data) {
            transformedData = this.transformMarketData(result.result.data);
        }
        return {
            errno: result.errno,
            errmsg: result.errmsg,
            result: {
                data: transformedData,
            },
        };
    }
    /**
     * Get orderbook for a token
     */
    async getOrderbook(tokenId) {
        if (!tokenId)
            throw new InvalidParamError('tokenId is required');
        const response = await apiGetOrderbook({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: { token_id: tokenId },
        });
        return this.handleApiResponse(response, 'Failed to get orderbook');
    }
    /**
     * Get latest price for a token
     */
    async getLatestPrice(tokenId) {
        if (!tokenId)
            throw new InvalidParamError('tokenId is required');
        const response = await apiGetLatestPrice({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: { token_id: tokenId },
        });
        return this.handleApiResponse(response, 'Failed to get latest price');
    }
    /**
     * Get price history for a token
     */
    async getPriceHistory(tokenId, options = {}) {
        if (!tokenId)
            throw new InvalidParamError('tokenId is required');
        const { interval = '1h', startAt, endAt } = options;
        const response = await apiGetPriceHistory({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: {
                token_id: tokenId,
                interval,
                start_at: startAt,
                end_at: endAt,
            },
        });
        return this.handleApiResponse(response, 'Failed to get price history');
    }
    /**
     * Get fee rates from on-chain FeeManager contract
     */
    async getFeeRates(tokenId) {
        if (!tokenId)
            throw new InvalidParamError('tokenId is required');
        return await this.contractCaller.getFeeRateSettings(BigInt(tokenId));
    }
    // ==================== User APIs ====================
    /**
     * Get user's positions
     */
    async getMyPositions(options = {}) {
        const { marketId, page = 1, limit = 10 } = options;
        const response = await getPositions({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: {
                chainId: String(this.chainId),
                page,
                limit,
                marketId: marketId && marketId > 0 ? marketId : undefined,
            },
        });
        return this.handleApiResponse(response, 'Failed to get positions');
    }
    /**
     * Get user's trade history
     */
    async getMyTrades(options = {}) {
        const { marketId, page = 1, limit = 10 } = options;
        const response = await getTrades({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: {
                chainId: String(this.chainId),
                page,
                limit,
                marketId: marketId && marketId > 0 ? marketId : undefined,
            },
        });
        const apiResponse = this.handleApiResponse(response, 'Failed to get trades');
        // Transform trade data to align with Python SDK:
        // - Use fee_formatted as fee (string, avoids precision loss)
        // - Remove feeFormatted field
        if (apiResponse.result?.list) {
            apiResponse.result.list = apiResponse.result.list.map((trade) => {
                const { feeFormatted, fee: _fee, ...rest } = trade;
                return {
                    ...rest,
                    fee: feeFormatted,
                };
            });
        }
        return apiResponse;
    }
    /**
     * Get user's orders with optional filters
     */
    async getMyOrders(options = {}) {
        const { marketId, status, page = 1, limit = 10 } = options;
        const response = await getOrders({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: {
                chainId: String(this.chainId),
                page,
                limit,
                marketId: marketId && marketId > 0 ? marketId : undefined,
                status,
            },
        });
        return this.handleApiResponse(response, 'Failed to get orders');
    }
    /**
     * Get detailed information about a specific order
     */
    async getOrderById(orderId) {
        if (!orderId)
            throw new InvalidParamError('orderId must be a non-empty string');
        const response = await apiGetOrderById({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            path: { orderId },
        });
        return this.handleApiResponse(response, 'Failed to get order');
    }
    /**
     * Get user's balances
     */
    async getMyBalances() {
        const response = await getUserBalances({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            query: {
                chain_id: String(this.chainId),
            },
        });
        return this.handleApiResponse(response, 'Failed to get balances');
    }
    /**
     * Get authenticated user information
     */
    async getUserAuth() {
        const response = await apiGetUserAuth({
            client: this.apiClient,
        });
        return this.handleApiResponse(response, 'Failed to get user auth');
    }
    // ==================== Trading APIs ====================
    /**
     * Enable trading by approving necessary tokens
     */
    async enableTrading() {
        const quoteTokensResponse = await this.getQuoteTokens();
        if (quoteTokensResponse.errno !== 0) {
            throw new OpenApiError(`Failed to get quote tokens: ${quoteTokensResponse.errmsg || 'Unknown error'}`);
        }
        const quoteTokens = quoteTokensResponse.result.list;
        if (!quoteTokens || quoteTokens.length === 0) {
            throw new OpenApiError('No supported quote tokens found');
        }
        const supportedQuoteTokens = new Map();
        for (const token of quoteTokens) {
            const qt = token;
            const tokenAddr = qt.quoteTokenAddress;
            const ctfExchangeAddr = qt.ctfExchangeAddress;
            if (tokenAddr) {
                // Use configured ctfExchangeAddress if provided, otherwise use API-provided address
                const exchangeAddr = this.ctfExchangeAddress ?? (ctfExchangeAddr ? getAddress(ctfExchangeAddr) : null);
                if (exchangeAddr) {
                    supportedQuoteTokens.set(getAddress(tokenAddr), exchangeAddr);
                }
            }
        }
        return await this.contractCaller.enableTrading(supportedQuoteTokens);
    }
    /**
     * Place an order
     */
    async placeOrder(data, checkApproval = false) {
        if (checkApproval) {
            await this.enableTrading();
        }
        // Get quote tokens and market data
        const quoteTokensResponse = await this.getQuoteTokens();
        if (quoteTokensResponse.errno !== 0) {
            throw new OpenApiError(`Failed to get quote tokens: ${quoteTokensResponse.errmsg || 'Unknown error'}`);
        }
        const quoteTokens = quoteTokensResponse.result.list;
        const marketResponse = await this.getMarket(data.marketId);
        if (marketResponse.errno !== 0) {
            throw new OpenApiError(`Failed to get market: ${marketResponse.errmsg || 'Unknown error'}`);
        }
        const market = marketResponse.result.data;
        const marketData = market;
        if (parseInt(marketData.chainId ?? '') !== this.chainId) {
            throw new OpenApiError('Cannot place order on different chain');
        }
        // Find quote token
        const quoteToken = quoteTokens.find((t) => {
            const qt = t;
            return qt.quoteTokenAddress?.toLowerCase() === marketData.quoteToken?.toLowerCase();
        });
        if (!quoteToken) {
            throw new OpenApiError('Quote token not found for this market');
        }
        const qt = quoteToken;
        const quoteTokenDecimal = qt.decimal ?? 6;
        const quoteTokenExchangeAddr = qt.ctfExchangeAddress ?? '';
        // Validate order inputs
        this.validateOrderInputs(data);
        // Calculate maker amount
        const makerAmount = this.calculateMakerAmount(data, quoteTokenDecimal);
        // Build and sign order - use configured ctfExchangeAddress if provided
        const exchangeAddress = this.ctfExchangeAddress ?? getAddress(quoteTokenExchangeAddr);
        const orderBuilder = new OrderBuilder(exchangeAddress, this.chainId, this.contractCaller.getSigner());
        let takerAmount = 0n;
        let recalculatedMakerAmount = safeAmountToWei(makerAmount, quoteTokenDecimal);
        if (data.orderType === OrderType.LIMIT_ORDER) {
            const amounts = calculateOrderAmounts(parseFloat(data.price), recalculatedMakerAmount, data.side, quoteTokenDecimal);
            recalculatedMakerAmount = amounts.recalculatedMakerAmount;
            takerAmount = amounts.takerAmount;
        }
        const orderData = {
            maker: this.contractCaller.getMultiSigAddress(),
            taker: ZERO_ADDRESS,
            tokenId: data.tokenId,
            makerAmount: recalculatedMakerAmount,
            takerAmount,
            side: data.side,
            feeRateBps: '0',
            signatureType: SignatureType.POLY_GNOSIS_SAFE,
            signer: this.contractCaller.getSigner().address(),
            nonce: '0',
            expiration: '0',
        };
        const signedOrder = await orderBuilder.buildSignedOrder(orderData);
        // Submit order to API
        const orderForApi = {
            salt: signedOrder.order.salt.toString(),
            topicId: data.marketId,
            maker: signedOrder.order.maker,
            signer: signedOrder.order.signer,
            taker: signedOrder.order.taker,
            tokenId: signedOrder.order.tokenId.toString(),
            makerAmount: signedOrder.order.makerAmount.toString(),
            takerAmount: signedOrder.order.takerAmount.toString(),
            expiration: signedOrder.order.expiration.toString(),
            nonce: signedOrder.order.nonce.toString(),
            feeRateBps: signedOrder.order.feeRateBps.toString(),
            side: signedOrder.order.side.toString(),
            signatureType: signedOrder.order.signatureType.toString(),
            signature: signedOrder.signature,
            sign: signedOrder.signature,
            contractAddress: '',
            currencyAddress: getAddress(marketData.quoteToken ?? ''),
            price: data.orderType === OrderType.MARKET_ORDER ? '0' : data.price,
            tradingMethod: data.orderType,
            timestamp: Math.floor(Date.now() / 1000),
            safeRate: '0',
            orderExpTime: '0',
        };
        const response = await createOrder({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            body: orderForApi,
        });
        return this.handleApiResponse(response, 'Failed to create order');
    }
    /**
     * Validate order inputs
     */
    validateOrderInputs(data) {
        // Reject market buy with base token amount
        if (data.side === OrderSide.BUY &&
            data.orderType === OrderType.MARKET_ORDER &&
            data.makerAmountInBaseToken) {
            throw new InvalidParamError('makerAmountInBaseToken is not allowed for market buy');
        }
        // Reject market sell with quote token amount
        if (data.side === OrderSide.SELL &&
            data.orderType === OrderType.MARKET_ORDER &&
            data.makerAmountInQuoteToken) {
            throw new InvalidParamError('makerAmountInQuoteToken is not allowed for market sell');
        }
        // Validate price for limit orders
        if (data.orderType === OrderType.LIMIT_ORDER) {
            const price = parseFloat(data.price);
            if (isNaN(price) || price <= 0) {
                throw new InvalidParamError(`Price must be positive for limit orders, got: ${data.price}`);
            }
        }
    }
    /**
     * Calculate maker amount from input
     */
    calculateMakerAmount(data, _decimals) {
        const minimalAmount = 1;
        if (data.side === OrderSide.BUY) {
            if (data.makerAmountInBaseToken) {
                const baseAmount = parseFloat(data.makerAmountInBaseToken);
                if (baseAmount < minimalAmount) {
                    throw new InvalidParamError('makerAmountInBaseToken must be at least 1');
                }
                return baseAmount * parseFloat(data.price);
            }
            else if (data.makerAmountInQuoteToken) {
                const quoteAmount = parseFloat(data.makerAmountInQuoteToken);
                if (quoteAmount < minimalAmount) {
                    throw new InvalidParamError('makerAmountInQuoteToken must be at least 1');
                }
                return quoteAmount;
            }
            throw new InvalidParamError('Either makerAmountInBaseToken or makerAmountInQuoteToken must be provided for BUY orders');
        }
        else {
            if (data.makerAmountInBaseToken) {
                const baseAmount = parseFloat(data.makerAmountInBaseToken);
                if (baseAmount < minimalAmount) {
                    throw new InvalidParamError('makerAmountInBaseToken must be at least 1');
                }
                return baseAmount;
            }
            else if (data.makerAmountInQuoteToken) {
                const quoteAmount = parseFloat(data.makerAmountInQuoteToken);
                if (quoteAmount < minimalAmount) {
                    throw new InvalidParamError('makerAmountInQuoteToken must be at least 1');
                }
                const price = parseFloat(data.price);
                if (price === 0) {
                    throw new InvalidParamError('Price cannot be zero for SELL orders with makerAmountInQuoteToken');
                }
                return quoteAmount / price;
            }
            throw new InvalidParamError('Either makerAmountInBaseToken or makerAmountInQuoteToken must be provided for SELL orders');
        }
    }
    /**
     * Cancel an order
     */
    async cancelOrder(orderId) {
        if (!orderId)
            throw new InvalidParamError('orderId must be a non-empty string');
        const response = await apiCancelOrder({
            client: this.apiClient,
            headers: { apikey: this.apiKey },
            body: { orderId },
        });
        return this.handleApiResponse(response, 'Failed to cancel order');
    }
    /**
     * Place multiple orders in batch
     */
    async placeOrdersBatch(orders, checkApproval = false) {
        if (!orders || orders.length === 0) {
            throw new InvalidParamError('orders list cannot be empty');
        }
        if (checkApproval) {
            await this.enableTrading();
        }
        const results = [];
        for (let i = 0; i < orders.length; i++) {
            try {
                const result = await this.placeOrder(orders[i], false);
                results.push({ index: i, success: true, result });
            }
            catch (error) {
                const errorMsg = error instanceof Error ? error.message : String(error);
                results.push({ index: i, success: false, error: errorMsg });
            }
        }
        return results;
    }
    /**
     * Cancel multiple orders in batch
     */
    async cancelOrdersBatch(orderIds) {
        if (!orderIds || orderIds.length === 0) {
            throw new InvalidParamError('orderIds list cannot be empty');
        }
        const results = [];
        for (let i = 0; i < orderIds.length; i++) {
            try {
                const result = await this.cancelOrder(orderIds[i]);
                results.push({ index: i, success: true, result });
            }
            catch (error) {
                const errorMsg = error instanceof Error ? error.message : String(error);
                results.push({ index: i, success: false, error: errorMsg });
            }
        }
        return results;
    }
    /**
     * Cancel all open orders, optionally filtered by market and/or side.
     * Uses pagination to fetch all orders (max 20 per page).
     */
    async cancelAllOrders(options = {}) {
        const { marketId, side } = options;
        // Collect all open orders using pagination
        const allOrdersList = [];
        let page = 1;
        const limit = 20; // Backend max page size
        const maxPages = 100; // Safety limit
        while (page <= maxPages) {
            const pageOrders = await this.getMyOrders({
                marketId: marketId,
                status: '1', // 1 = pending/open orders
                limit,
                page,
            });
            // Check for API errors
            if (pageOrders.errno !== 0) {
                throw new OpenApiError(`Failed to get orders: ${pageOrders.errmsg || 'Unknown error'}`);
            }
            const ordersList = pageOrders.result?.list;
            if (!ordersList || ordersList.length === 0) {
                break;
            }
            for (const order of ordersList) {
                if (order.orderId) {
                    allOrdersList.push({ orderId: order.orderId, side: order.side ?? 0 });
                }
            }
            if (ordersList.length < limit) {
                break;
            }
            page++;
        }
        if (allOrdersList.length === 0) {
            return {
                totalOrders: 0,
                cancelled: 0,
                failed: 0,
                results: [],
            };
        }
        // Filter by side if specified
        let filteredOrders = allOrdersList;
        if (side !== undefined) {
            filteredOrders = allOrdersList.filter((order) => order.side === side);
        }
        // Extract order IDs
        const orderIds = filteredOrders.map((order) => order.orderId).filter(Boolean);
        if (orderIds.length === 0) {
            return {
                totalOrders: 0,
                cancelled: 0,
                failed: 0,
                results: [],
            };
        }
        // Cancel all orders in batch
        const results = await this.cancelOrdersBatch(orderIds);
        // Count successes and failures
        const cancelled = results.filter((r) => r.success).length;
        const failed = results.filter((r) => !r.success).length;
        return {
            totalOrders: orderIds.length,
            cancelled,
            failed,
            results,
        };
    }
    // ==================== Token Operations ====================
    /**
     * Split collateral into outcome tokens
     */
    async split(marketId, amount, checkApproval = true) {
        if (!marketId || marketId <= 0) {
            throw new InvalidParamError('marketId must be a positive integer');
        }
        if (!amount || amount <= 0n) {
            throw new InvalidParamError('amount must be a positive integer');
        }
        if (checkApproval) {
            await this.enableTrading();
        }
        const marketResponse = await this.getMarket(marketId);
        if (marketResponse.errno !== 0) {
            throw new OpenApiError(`Failed to get market: ${marketResponse.errmsg || 'Unknown error'}`);
        }
        const marketData = marketResponse.result.data;
        if (parseInt(marketData.chainId ?? '') !== this.chainId) {
            throw new OpenApiError('Cannot split on different chain');
        }
        const validStatuses = [TopicStatus.ACTIVATED, TopicStatus.RESOLVING, TopicStatus.RESOLVED];
        if (!validStatuses.includes(marketData.status)) {
            throw new OpenApiError('Cannot split on non-activated/resolving/resolved market');
        }
        return await this.contractCaller.split(getAddress(marketData.quoteToken ?? ''), `0x${marketData.conditionId}`, amount);
    }
    /**
     * Merge outcome tokens back to collateral
     */
    async merge(marketId, amount, checkApproval = true) {
        if (!marketId || marketId <= 0) {
            throw new InvalidParamError('marketId must be a positive integer');
        }
        if (!amount || amount <= 0n) {
            throw new InvalidParamError('amount must be a positive integer');
        }
        if (checkApproval) {
            await this.enableTrading();
        }
        const marketResponse = await this.getMarket(marketId);
        if (marketResponse.errno !== 0) {
            throw new OpenApiError(`Failed to get market: ${marketResponse.errmsg || 'Unknown error'}`);
        }
        const marketData = marketResponse.result.data;
        if (parseInt(marketData.chainId ?? '') !== this.chainId) {
            throw new OpenApiError('Cannot merge on different chain');
        }
        const validStatuses = [TopicStatus.ACTIVATED, TopicStatus.RESOLVING, TopicStatus.RESOLVED];
        if (!validStatuses.includes(marketData.status)) {
            throw new OpenApiError('Cannot merge on non-activated/resolving/resolved market');
        }
        return await this.contractCaller.merge(getAddress(marketData.quoteToken ?? ''), `0x${marketData.conditionId}`, amount);
    }
    /**
     * Redeem winning outcome tokens
     */
    async redeem(marketId, checkApproval = true) {
        if (!marketId || marketId <= 0) {
            throw new InvalidParamError('marketId must be a positive integer');
        }
        if (checkApproval) {
            await this.enableTrading();
        }
        const marketResponse = await this.getMarket(marketId);
        if (marketResponse.errno !== 0) {
            throw new OpenApiError(`Failed to get market: ${marketResponse.errmsg || 'Unknown error'}`);
        }
        const marketData = marketResponse.result.data;
        if (parseInt(marketData.chainId ?? '') !== this.chainId) {
            throw new OpenApiError('Cannot redeem on different chain');
        }
        if (marketData.status !== TopicStatus.RESOLVED) {
            throw new OpenApiError('Cannot redeem on non-resolved market');
        }
        return await this.contractCaller.redeem(getAddress(marketData.quoteToken ?? ''), `0x${marketData.conditionId}`);
    }
    // ==================== WebSocket ====================
    /**
     * Create a WebSocket client for real-time market data.
     *
     * This method creates a WebSocket client that is pre-configured with the
     * API key and wallet address from this client instance.
     *
     * @param options - WebSocket client configuration options
     * @returns WebSocketClient instance
     *
     * @example
     * ```typescript
     * const wsClient = client.createWebSocketClient({
     *   heartbeatInterval: 30,
     *   onMessage: (msg) => console.log('Received:', msg),
     *   onOpen: () => console.log('Connected'),
     *   onClose: () => console.log('Disconnected'),
     *   onError: (error) => console.log('Error:', error),
     * });
     *
     * await wsClient.connect();
     *
     * // Subscribe to market updates
     * wsClient.subscribeMarketDepthDiff(1274, (msg) => {
     *   console.log(`Depth update: ${msg.side} ${msg.price} @ ${msg.size}`);
     * });
     * ```
     */
    createWebSocketClient(options = {}) {
        const walletAddress = this.contractCaller.getMultiSigAddress();
        return new WebSocketClient({
            apiKey: this.apiKey,
            walletAddress,
            proxyUrl: this.proxyUrl,
            ...options,
        });
    }
}
