// @opinion-labs/opinion-clob-sdk/dist/errors.js
/**
 * Base SDK error class
 */
export class SdkError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SdkError';
    }
}
/**
 * Invalid parameter error
 */
export class InvalidParamError extends SdkError {
    constructor(message) {
        super(message);
        this.name = 'InvalidParamError';
    }
}
/**
 * OpenAPI error
 */
export class OpenApiError extends SdkError {
    constructor(message) {
        super(message);
        this.name = 'OpenApiError';
    }
}
/**
 * Insufficient balance error
 */
export class BalanceNotEnough extends SdkError {
    constructor(message = 'Insufficient balance') {
        super(message);
        this.name = 'BalanceNotEnough';
    }
}
/**
 * No positions to redeem error
 */
export class NoPositionsToRedeem extends SdkError {
    constructor(message = 'No positions to redeem') {
        super(message);
        this.name = 'NoPositionsToRedeem';
    }
}
/**
 * Insufficient gas balance error
 */
export class InsufficientGasBalance extends SdkError {
    constructor(message) {
        super(message);
        this.name = 'InsufficientGasBalance';
    }
}
/**
 * Validation exception for order building
 */
export class ValidationException extends SdkError {
    constructor(message) {
        super(message);
        this.name = 'ValidationException';
    }
}
