// polymarket-us/dist/index.mjs
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);

// src/auth.ts
import * as ed25519 from "@noble/ed25519";
async function createAuthHeaders(keyId, secretKey, method, path) {
  const timestamp = Date.now().toString();
  const message = `${timestamp}${method}${path}`;
  const secretKeyBytes = base64ToBytes(secretKey);
  const privateKey = secretKeyBytes.length === 64 ? secretKeyBytes.slice(0, 32) : secretKeyBytes;
  const messageBytes = new TextEncoder().encode(message);
  const signatureBytes = await ed25519.signAsync(messageBytes, privateKey);
  const signature = bytesToBase64(signatureBytes);
  return {
    "X-PM-Access-Key": keyId,
    "X-PM-Timestamp": timestamp,
    "X-PM-Signature": signature
  };
}
function base64ToBytes(base64) {
  const binaryString = atob(base64);
  const bytes = new Uint8Array(binaryString.length);
  for (let i = 0; i < binaryString.length; i++) {
    bytes[i] = binaryString.charCodeAt(i);
  }
  return bytes;
}
function bytesToBase64(bytes) {
  const binaryString = Array.from(
    bytes,
    (byte) => String.fromCharCode(byte)
  ).join("");
  return btoa(binaryString);
}

// src/error.ts
var PolymarketUSError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "PolymarketUSError";
  }
};
var APIError = class extends PolymarketUSError {
  constructor(status, message, code) {
    super(message);
    this.name = "APIError";
    this.status = status;
    this.code = code;
  }
};
var AuthenticationError = class extends APIError {
  constructor(message = "Authentication failed") {
    super(401, message, "authentication_error");
    this.name = "AuthenticationError";
  }
};
var BadRequestError = class extends APIError {
  constructor(message = "Bad request") {
    super(400, message, "bad_request");
    this.name = "BadRequestError";
  }
};
var NotFoundError = class extends APIError {
  constructor(message = "Resource not found") {
    super(404, message, "not_found");
    this.name = "NotFoundError";
  }
};
var RateLimitError = class extends APIError {
  constructor(message = "Rate limit exceeded") {
    super(429, message, "rate_limit_exceeded");
    this.name = "RateLimitError";
  }
};
var InternalServerError = class extends APIError {
  constructor(message = "Internal server error") {
    super(500, message, "internal_server_error");
    this.name = "InternalServerError";
  }
};
var WebSocketError = class extends PolymarketUSError {
  constructor(message, requestId) {
    super(message);
    this.name = "WebSocketError";
    this.requestId = requestId;
  }
};

// src/resource.ts
var APIResource = class {
  constructor(client) {
    this.client = client;
  }
};

// src/resources/account.ts
var Account = class extends APIResource {
  async balances() {
    return this.client.get("/v1/account/balances", { authenticated: true });
  }
};

// src/resources/events.ts
var Events = class extends APIResource {
  async list(params) {
    return this.client.get("/v1/events", { query: params });
  }
  async retrieve(id) {
    return this.client.get(`/v1/events/${id}`);
  }
  async retrieveBySlug(slug) {
    return this.client.get(`/v1/events/slug/${slug}`);
  }
};

// src/resources/markets.ts
var Markets = class extends APIResource {
  async list(params) {
    return this.client.get("/v1/markets", { query: params });
  }
  async retrieve(id) {
    return this.client.get(`/v1/market/id/${id}`);
  }
  async retrieveBySlug(slug) {
    return this.client.get(`/v1/market/slug/${slug}`);
  }
  async book(slug) {
    return this.client.get(`/v1/markets/${slug}/book`);
  }
  async bbo(slug) {
    return this.client.get(`/v1/markets/${slug}/bbo`);
  }
  async settlement(slug) {
    return this.client.get(`/v1/markets/${slug}/settlement`);
  }
};

// src/resources/orders.ts
var Orders = class extends APIResource {
  async create(params) {
    return this.client.post("/v1/orders", {
      body: params,
      authenticated: true
    });
  }
  async list(params) {
    return this.client.get("/v1/orders/open", {
      query: params,
      authenticated: true
    });
  }
  async retrieve(orderId) {
    return this.client.get(`/v1/order/${orderId}`, { authenticated: true });
  }
  async cancel(orderId, params) {
    await this.client.post(`/v1/order/${orderId}/cancel`, {
      body: params,
      authenticated: true
    });
  }
  async modify(orderId, params) {
    await this.client.post(`/v1/order/${orderId}/modify`, {
      body: params,
      authenticated: true
    });
  }
  async cancelAll(params) {
    return this.client.post("/v1/orders/open/cancel", {
      body: params || {},
      authenticated: true
    });
  }
  async preview(params) {
    return this.client.post("/v1/order/preview", {
      body: params,
      authenticated: true
    });
  }
  async closePosition(params) {
    return this.client.post("/v1/order/close-position", {
      body: params,
      authenticated: true
    });
  }
};

// src/resources/portfolio.ts
var Portfolio = class extends APIResource {
  async positions(params) {
    return this.client.get("/v1/portfolio/positions", {
      query: params,
      authenticated: true
    });
  }
  async activities(params) {
    return this.client.get("/v1/portfolio/activities", {
      query: params,
      authenticated: true
    });
  }
};

// src/resources/search.ts
var Search = class extends APIResource {
  async query(params) {
    return this.client.get("/v1/search", { query: params });
  }
};

// src/resources/series.ts
var Series = class extends APIResource {
  async list(params) {
    return this.client.get("/v1/series", { query: params });
  }
  async retrieve(id) {
    return this.client.get(`/v1/series/id/${id}`);
  }
};

// src/resources/sports.ts
var Sports = class extends APIResource {
  async list() {
    return this.client.get("/v1/sports");
  }
  async teams(params) {
    return this.client.get("/v1/sports/teams/provider", { query: params });
  }
};

// src/websocket/emitter.ts
var _listeners;
var EventEmitter = class {
  constructor() {
    __privateAdd(this, _listeners, {});
  }
  on(event, listener) {
    if (!__privateGet(this, _listeners)[event]) {
      __privateGet(this, _listeners)[event] = [];
    }
    __privateGet(this, _listeners)[event].push({ listener });
    return this;
  }
  off(event, listener) {
    const listeners = __privateGet(this, _listeners)[event];
    if (!listeners) return this;
    const index = listeners.findIndex((l) => l.listener === listener);
    if (index >= 0) listeners.splice(index, 1);
    return this;
  }
  once(event, listener) {
    if (!__privateGet(this, _listeners)[event]) {
      __privateGet(this, _listeners)[event] = [];
    }
    __privateGet(this, _listeners)[event].push({ listener, once: true });
    return this;
  }
  _emit(event, ...args) {
    const listeners = __privateGet(this, _listeners)[event];
    if (listeners) {
      __privateGet(this, _listeners)[event] = listeners.filter((l) => !l.once);
      for (const { listener } of listeners) {
        listener(...args);
      }
    }
  }
  _hasListener(event) {
    const listeners = __privateGet(this, _listeners)[event];
    return !!listeners && listeners.length > 0;
  }
};
_listeners = new WeakMap();

// src/websocket/base.ts
var WS_OPEN = 1;
function isBrowser() {
  return typeof window !== "undefined" && typeof window.document !== "undefined";
}
async function getWebSocketImpl() {
  if (isBrowser()) {
    throw new PolymarketUSError(
      "WebSocket is not supported in browsers. The Polymarket US API requires header-based authentication which browser WebSocket cannot send. Use this SDK in a Node.js environment."
    );
  }
  const ws = await import("ws");
  return ws.default;
}
var BaseWebSocket = class extends EventEmitter {
  constructor(options, path) {
    super();
    this.socket = null;
    this.options = options;
    this.path = path;
  }
  async connect() {
    const baseUrl = this.options.baseUrl || "wss://api.polymarket.us";
    const url = `${baseUrl}${this.path}`;
    const authHeaders = await createAuthHeaders(
      this.options.keyId,
      this.options.secretKey,
      "GET",
      this.path
    );
    const WebSocketImpl = await getWebSocketImpl();
    this.socket = new WebSocketImpl(url, [], {
      headers: authHeaders
    });
    this.socket.addEventListener("message", (event) => {
      this.handleMessage(event.data.toString());
    });
    this.socket.addEventListener("error", (event) => {
      this.handleError(event);
    });
    this.socket.addEventListener("close", () => {
      this.handleClose();
    });
    const socket = this.socket;
    return new Promise((resolve, reject) => {
      socket.addEventListener("open", () => {
        this._emit("open");
        resolve();
      });
      socket.addEventListener(
        "error",
        (e) => reject(new PolymarketUSError(`WebSocket connection failed: ${e}`))
      );
    });
  }
  send(request) {
    if (!this.socket || this.socket.readyState !== WS_OPEN) {
      throw new PolymarketUSError("WebSocket is not connected");
    }
    this.socket.send(JSON.stringify(request));
  }
  subscribe(requestId, subscriptionType, marketSlugs) {
    this.send({
      subscribe: {
        requestId,
        subscriptionType,
        marketSlugs
      }
    });
  }
  unsubscribe(requestId) {
    this.send({
      unsubscribe: { requestId }
    });
  }
  close() {
    if (this.socket) {
      this.socket.close(1e3, "OK");
      this.socket = null;
    }
  }
  get isConnected() {
    return this.socket !== null && this.socket.readyState === WS_OPEN;
  }
};

// src/websocket/markets.ts
var MarketsWebSocket = class extends BaseWebSocket {
  constructor(options) {
    super(options, "/v1/ws/markets");
  }
  subscribeMarketData(requestId, marketSlugs) {
    this.subscribe(requestId, "SUBSCRIPTION_TYPE_MARKET_DATA", marketSlugs);
  }
  subscribeMarketDataLite(requestId, marketSlugs) {
    this.subscribe(
      requestId,
      "SUBSCRIPTION_TYPE_MARKET_DATA_LITE",
      marketSlugs
    );
  }
  subscribeTrades(requestId, marketSlugs) {
    this.subscribe(requestId, "SUBSCRIPTION_TYPE_TRADE", marketSlugs);
  }
  subscribeAll(requestId, subscriptionType, marketSlugs) {
    this.subscribe(requestId, subscriptionType, marketSlugs);
  }
  handleMessage(data) {
    let message;
    try {
      message = JSON.parse(data);
    } catch {
      this._emit(
        "error",
        new PolymarketUSError(`Failed to parse WebSocket message: ${data}`)
      );
      return;
    }
    this._emit("message", message);
    if ("heartbeat" in message) {
      this._emit("heartbeat");
      return;
    }
    if ("error" in message) {
      this._emit("error", new WebSocketError(message.error, message.requestId));
      return;
    }
    if ("marketData" in message) {
      this._emit("marketData", message);
    } else if ("marketDataLite" in message) {
      this._emit("marketDataLite", message);
    } else if ("trade" in message) {
      this._emit("trade", message);
    }
  }
  handleError(event) {
    const message = event instanceof Error ? event.message : "WebSocket error occurred";
    this._emit("error", new PolymarketUSError(message));
  }
  handleClose() {
    this._emit("close");
  }
};

// src/websocket/private.ts
var PrivateWebSocket = class extends BaseWebSocket {
  constructor(options) {
    super(options, "/v1/ws/private");
  }
  subscribeOrders(requestId, marketSlugs) {
    this.subscribe(requestId, "SUBSCRIPTION_TYPE_ORDER", marketSlugs);
  }
  subscribePositions(requestId, marketSlugs) {
    this.subscribe(requestId, "SUBSCRIPTION_TYPE_POSITION", marketSlugs);
  }
  subscribeAccountBalance(requestId) {
    this.subscribe(requestId, "SUBSCRIPTION_TYPE_ACCOUNT_BALANCE");
  }
  subscribeAll(requestId, subscriptionType, marketSlugs) {
    this.subscribe(requestId, subscriptionType, marketSlugs);
  }
  handleMessage(data) {
    let message;
    try {
      message = JSON.parse(data);
    } catch {
      this._emit(
        "error",
        new PolymarketUSError(`Failed to parse WebSocket message: ${data}`)
      );
      return;
    }
    this._emit("message", message);
    if ("heartbeat" in message) {
      this._emit("heartbeat");
      return;
    }
    if ("error" in message) {
      this._emit("error", new WebSocketError(message.error, message.requestId));
      return;
    }
    if ("orderSubscriptionSnapshot" in message || "ordersSnapshot" in message) {
      this._emit("orderSnapshot", message);
    } else if ("orderSubscriptionUpdate" in message || "orderUpdate" in message) {
      this._emit("orderUpdate", message);
    } else if ("positionSubscriptionSnapshot" in message || "positionsSnapshot" in message) {
      this._emit("positionSnapshot", message);
    } else if ("positionSubscriptionUpdate" in message || "positionUpdate" in message) {
      this._emit("positionUpdate", message);
    } else if ("accountBalanceSubscriptionSnapshot" in message || "accountBalancesSnapshot" in message) {
      this._emit("accountBalanceSnapshot", message);
    } else if ("accountBalanceSubscriptionUpdate" in message || "accountBalanceUpdate" in message) {
      this._emit("accountBalanceUpdate", message);
    }
  }
  handleError(event) {
    const message = event instanceof Error ? event.message : "WebSocket error occurred";
    this._emit("error", new PolymarketUSError(message));
  }
  handleClose() {
    this._emit("close");
  }
};

// src/client.ts
var GATEWAY_BASE_URL = "https://gateway.polymarket.us";
var API_BASE_URL = "https://api.polymarket.us";
function getFetch() {
  if (typeof fetch !== "undefined") {
    return fetch;
  }
  throw new PolymarketUSError(
    '`fetch` is not defined. This SDK requires Node.js 18+ or a browser environment. If using an older Node.js version, polyfill fetch: `globalThis.fetch = require("node-fetch")`'
  );
}
var PolymarketUS = class {
  constructor(options = {}) {
    this.keyId = options.keyId;
    this.secretKey = options.secretKey;
    this.gatewayBaseUrl = options.gatewayBaseUrl || GATEWAY_BASE_URL;
    this.apiBaseUrl = options.apiBaseUrl || API_BASE_URL;
    this.timeout = options.timeout || 3e4;
    this.events = new Events(this);
    this.markets = new Markets(this);
    this.orders = new Orders(this);
    this.portfolio = new Portfolio(this);
    this.account = new Account(this);
    this.series = new Series(this);
    this.sports = new Sports(this);
    this.search = new Search(this);
    this.ws = new WebSocketFactory(this);
  }
  async get(path, options) {
    return this.request("GET", path, options);
  }
  async post(path, options) {
    return this.request("POST", path, options);
  }
  async delete(path, options) {
    return this.request("DELETE", path, options);
  }
  async request(method, path, options = {}) {
    const { query, body, authenticated } = options;
    const baseUrl = authenticated ? this.apiBaseUrl : this.gatewayBaseUrl;
    const url = new URL(path, baseUrl);
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value !== void 0 && value !== null) {
          if (Array.isArray(value)) {
            for (const item of value) {
              url.searchParams.append(key, String(item));
            }
          } else {
            url.searchParams.set(key, String(value));
          }
        }
      }
    }
    const headers = {
      "Content-Type": "application/json"
    };
    if (authenticated) {
      if (!this.keyId || !this.secretKey) {
        throw new AuthenticationError(
          "API key credentials required for authenticated endpoints. Provide keyId and secretKey when initializing the client."
        );
      }
      const authHeaders = await createAuthHeaders(
        this.keyId,
        this.secretKey,
        method,
        url.pathname
      );
      Object.assign(headers, authHeaders);
    }
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);
    try {
      const response = await getFetch()(url.toString(), {
        method,
        headers,
        body: body ? JSON.stringify(body) : void 0,
        signal: controller.signal
      });
      clearTimeout(timeoutId);
      if (!response.ok) {
        await this.handleErrorResponse(response);
      }
      const text = await response.text();
      if (!text) {
        return {};
      }
      return JSON.parse(text);
    } catch (error) {
      clearTimeout(timeoutId);
      if (error instanceof APIError) {
        throw error;
      }
      if (error instanceof Error) {
        if (error.name === "AbortError") {
          throw new APIError(408, "Request timeout");
        }
        throw new APIError(0, error.message);
      }
      throw error;
    }
  }
  async handleErrorResponse(response) {
    const text = await response.text();
    let message;
    try {
      const data = JSON.parse(text);
      message = data.message || data.error || response.statusText;
    } catch {
      message = text || response.statusText;
    }
    switch (response.status) {
      case 400:
        throw new BadRequestError(message);
      case 401:
        throw new AuthenticationError(message);
      case 404:
        throw new NotFoundError(message);
      case 429:
        throw new RateLimitError(message);
      case 500:
      case 502:
      case 503:
      case 504:
        throw new InternalServerError(message);
      default:
        throw new APIError(response.status, message);
    }
  }
};
var WebSocketFactory = class {
  constructor(client) {
    this.client = client;
  }
  getWebSocketOptions() {
    if (!this.client.keyId || !this.client.secretKey) {
      throw new AuthenticationError(
        "API key credentials required for WebSocket connections. Provide keyId and secretKey when initializing the client."
      );
    }
    const url = new URL(this.client.apiBaseUrl);
    url.protocol = url.protocol === "http:" ? "ws:" : "wss:";
    return {
      keyId: this.client.keyId,
      secretKey: this.client.secretKey,
      baseUrl: url.origin
    };
  }
  private() {
    return new PrivateWebSocket(this.getWebSocketOptions());
  }
  markets() {
    return new MarketsWebSocket(this.getWebSocketOptions());
  }
};
export {
  APIError,
  AuthenticationError,
  BadRequestError,
  InternalServerError,
  MarketsWebSocket,
  NotFoundError,
  PolymarketUS,
  PolymarketUSError,
  PrivateWebSocket,
  RateLimitError,
  WebSocketError
};
