// @opinion-labs/opinion-clob-sdk/src/client.ts
import { type Address, type Hex, getAddress } from 'viem';
import { fetch as undiciFetch, ProxyAgent, type Dispatcher } from 'undici';
import {
  ContractCaller,
  type TransactionResult,
  type FeeRateSettings,
} from './chain/contract_caller.js';
import { OrderBuilder } from './chain/order_utils/order_builder.js';
import { calculateOrderAmounts, safeAmountToWei } from './chain/order_utils/utils.js';
import {
  type PlaceOrderDataInput,
  type OrderData,
  type OrderForApi,
  type ApiMarketData,
  type ApiQuoteToken,
  ZERO_ADDRESS,
} from './models/order.js';
import {
  TopicStatus,
  TopicStatusFilter,
  TopicType,
  TopicSortType,
  OrderSide,
  OrderType,
  SignatureType,
} from './models/enums.js';
import {
  DEFAULT_CONTRACT_ADDRESSES,
  SUPPORTED_CHAIN_IDS,
  type SupportedChainId,
} from './config.js';
import { InvalidParamError, OpenApiError } from './errors.js';
import { WebSocketClient, type WebSocketClientConfig } from './websocket/index.js';

// Import generated API client and types
import {
  getMarket as getMarketList,
  getMarketByMarketId as getBinaryMarketDetail,
  getMarketCategoricalByMarketId as getCategoricalMarketDetail,
  getMarketSlugBySlug as getMarketDetailBySlug,
  getQuoteToken as getQuoteTokenList,
  getTokenOrderbook as apiGetOrderbook,
  getTokenLatestPrice as apiGetLatestPrice,
  getTokenPriceHistory as apiGetPriceHistory,
  getPositions,
  getTrade as getTrades,
  getOrder as getOrders,
  getOrderByOrderId as apiGetOrderById,
  getUserBalance as getUserBalances,
  getUserAuth as apiGetUserAuth,
  postOrder as createOrder,
  postOrderCancel as apiCancelOrder,
  type OpenapiMarketDataOpenApi as MarketData,
  type OpenapiMarketListRespOpenApi as MarketListResponse,
  type OpenapiMarketDetailRespOpenApi as MarketDetailResponse,
  type OpenapiQuoteTokenDataOpenApi as QuoteTokenData,
  type OpenapiQuoteTokenListRespOpenApi as QuoteTokenListResponse,
  type OpenapiOrderbookRespOpenApi as OrderbookResponse,
  type OpenapiLatestPriceRespOpenApi as LatestPriceResponse,
  type OpenapiPriceHistoryRespOpenApi as PriceHistoryResponse,
  type OpenapiPositionsRespOpenApi as PositionsResponse,
  type OpenapiTradeListRespOpenApi as TradeListResponse,
  type OpenapiOrderListRespOpenApi as OrderListResponse,
  type OpenapiOrderDetailRespOpenApi as OrderDetailResponse,
  type OpenapiBalanceRespOpenApi as UserBalanceResponse,
  type V2GetApiKeyResp as UserAuthResponse,
  type OpenapiAddOrderRespOpenApi as CreateOrderResponse,
  type OpenapiCancelOrderRespOpenApi as CancelOrderResponse,
} from '@opinion-labs/opinion-api';
import { createClient, createConfig } from '@opinion-labs/opinion-api/client';
import type { Client as ApiClient } from '@opinion-labs/opinion-api/client';

/**
 * Client configuration options
 */
export interface ClientConfig {
  /** API host URL */
  host: string;
  /** API authentication key */
  apiKey: string;
  /** Blockchain chain ID (56 for BNB Chain) */
  chainId: SupportedChainId;
  /** RPC endpoint URL */
  rpcUrl: string;
  /** Private key for signing transactions */
  privateKey: Hex;
  /** Multi-signature wallet address */
  multiSigAddress: Address;
  /** Conditional tokens contract address (optional) */
  conditionalTokensAddress?: Address;
  /** MultiSend contract address (optional) */
  multiSendAddress?: Address;
  /** FeeManager contract address (optional) */
  feeManagerAddress?: Address;
  /** CTF Exchange contract address (optional, overrides API-provided address) */
  ctfExchangeAddress?: Address;
  /** Cache TTL for enable trading checks (seconds, default 3600) */
  enableTradingCheckInterval?: number;
  /** Cache TTL for quote tokens (seconds, default 3600) */
  quoteTokensCacheTtl?: number;
  /** Cache TTL for market data (seconds, default 300) */
  marketCacheTtl?: number;
  /** HTTP proxy URL for API requests (optional, e.g., 'http://127.0.0.1:7890') */
  proxyUrl?: string;
}

/**
 * Cache entry with timestamp
 */
interface CacheEntry<T> {
  data: T;
  timestamp: number;
}

/**
 * API response wrapper
 */
export interface ApiResponse<T> {
  errno: number;
  errmsg: string;
  result: T;
}

/**
 * Market list result
 */
export interface MarketListResult {
  total: number;
  list: MarketData[];
}

/**
 * Market detail result
 */
export interface MarketDetailResult {
  data: MarketData;
}

/**
 * Quote token list result
 */
export interface QuoteTokenListResult {
  total: number;
  list: QuoteTokenData[];
}

/**
 * Main SDK Client for Opinion CLOB operations
 */
export class Client {
  private apiClient: ApiClient;
  private chainId: SupportedChainId;
  private contractCaller: ContractCaller;
  private apiKey: string;
  private ctfExchangeAddress?: Address;
  private proxyUrl?: string;

  // Cache settings
  private quoteTokensCacheTtl: number;
  private marketCacheTtl: number;
  private quoteTokensCache: CacheEntry<QuoteTokenListResult> | null = null;
  private marketCache: Map<number, CacheEntry<MarketDetailResult>> = new Map();

  constructor(config: ClientConfig) {
    // Validate chain ID
    if (!SUPPORTED_CHAIN_IDS.includes(config.chainId)) {
      throw new InvalidParamError(`chainId must be one of ${SUPPORTED_CHAIN_IDS.join(', ')}`);
    }

    this.chainId = config.chainId;
    this.apiKey = config.apiKey;
    this.quoteTokensCacheTtl = config.quoteTokensCacheTtl ?? 3600;
    this.marketCacheTtl = config.marketCacheTtl ?? 300;
    this.ctfExchangeAddress = config.ctfExchangeAddress;
    this.proxyUrl = config.proxyUrl;

    // Create custom fetch function with proxy support
    let customFetch:
      | ((input: string | URL | Request, init?: RequestInit) => Promise<Response>)
      | undefined;
    if (config.proxyUrl) {
      const proxyDispatcher: Dispatcher = new ProxyAgent(config.proxyUrl);
      customFetch = async (
        input: string | URL | Request,
        init?: RequestInit
      ): Promise<Response> => {
        const request = new Request(input, init);
        const response = await undiciFetch(request.url, {
          method: request.method,
          headers: Object.fromEntries(request.headers.entries()),
          body: request.body ? await request.text() : undefined,
          dispatcher: proxyDispatcher,
        } as Parameters<typeof undiciFetch>[1]);
        return response as unknown as Response;
      };
    }

    // Initialize API client with configuration
    this.apiClient = createClient(
      createConfig({
        baseUrl: config.host,
        headers: {
          apikey: config.apiKey,
        },
        fetch: customFetch,
      })
    );

    // Get default contract addresses
    const defaults = DEFAULT_CONTRACT_ADDRESSES[config.chainId];

    this.contractCaller = new ContractCaller(
      config.rpcUrl,
      config.privateKey,
      getAddress(config.multiSigAddress),
      config.conditionalTokensAddress ?? defaults.conditionalTokens,
      config.multiSendAddress ?? defaults.multisend,
      config.feeManagerAddress ?? defaults.feeManager,
      config.chainId,
      config.enableTradingCheckInterval ?? 3600,
      config.proxyUrl
    );
  }

  /**
   * Check if cache is valid
   */
  private isCacheValid<T>(cache: CacheEntry<T> | null, ttl: number): boolean {
    if (!cache || ttl <= 0) return false;
    return Date.now() - cache.timestamp < ttl * 1000;
  }

  /**
   * Transform market data - filter to known fields and compute isIncentivized.
   */
  private transformMarketData(market: Record<string, unknown>): Record<string, unknown> {
    const incentiveFactor = market['incentiveFactor'];
    const isIncentivized = incentiveFactor !== null && incentiveFactor !== undefined;

    const transformed: Record<string, unknown> = {
      chainId: market['chainId'],
      conditionId: market['conditionId'],
      createdAt: market['createdAt'],
      cutoffAt: market['cutoffAt'],
      isIncentivized,
      marketId: market['marketId'],
      marketTitle: market['marketTitle'],
      noLabel: market['noLabel'],
      noTokenId: market['noTokenId'],
      questionId: market['questionId'],
      quoteToken: market['quoteToken'],
      resolvedAt: market['resolvedAt'],
      resultTokenId: market['resultTokenId'],
      rules: market['rules'],
      slug: market['slug'],
      status: market['status'],
      statusEnum: market['statusEnum'],
      volume: market['volume'],
      yesLabel: market['yesLabel'],
      yesTokenId: market['yesTokenId'],
    };

    if (market['collection'] !== undefined) {
      transformed['collection'] = market['collection'];
    }

    if (market['childMarkets'] && Array.isArray(market['childMarkets'])) {
      transformed['childMarkets'] = (market['childMarkets'] as Record<string, unknown>[]).map(
        (child) => this.transformChildMarketData(child)
      );
    }

    return transformed;
  }

  /**
   * Transform child market data - filter to known fields.
   */
  private transformChildMarketData(child: Record<string, unknown>): Record<string, unknown> {
    return {
      chainId: child['chainId'],
      conditionId: child['conditionId'],
      createdAt: child['createdAt'],
      cutoffAt: child['cutoffAt'],
      marketId: child['marketId'],
      marketTitle: child['marketTitle'],
      noLabel: child['noLabel'],
      noTokenId: child['noTokenId'],
      questionId: child['questionId'],
      quoteToken: child['quoteToken'],
      resolvedAt: child['resolvedAt'],
      resultTokenId: child['resultTokenId'],
      rules: child['rules'],
      slug: child['slug'],
      status: child['status'],
      statusEnum: child['statusEnum'],
      volume: child['volume'],
      yesLabel: child['yesLabel'],
      yesTokenId: child['yesTokenId'],
    };
  }

  /**
   * Handle API response - returns response as-is to match Python SDK behavior.
   * Does not throw on non-zero errno, caller must check errno if needed.
   */
  private handleApiResponse<T>(
    response: { data?: { errno?: number; errmsg?: string; result?: T }; error?: unknown },
    defaultErrorMsg: string
  ): ApiResponse<T> {
    if (response.error) {
      const errorData = response.error as { errno?: number; errmsg?: string };
      throw new OpenApiError(errorData?.errmsg || defaultErrorMsg);
    }

    const data = response.data;
    if (!data) {
      throw new OpenApiError(defaultErrorMsg);
    }

    // Return the full response including errno/errmsg to match Python SDK behavior.
    // Python SDK does not throw on non-zero errno, it returns the response as-is.
    return {
      errno: data.errno ?? 0,
      errmsg: data.errmsg ?? '',
      result: data.result as T,
    };
  }

  // ==================== Market Data APIs ====================

  /**
   * Get list of supported quote tokens
   */
  async getQuoteTokens(useCache = true): Promise<ApiResponse<QuoteTokenListResult>> {
    if (useCache && this.isCacheValid(this.quoteTokensCache, this.quoteTokensCacheTtl)) {
      return { errno: 0, errmsg: '', result: this.quoteTokensCache!.data };
    }

    const response = await getQuoteTokenList({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: {
        chainId: String(this.chainId),
      },
    });

    const result = this.handleApiResponse<QuoteTokenListResponse>(
      response,
      'Failed to get quote tokens'
    );

    const quoteTokenResult: QuoteTokenListResult = {
      total: result.result?.total ?? 0,
      list: result.result?.list ?? [],
    };

    if (this.quoteTokensCacheTtl > 0) {
      this.quoteTokensCache = { data: quoteTokenResult, timestamp: Date.now() };
    }

    return { errno: 0, errmsg: '', result: quoteTokenResult };
  }

  /**
   * Get markets with pagination
   */
  async getMarkets(
    options: {
      topicType?: TopicType;
      page?: number;
      limit?: number;
      status?: TopicStatusFilter;
      sortBy?: TopicSortType;
    } = {}
  ): Promise<ApiResponse<MarketListResult>> {
    const {
      topicType = TopicType.ALL,
      page = 1,
      limit = 20,
      status,
      sortBy = TopicSortType.BY_TIME_DESC,
    } = options;

    if (page < 1) throw new InvalidParamError('page must be >= 1');
    if (limit < 1 || limit > 20) throw new InvalidParamError('limit must be between 1 and 20');

    const queryParams: {
      chainId: string;
      marketType: number;
      page: number;
      limit: number;
      status?: string;
      sortBy: number;
    } = {
      chainId: String(this.chainId),
      marketType: topicType,
      page,
      limit,
      sortBy,
    };

    if (status && status !== (TopicStatusFilter.ALL as TopicStatusFilter)) {
      queryParams.status = status;
    }

    const response = await getMarketList({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: queryParams,
    });

    const result = this.handleApiResponse<MarketListResponse>(response, 'Failed to get markets');

    // Transform market data to match Python SDK's model format
    const rawList = (result.result?.list ?? []) as Record<string, unknown>[];
    const transformedList = rawList.map((market) => this.transformMarketData(market));

    return {
      errno: result.errno,
      errmsg: result.errmsg,
      result: {
        total: result.result?.total ?? 0,
        list: transformedList as MarketData[],
      },
    };
  }

  /**
   * Get market details by ID
   */
  async getMarket(marketId: number, useCache = true): Promise<ApiResponse<MarketDetailResult>> {
    if (!marketId) throw new InvalidParamError('marketId is required');

    if (useCache) {
      const cached = this.marketCache.get(marketId);
      if (this.isCacheValid(cached ?? null, this.marketCacheTtl)) {
        return { errno: 0, errmsg: '', result: cached!.data };
      }
    }

    const response = await getBinaryMarketDetail({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      path: { marketId },
    });

    const result = this.handleApiResponse<MarketDetailResponse>(
      response,
      'Failed to get market details'
    );

    // Return the response with errno/errmsg preserved (match Python SDK)
    // Transform market data if present
    let transformedData: MarketData | undefined;
    if (result.result?.data) {
      transformedData = this.transformMarketData(
        result.result.data as unknown as Record<string, unknown>
      ) as unknown as MarketData;
    }

    const marketResult: MarketDetailResult = {
      data: transformedData as MarketData,
    };

    if (result.errno === 0 && this.marketCacheTtl > 0) {
      this.marketCache.set(marketId, { data: marketResult, timestamp: Date.now() });
    }

    return { errno: result.errno, errmsg: result.errmsg, result: marketResult };
  }

  /**
   * Get categorical market details
   */
  async getCategoricalMarket(marketId: number): Promise<ApiResponse<MarketDetailResult>> {
    if (!marketId) throw new InvalidParamError('marketId is required');

    const response = await getCategoricalMarketDetail({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      path: { marketId },
    });

    const result = this.handleApiResponse<MarketDetailResponse>(
      response,
      'Failed to get categorical market details'
    );

    // Transform market data if present
    let transformedData: MarketData | undefined;
    if (result.result?.data) {
      transformedData = this.transformMarketData(
        result.result.data as unknown as Record<string, unknown>
      ) as unknown as MarketData;
    }

    return {
      errno: result.errno,
      errmsg: result.errmsg,
      result: {
        data: transformedData as MarketData,
      },
    };
  }

  /**
   * Get market details by slug (supports both binary and categorical markets)
   */
  async getMarketBySlug(slug: string): Promise<ApiResponse<MarketDetailResult>> {
    if (!slug || !slug.trim()) throw new InvalidParamError('slug is required');

    const trimmedSlug = slug.trim();
    const response = await getMarketDetailBySlug({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      path: { slug: trimmedSlug },
    });

    const result = this.handleApiResponse<MarketDetailResponse>(
      response,
      'Failed to get market details by slug'
    );

    let transformedData: MarketData | undefined;
    if (result.result?.data) {
      transformedData = this.transformMarketData(
        result.result.data as unknown as Record<string, unknown>
      ) as unknown as MarketData;
    }

    return {
      errno: result.errno,
      errmsg: result.errmsg,
      result: {
        data: transformedData as MarketData,
      },
    };
  }

  /**
   * Get orderbook for a token
   */
  async getOrderbook(tokenId: string): Promise<ApiResponse<OrderbookResponse>> {
    if (!tokenId) throw new InvalidParamError('tokenId is required');

    const response = await apiGetOrderbook({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: { token_id: tokenId },
    });

    return this.handleApiResponse<OrderbookResponse>(response, 'Failed to get orderbook');
  }

  /**
   * Get latest price for a token
   */
  async getLatestPrice(tokenId: string): Promise<ApiResponse<LatestPriceResponse>> {
    if (!tokenId) throw new InvalidParamError('tokenId is required');

    const response = await apiGetLatestPrice({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: { token_id: tokenId },
    });

    return this.handleApiResponse<LatestPriceResponse>(response, 'Failed to get latest price');
  }

  /**
   * Get price history for a token
   */
  async getPriceHistory(
    tokenId: string,
    options: {
      interval?: '1m' | '1h' | '1d' | '1w' | 'max';
      startAt?: number;
      endAt?: number;
    } = {}
  ): Promise<ApiResponse<PriceHistoryResponse>> {
    if (!tokenId) throw new InvalidParamError('tokenId is required');

    const { interval = '1h', startAt, endAt } = options;

    const response = await apiGetPriceHistory({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: {
        token_id: tokenId,
        interval,
        start_at: startAt,
        end_at: endAt,
      },
    });

    return this.handleApiResponse<PriceHistoryResponse>(response, 'Failed to get price history');
  }

  /**
   * Get fee rates from on-chain FeeManager contract
   */
  async getFeeRates(tokenId: string): Promise<FeeRateSettings> {
    if (!tokenId) throw new InvalidParamError('tokenId is required');
    return await this.contractCaller.getFeeRateSettings(BigInt(tokenId));
  }

  // ==================== User APIs ====================

  /**
   * Get user's positions
   */
  async getMyPositions(
    options: {
      marketId?: number;
      page?: number;
      limit?: number;
    } = {}
  ): Promise<ApiResponse<PositionsResponse>> {
    const { marketId, page = 1, limit = 10 } = options;

    const response = await getPositions({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: {
        chainId: String(this.chainId),
        page,
        limit,
        marketId: marketId && marketId > 0 ? marketId : undefined,
      },
    });

    return this.handleApiResponse<PositionsResponse>(response, 'Failed to get positions');
  }

  /**
   * Get user's trade history
   */
  async getMyTrades(
    options: {
      marketId?: number;
      page?: number;
      limit?: number;
    } = {}
  ): Promise<ApiResponse<TradeListResponse>> {
    const { marketId, page = 1, limit = 10 } = options;

    const response = await getTrades({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: {
        chainId: String(this.chainId),
        page,
        limit,
        marketId: marketId && marketId > 0 ? marketId : undefined,
      },
    });

    const apiResponse = this.handleApiResponse<TradeListResponse>(response, 'Failed to get trades');

    // Transform trade data to align with Python SDK:
    // - Use fee_formatted as fee (string, avoids precision loss)
    // - Remove feeFormatted field
    if (apiResponse.result?.list) {
      apiResponse.result.list = (apiResponse.result.list as Record<string, unknown>[]).map(
        (trade) => {
          const { feeFormatted, fee: _fee, ...rest } = trade;
          return {
            ...rest,
            fee: feeFormatted,
          };
        }
      ) as typeof apiResponse.result.list;
    }

    return apiResponse;
  }

  /**
   * Get user's orders with optional filters
   */
  async getMyOrders(
    options: {
      marketId?: number;
      status?: string;
      page?: number;
      limit?: number;
    } = {}
  ): Promise<ApiResponse<OrderListResponse>> {
    const { marketId, status, page = 1, limit = 10 } = options;

    const response = await getOrders({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: {
        chainId: String(this.chainId),
        page,
        limit,
        marketId: marketId && marketId > 0 ? marketId : undefined,
        status,
      },
    });

    return this.handleApiResponse<OrderListResponse>(response, 'Failed to get orders');
  }

  /**
   * Get detailed information about a specific order
   */
  async getOrderById(orderId: string): Promise<ApiResponse<OrderDetailResponse>> {
    if (!orderId) throw new InvalidParamError('orderId must be a non-empty string');

    const response = await apiGetOrderById({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      path: { orderId },
    });

    return this.handleApiResponse<OrderDetailResponse>(response, 'Failed to get order');
  }

  /**
   * Get user's balances
   */
  async getMyBalances(): Promise<ApiResponse<UserBalanceResponse>> {
    const response = await getUserBalances({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      query: {
        chain_id: String(this.chainId),
      },
    });

    return this.handleApiResponse<UserBalanceResponse>(response, 'Failed to get balances');
  }

  /**
   * Get authenticated user information
   */
  async getUserAuth(): Promise<ApiResponse<UserAuthResponse>> {
    const response = await apiGetUserAuth({
      client: this.apiClient,
    });

    return this.handleApiResponse<UserAuthResponse>(response, 'Failed to get user auth');
  }

  // ==================== Trading APIs ====================

  /**
   * Enable trading by approving necessary tokens
   */
  async enableTrading(): Promise<TransactionResult> {
    const quoteTokensResponse = await this.getQuoteTokens();
    if (quoteTokensResponse.errno !== 0) {
      throw new OpenApiError(
        `Failed to get quote tokens: ${quoteTokensResponse.errmsg || 'Unknown error'}`
      );
    }
    const quoteTokens = quoteTokensResponse.result.list;

    if (!quoteTokens || quoteTokens.length === 0) {
      throw new OpenApiError('No supported quote tokens found');
    }

    const supportedQuoteTokens = new Map<Address, Address>();
    for (const token of quoteTokens) {
      const qt = token as ApiQuoteToken;
      const tokenAddr = qt.quoteTokenAddress;
      const ctfExchangeAddr = qt.ctfExchangeAddress;

      if (tokenAddr) {
        // Use configured ctfExchangeAddress if provided, otherwise use API-provided address
        const exchangeAddr =
          this.ctfExchangeAddress ?? (ctfExchangeAddr ? getAddress(ctfExchangeAddr) : null);
        if (exchangeAddr) {
          supportedQuoteTokens.set(getAddress(tokenAddr), exchangeAddr);
        }
      }
    }

    return await this.contractCaller.enableTrading(supportedQuoteTokens);
  }

  /**
   * Place an order
   */
  async placeOrder(
    data: PlaceOrderDataInput,
    checkApproval = false
  ): Promise<ApiResponse<CreateOrderResponse>> {
    if (checkApproval) {
      await this.enableTrading();
    }

    // Get quote tokens and market data
    const quoteTokensResponse = await this.getQuoteTokens();
    if (quoteTokensResponse.errno !== 0) {
      throw new OpenApiError(
        `Failed to get quote tokens: ${quoteTokensResponse.errmsg || 'Unknown error'}`
      );
    }
    const quoteTokens = quoteTokensResponse.result.list;

    const marketResponse = await this.getMarket(data.marketId);
    if (marketResponse.errno !== 0) {
      throw new OpenApiError(`Failed to get market: ${marketResponse.errmsg || 'Unknown error'}`);
    }
    const market = marketResponse.result.data;

    const marketData = market as ApiMarketData;
    if (parseInt(marketData.chainId ?? '') !== this.chainId) {
      throw new OpenApiError('Cannot place order on different chain');
    }

    // Find quote token
    const quoteToken = quoteTokens.find((t) => {
      const qt = t as ApiQuoteToken;
      return qt.quoteTokenAddress?.toLowerCase() === marketData.quoteToken?.toLowerCase();
    });
    if (!quoteToken) {
      throw new OpenApiError('Quote token not found for this market');
    }

    const qt = quoteToken as ApiQuoteToken;
    const quoteTokenDecimal = qt.decimal ?? 6;
    const quoteTokenExchangeAddr = qt.ctfExchangeAddress ?? '';

    // Validate order inputs
    this.validateOrderInputs(data);

    // Calculate maker amount
    const makerAmount = this.calculateMakerAmount(data, quoteTokenDecimal);

    // Build and sign order - use configured ctfExchangeAddress if provided
    const exchangeAddress = this.ctfExchangeAddress ?? getAddress(quoteTokenExchangeAddr);
    const orderBuilder = new OrderBuilder(
      exchangeAddress,
      this.chainId,
      this.contractCaller.getSigner()
    );

    let takerAmount = 0n;
    let recalculatedMakerAmount = safeAmountToWei(makerAmount, quoteTokenDecimal);

    if (data.orderType === OrderType.LIMIT_ORDER) {
      const amounts = calculateOrderAmounts(
        parseFloat(data.price),
        recalculatedMakerAmount,
        data.side,
        quoteTokenDecimal
      );
      recalculatedMakerAmount = amounts.recalculatedMakerAmount;
      takerAmount = amounts.takerAmount;
    }

    const orderData: OrderData = {
      maker: this.contractCaller.getMultiSigAddress(),
      taker: ZERO_ADDRESS,
      tokenId: data.tokenId,
      makerAmount: recalculatedMakerAmount,
      takerAmount,
      side: data.side,
      feeRateBps: '0',
      signatureType: SignatureType.POLY_GNOSIS_SAFE,
      signer: this.contractCaller.getSigner().address(),
      nonce: '0',
      expiration: '0',
    };

    const signedOrder = await orderBuilder.buildSignedOrder(orderData);

    // Submit order to API
    const orderForApi: OrderForApi = {
      salt: signedOrder.order.salt.toString(),
      topicId: data.marketId,
      maker: signedOrder.order.maker,
      signer: signedOrder.order.signer,
      taker: signedOrder.order.taker,
      tokenId: signedOrder.order.tokenId.toString(),
      makerAmount: signedOrder.order.makerAmount.toString(),
      takerAmount: signedOrder.order.takerAmount.toString(),
      expiration: signedOrder.order.expiration.toString(),
      nonce: signedOrder.order.nonce.toString(),
      feeRateBps: signedOrder.order.feeRateBps.toString(),
      side: signedOrder.order.side.toString(),
      signatureType: signedOrder.order.signatureType.toString(),
      signature: signedOrder.signature,
      sign: signedOrder.signature,
      contractAddress: '',
      currencyAddress: getAddress(marketData.quoteToken ?? ''),
      price: data.orderType === OrderType.MARKET_ORDER ? '0' : data.price,
      tradingMethod: data.orderType,
      timestamp: Math.floor(Date.now() / 1000),
      safeRate: '0',
      orderExpTime: '0',
    };

    const response = await createOrder({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      body: orderForApi,
    });

    return this.handleApiResponse<CreateOrderResponse>(response, 'Failed to create order');
  }

  /**
   * Validate order inputs
   */
  private validateOrderInputs(data: PlaceOrderDataInput): void {
    // Reject market buy with base token amount
    if (
      data.side === OrderSide.BUY &&
      data.orderType === OrderType.MARKET_ORDER &&
      data.makerAmountInBaseToken
    ) {
      throw new InvalidParamError('makerAmountInBaseToken is not allowed for market buy');
    }

    // Reject market sell with quote token amount
    if (
      data.side === OrderSide.SELL &&
      data.orderType === OrderType.MARKET_ORDER &&
      data.makerAmountInQuoteToken
    ) {
      throw new InvalidParamError('makerAmountInQuoteToken is not allowed for market sell');
    }

    // Validate price for limit orders
    if (data.orderType === OrderType.LIMIT_ORDER) {
      const price = parseFloat(data.price);
      if (isNaN(price) || price <= 0) {
        throw new InvalidParamError(`Price must be positive for limit orders, got: ${data.price}`);
      }
    }
  }

  /**
   * Calculate maker amount from input
   */
  private calculateMakerAmount(data: PlaceOrderDataInput, _decimals: number): number {
    const minimalAmount = 1;

    if (data.side === OrderSide.BUY) {
      if (data.makerAmountInBaseToken) {
        const baseAmount = parseFloat(data.makerAmountInBaseToken);
        if (baseAmount < minimalAmount) {
          throw new InvalidParamError('makerAmountInBaseToken must be at least 1');
        }
        return baseAmount * parseFloat(data.price);
      } else if (data.makerAmountInQuoteToken) {
        const quoteAmount = parseFloat(data.makerAmountInQuoteToken);
        if (quoteAmount < minimalAmount) {
          throw new InvalidParamError('makerAmountInQuoteToken must be at least 1');
        }
        return quoteAmount;
      }
      throw new InvalidParamError(
        'Either makerAmountInBaseToken or makerAmountInQuoteToken must be provided for BUY orders'
      );
    } else {
      if (data.makerAmountInBaseToken) {
        const baseAmount = parseFloat(data.makerAmountInBaseToken);
        if (baseAmount < minimalAmount) {
          throw new InvalidParamError('makerAmountInBaseToken must be at least 1');
        }
        return baseAmount;
      } else if (data.makerAmountInQuoteToken) {
        const quoteAmount = parseFloat(data.makerAmountInQuoteToken);
        if (quoteAmount < minimalAmount) {
          throw new InvalidParamError('makerAmountInQuoteToken must be at least 1');
        }
        const price = parseFloat(data.price);
        if (price === 0) {
          throw new InvalidParamError(
            'Price cannot be zero for SELL orders with makerAmountInQuoteToken'
          );
        }
        return quoteAmount / price;
      }
      throw new InvalidParamError(
        'Either makerAmountInBaseToken or makerAmountInQuoteToken must be provided for SELL orders'
      );
    }
  }

  /**
   * Cancel an order
   */
  async cancelOrder(orderId: string): Promise<ApiResponse<CancelOrderResponse>> {
    if (!orderId) throw new InvalidParamError('orderId must be a non-empty string');

    const response = await apiCancelOrder({
      client: this.apiClient,
      headers: { apikey: this.apiKey },
      body: { orderId },
    });

    return this.handleApiResponse<CancelOrderResponse>(response, 'Failed to cancel order');
  }

  /**
   * Place multiple orders in batch
   */
  async placeOrdersBatch(
    orders: PlaceOrderDataInput[],
    checkApproval = false
  ): Promise<Array<{ index: number; success: boolean; result?: unknown; error?: string }>> {
    if (!orders || orders.length === 0) {
      throw new InvalidParamError('orders list cannot be empty');
    }

    if (checkApproval) {
      await this.enableTrading();
    }

    const results: Array<{ index: number; success: boolean; result?: unknown; error?: string }> =
      [];

    for (let i = 0; i < orders.length; i++) {
      try {
        const result = await this.placeOrder(orders[i], false);
        results.push({ index: i, success: true, result });
      } catch (error) {
        const errorMsg = error instanceof Error ? error.message : String(error);
        results.push({ index: i, success: false, error: errorMsg });
      }
    }

    return results;
  }

  /**
   * Cancel multiple orders in batch
   */
  async cancelOrdersBatch(
    orderIds: string[]
  ): Promise<Array<{ index: number; success: boolean; result?: unknown; error?: string }>> {
    if (!orderIds || orderIds.length === 0) {
      throw new InvalidParamError('orderIds list cannot be empty');
    }

    const results: Array<{ index: number; success: boolean; result?: unknown; error?: string }> =
      [];

    for (let i = 0; i < orderIds.length; i++) {
      try {
        const result = await this.cancelOrder(orderIds[i]);
        results.push({ index: i, success: true, result });
      } catch (error) {
        const errorMsg = error instanceof Error ? error.message : String(error);
        results.push({ index: i, success: false, error: errorMsg });
      }
    }

    return results;
  }

  /**
   * Cancel all open orders, optionally filtered by market and/or side.
   * Uses pagination to fetch all orders (max 20 per page).
   */
  async cancelAllOrders(
    options: {
      marketId?: number;
      side?: OrderSide;
    } = {}
  ): Promise<{
    totalOrders: number;
    cancelled: number;
    failed: number;
    results: Array<{ index: number; success: boolean; result?: unknown; error?: string }>;
  }> {
    const { marketId, side } = options;

    // Collect all open orders using pagination
    const allOrdersList: Array<{ orderId: string; side: number }> = [];
    let page = 1;
    const limit = 20; // Backend max page size
    const maxPages = 100; // Safety limit

    while (page <= maxPages) {
      const pageOrders = await this.getMyOrders({
        marketId: marketId,
        status: '1', // 1 = pending/open orders
        limit,
        page,
      });

      // Check for API errors
      if (pageOrders.errno !== 0) {
        throw new OpenApiError(`Failed to get orders: ${pageOrders.errmsg || 'Unknown error'}`);
      }

      const ordersList = pageOrders.result?.list as
        | Array<{ orderId?: string; side?: number }>
        | undefined;

      if (!ordersList || ordersList.length === 0) {
        break;
      }

      for (const order of ordersList) {
        if (order.orderId) {
          allOrdersList.push({ orderId: order.orderId, side: order.side ?? 0 });
        }
      }

      if (ordersList.length < limit) {
        break;
      }

      page++;
    }

    if (allOrdersList.length === 0) {
      return {
        totalOrders: 0,
        cancelled: 0,
        failed: 0,
        results: [],
      };
    }

    // Filter by side if specified
    let filteredOrders = allOrdersList;
    if (side !== undefined) {
      filteredOrders = allOrdersList.filter((order) => order.side === (side as number));
    }

    // Extract order IDs
    const orderIds = filteredOrders.map((order) => order.orderId).filter(Boolean);

    if (orderIds.length === 0) {
      return {
        totalOrders: 0,
        cancelled: 0,
        failed: 0,
        results: [],
      };
    }

    // Cancel all orders in batch
    const results = await this.cancelOrdersBatch(orderIds);

    // Count successes and failures
    const cancelled = results.filter((r) => r.success).length;
    const failed = results.filter((r) => !r.success).length;

    return {
      totalOrders: orderIds.length,
      cancelled,
      failed,
      results,
    };
  }

  // ==================== Token Operations ====================

  /**
   * Split collateral into outcome tokens
   */
  async split(marketId: number, amount: bigint, checkApproval = true): Promise<TransactionResult> {
    if (!marketId || marketId <= 0) {
      throw new InvalidParamError('marketId must be a positive integer');
    }
    if (!amount || amount <= 0n) {
      throw new InvalidParamError('amount must be a positive integer');
    }

    if (checkApproval) {
      await this.enableTrading();
    }

    const marketResponse = await this.getMarket(marketId);
    if (marketResponse.errno !== 0) {
      throw new OpenApiError(`Failed to get market: ${marketResponse.errmsg || 'Unknown error'}`);
    }
    const marketData = marketResponse.result.data as ApiMarketData;

    if (parseInt(marketData.chainId ?? '') !== this.chainId) {
      throw new OpenApiError('Cannot split on different chain');
    }

    const validStatuses = [TopicStatus.ACTIVATED, TopicStatus.RESOLVING, TopicStatus.RESOLVED];
    if (!validStatuses.includes(marketData.status as TopicStatus)) {
      throw new OpenApiError('Cannot split on non-activated/resolving/resolved market');
    }

    return await this.contractCaller.split(
      getAddress(marketData.quoteToken ?? ''),
      `0x${marketData.conditionId}`,
      amount
    );
  }

  /**
   * Merge outcome tokens back to collateral
   */
  async merge(marketId: number, amount: bigint, checkApproval = true): Promise<TransactionResult> {
    if (!marketId || marketId <= 0) {
      throw new InvalidParamError('marketId must be a positive integer');
    }
    if (!amount || amount <= 0n) {
      throw new InvalidParamError('amount must be a positive integer');
    }

    if (checkApproval) {
      await this.enableTrading();
    }

    const marketResponse = await this.getMarket(marketId);
    if (marketResponse.errno !== 0) {
      throw new OpenApiError(`Failed to get market: ${marketResponse.errmsg || 'Unknown error'}`);
    }
    const marketData = marketResponse.result.data as ApiMarketData;

    if (parseInt(marketData.chainId ?? '') !== this.chainId) {
      throw new OpenApiError('Cannot merge on different chain');
    }

    const validStatuses = [TopicStatus.ACTIVATED, TopicStatus.RESOLVING, TopicStatus.RESOLVED];
    if (!validStatuses.includes(marketData.status as TopicStatus)) {
      throw new OpenApiError('Cannot merge on non-activated/resolving/resolved market');
    }

    return await this.contractCaller.merge(
      getAddress(marketData.quoteToken ?? ''),
      `0x${marketData.conditionId}`,
      amount
    );
  }

  /**
   * Redeem winning outcome tokens
   */
  async redeem(marketId: number, checkApproval = true): Promise<TransactionResult> {
    if (!marketId || marketId <= 0) {
      throw new InvalidParamError('marketId must be a positive integer');
    }

    if (checkApproval) {
      await this.enableTrading();
    }

    const marketResponse = await this.getMarket(marketId);
    if (marketResponse.errno !== 0) {
      throw new OpenApiError(`Failed to get market: ${marketResponse.errmsg || 'Unknown error'}`);
    }
    const marketData = marketResponse.result.data as ApiMarketData;

    if (parseInt(marketData.chainId ?? '') !== this.chainId) {
      throw new OpenApiError('Cannot redeem on different chain');
    }

    if (marketData.status !== (TopicStatus.RESOLVED as number)) {
      throw new OpenApiError('Cannot redeem on non-resolved market');
    }

    return await this.contractCaller.redeem(
      getAddress(marketData.quoteToken ?? ''),
      `0x${marketData.conditionId}`
    );
  }

  // ==================== WebSocket ====================

  /**
   * Create a WebSocket client for real-time market data.
   *
   * This method creates a WebSocket client that is pre-configured with the
   * API key and wallet address from this client instance.
   *
   * @param options - WebSocket client configuration options
   * @returns WebSocketClient instance
   *
   * @example
   * ```typescript
   * const wsClient = client.createWebSocketClient({
   *   heartbeatInterval: 30,
   *   onMessage: (msg) => console.log('Received:', msg),
   *   onOpen: () => console.log('Connected'),
   *   onClose: () => console.log('Disconnected'),
   *   onError: (error) => console.log('Error:', error),
   * });
   *
   * await wsClient.connect();
   *
   * // Subscribe to market updates
   * wsClient.subscribeMarketDepthDiff(1274, (msg) => {
   *   console.log(`Depth update: ${msg.side} ${msg.price} @ ${msg.size}`);
   * });
   * ```
   */
  createWebSocketClient(
    options: Omit<WebSocketClientConfig, 'apiKey' | 'walletAddress'> & {
      wsUrl?: string;
    } = {}
  ): WebSocketClient {
    const walletAddress = this.contractCaller.getMultiSigAddress();

    return new WebSocketClient({
      apiKey: this.apiKey,
      walletAddress,
      proxyUrl: this.proxyUrl,
      ...options,
    });
  }
}
