// @buidlrrr/rain-sdk/dist/markets/getMarketVolume.js
import { TradePoolAbi } from '../abi/TradeMarketsAbi.js';
import { multicallRead } from '../utils/multicall.js';
export async function getMarketVolume(params) {
    const { marketId, apiUrl, rpcUrl } = params;
    if (!apiUrl)
        throw new Error('apiUrl is required');
    if (!rpcUrl)
        throw new Error('rpcUrl is required');
    if (!marketId)
        throw new Error('marketId is required');
    // 1. Fetch from API
    const res = await fetch(`${apiUrl}/pools/pool/${marketId}`);
    if (!res.ok) {
        throw new Error(`Failed to fetch market: ${res.status}`);
    }
    const json = await res.json();
    const apiData = json.data ?? json;
    const contractAddress = apiData.contractAddress;
    if (!contractAddress) {
        throw new Error('Market response missing contractAddress');
    }
    const apiOptions = apiData.options ?? apiData.choices ?? [];
    // 2. Multicall batch: allFunds() + totalFunds(optionId) per option
    const contracts = [
        {
            address: contractAddress,
            abi: TradePoolAbi,
            functionName: 'allFunds',
            args: [],
        },
        ...apiOptions.map((opt, i) => ({
            address: contractAddress,
            abi: TradePoolAbi,
            functionName: 'totalFunds',
            args: [BigInt(opt.choiceIndex ?? i)],
        })),
    ];
    const results = await multicallRead(rpcUrl, contracts);
    // 3. Parse results
    const allFundsResult = results[0];
    const totalVolume = allFundsResult?.status === 'success' ? BigInt(allFundsResult.result) : 0n;
    const optionVolumes = apiOptions.map((opt, i) => {
        const result = results[i + 1];
        return {
            choiceIndex: opt.choiceIndex ?? i,
            optionName: opt.optionName ?? `Option ${i}`,
            volume: result?.status === 'success' ? BigInt(result.result) : 0n,
        };
    });
    return {
        marketId,
        contractAddress,
        totalVolume,
        optionVolumes,
    };
}
