// @buidlrrr/rain-sdk/dist/Rain.js
import { getMarkets } from './markets/getMarkets.js';
import { getMarketById } from './markets/getMarketById.js';
import { getUserInvestments } from './markets/getUserInvestments.js';
import { getMarketDetails } from './markets/getMarketDetails.js';
import { getMarketPrices } from './markets/getMarketPrices.js';
import { getMarketVolume } from './markets/getMarketVolume.js';
import { getMarketLiquidity } from './markets/getMarketLiquidity.js';
import { getMarketAddress } from './markets/getMarketAddress.js';
import { getMarketId } from './markets/getMarketId.js';
import { getProtocolStats } from './markets/getProtocolStats.js';
import { buildAddLiquidityRawTx, buildEnterOptionRawTx, buildLimitBuyOrderRawTx, buildSellOptionRawTx, buildCancelBuyOrdersRawTx, buildCancelSellOrdersRawTx } from './tx/buildRawTransactions.js';
import { buildCancelAllOpenOrdersRawTx } from './tx/buildCancelAllOpenOrdersRawTx.js';
import { buildCreateDisputeRawTx, buildCreateAppealRawTx } from './tx/Dispute/buildDisputeRawTx.js';
import { buildExtendTimeRawTx } from './tx/ExtendTime/buildExtendTimeRawTx.js';
import { buildApproveRawTx } from './tx/buildApprovalRawTx.js';
import { buildCreateMarketRawTx } from './tx/CreateMarket/buildCreateMarketRawTx.js';
import { ALLOWED_ENVIRONMENTS, ENV_CONFIG, getRandomRpc } from './config/environments.js';
import { buildClaimRawTx } from './tx/ClaimFunds/buildClaimFundsRawTx.js';
import { buildCloseMarketRawTx, buildChooseWinnerRawTx, buildResolveMarketRawTx } from './tx/ResolveMarket/buildResolveMarketRawTx.js';
import { buildTransferRawTx } from './tx/buildTransferRawTx.js';
import { getSmartAccountBalance } from './accounts/getSmartAccountBalance.js';
import { getEOAFromSmartAccount } from './accounts/getEOAFromSmartAccount.js';
import { getPositions } from './positions/getPositions.js';
import { getPositionByMarket } from './positions/getPositionByMarket.js';
import { getPortfolioValue } from './positions/getPortfolioValue.js';
import { getLPPosition } from './positions/getLPPosition.js';
import { getTransactions } from './transactions/getTransactions.js';
import { getTransactionDetails } from './transactions/getTransactionDetails.js';
import { getMarketTransactions } from './transactions/getMarketTransactions.js';
import { getTradeHistory } from './transactions/getTradeHistory.js';
import { createWsClient, subscribeToMarketEvents, subscribePriceUpdates } from './utils/websocket.js';
import { getPriceHistory } from './priceHistory/getPriceHistory.js';
import { getPnL } from './pnl/getPnL.js';
import { getLeaderboard } from './leaderboard/getLeaderboard.js';
import { loginUser } from './auth/login.js';
export class Rain {
    environment;
    marketFactory;
    apiUrl;
    distute_initial_timer;
    rpcUrl;
    subgraphUrl;
    subgraphApiKey;
    wsRpcUrl;
    wsReconnect;
    wsClient = null;
    usdtSymbol;
    constructor(config = {}) {
        const { environment = "development", rpcUrl, apiUrl, subgraphUrl, subgraphApiKey, wsRpcUrl, wsReconnect } = config;
        function isValidEnvironment(env) {
            return ALLOWED_ENVIRONMENTS.includes(env);
        }
        if (!isValidEnvironment(environment)) {
            throw new Error(`Invalid environment "${environment}". Allowed values: ${ALLOWED_ENVIRONMENTS.join(", ")}`);
        }
        this.environment = environment;
        this.rpcUrl = rpcUrl ?? getRandomRpc();
        const envConfig = ENV_CONFIG[this.environment];
        this.marketFactory = envConfig.market_factory_address;
        this.apiUrl = apiUrl ?? envConfig.apiUrl;
        this.distute_initial_timer = envConfig.dispute_initial_timer;
        this.subgraphUrl = subgraphUrl ?? envConfig.subgraphUrl;
        this.subgraphApiKey = subgraphApiKey;
        this.wsRpcUrl = wsRpcUrl;
        this.wsReconnect = wsReconnect;
        this.usdtSymbol = envConfig.usdt_symbol;
    }
    async getPublicMarkets(params) {
        return getMarkets({ ...params, apiUrl: this.apiUrl });
    }
    async getMarketById(params) {
        return getMarketById({ ...params, apiUrl: this.apiUrl });
    }
    async getUserInvestments(params) {
        return getUserInvestments({ ...params, apiUrl: this.apiUrl });
    }
    buildApprovalTx(params) {
        return buildApproveRawTx(params);
    }
    buildBuyOptionRawTx(params) {
        return buildEnterOptionRawTx(params);
    }
    buildLimitBuyOptionTx(params) {
        return buildLimitBuyOrderRawTx(params);
    }
    buildSellOptionTx(params) {
        return buildSellOptionRawTx(params);
    }
    buildCancelBuyOrdersTx(params) {
        return buildCancelBuyOrdersRawTx(params);
    }
    buildCancelSellOrdersTx(params) {
        return buildCancelSellOrdersRawTx(params);
    }
    async buildCancelAllOpenOrdersTx(params) {
        return buildCancelAllOpenOrdersRawTx({ ...params, apiUrl: this.apiUrl });
    }
    buildAddLiquidityTx(params) {
        return buildAddLiquidityRawTx(params);
    }
    buildCreateMarketTx(params) {
        return buildCreateMarketRawTx({ ...params, factoryContractAddress: this.marketFactory, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl, disputeTimer: this.distute_initial_timer });
    }
    async buildClaimTx(params) {
        return buildClaimRawTx({ ...params, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async buildCloseMarketTx(params) {
        return buildCloseMarketRawTx({ ...params, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async buildChooseWinnerTx(params) {
        return buildChooseWinnerRawTx({ ...params, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async buildResolveMarketTx(params) {
        return buildResolveMarketRawTx({ ...params, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async buildCreateDisputeTx(params) {
        return buildCreateDisputeRawTx({ ...params, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl, usdtSymbol: this.usdtSymbol });
    }
    async buildCreateAppealTx(params) {
        return buildCreateAppealRawTx({ ...params, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl, usdtSymbol: this.usdtSymbol });
    }
    async buildExtendTimeTx(params) {
        return buildExtendTimeRawTx({ ...params, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    buildDepositToSmartAccountTx(params) {
        return buildTransferRawTx({
            tokenAddress: params.tokenAddress,
            recipient: params.smartAccountAddress,
            amount: params.amount,
        });
    }
    buildWithdrawFromSmartAccountTx(params) {
        return buildTransferRawTx({
            tokenAddress: params.tokenAddress,
            recipient: params.eoaAddress,
            amount: params.amount,
        });
    }
    async getMarketAddress(marketId) {
        return getMarketAddress({ marketId, apiUrl: this.apiUrl });
    }
    async getMarketId(marketAddress) {
        return getMarketId({ marketAddress, apiUrl: this.apiUrl });
    }
    async getMarketDetails(marketId) {
        return getMarketDetails({ marketId, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async getMarketPrices(marketId) {
        return getMarketPrices({ marketId, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async getMarketVolume(marketId) {
        return getMarketVolume({ marketId, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async getMarketLiquidity(marketId) {
        return getMarketLiquidity({ marketId, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async getProtocolStats() {
        return getProtocolStats({
            apiUrl: this.apiUrl,
            rpcUrl: this.rpcUrl,
            subgraphUrl: this.subgraphUrl,
            subgraphApiKey: this.subgraphApiKey,
        });
    }
    async getSmartAccountBalance(params) {
        return getSmartAccountBalance({ ...params, rpcUrl: this.rpcUrl });
    }
    async getEOAFromSmartAccount(smartAccountAddress) {
        return getEOAFromSmartAccount({ smartAccountAddress, rpcUrl: this.rpcUrl });
    }
    async getPositions(address) {
        return getPositions({ address, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async getPositionByMarket(address, marketId) {
        return getPositionByMarket({ address, marketId, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async getLPPosition(address, marketId) {
        return getLPPosition({ address, marketId, apiUrl: this.apiUrl, rpcUrl: this.rpcUrl });
    }
    async getPortfolioValue(params) {
        return getPortfolioValue({
            ...params,
            apiUrl: this.apiUrl,
            rpcUrl: this.rpcUrl,
        });
    }
    async getTransactions(params) {
        const subgraphUrl = params.subgraphUrl ?? this.subgraphUrl;
        if (!subgraphUrl) {
            throw new Error('subgraphUrl is required — pass it in the Rain constructor config or in the method params');
        }
        return getTransactions({ ...params, subgraphUrl, subgraphApiKey: params.subgraphApiKey ?? this.subgraphApiKey });
    }
    async getTransactionDetails(params) {
        const subgraphUrl = params.subgraphUrl ?? this.subgraphUrl;
        const rpcUrl = params.rpcUrl ?? this.rpcUrl;
        if (!subgraphUrl) {
            throw new Error('subgraphUrl is required — pass it in the Rain constructor config or in the method params');
        }
        if (!rpcUrl) {
            throw new Error('rpcUrl is required — pass it in the Rain constructor config or in the method params');
        }
        return getTransactionDetails({ ...params, subgraphUrl, subgraphApiKey: params.subgraphApiKey ?? this.subgraphApiKey, rpcUrl });
    }
    async getMarketTransactions(params) {
        const subgraphUrl = params.subgraphUrl ?? this.subgraphUrl;
        if (!subgraphUrl) {
            throw new Error('subgraphUrl is required — pass it in the Rain constructor config or in the method params');
        }
        return getMarketTransactions({ ...params, subgraphUrl, subgraphApiKey: params.subgraphApiKey ?? this.subgraphApiKey });
    }
    async getPriceHistory(params) {
        const subgraphUrl = params.subgraphUrl ?? this.subgraphUrl;
        if (!subgraphUrl) {
            throw new Error('subgraphUrl is required — pass it in the Rain constructor config or in the method params');
        }
        return getPriceHistory({ ...params, subgraphUrl, subgraphApiKey: params.subgraphApiKey ?? this.subgraphApiKey, apiUrl: params.apiUrl ?? this.apiUrl });
    }
    async getTradeHistory(params) {
        const subgraphUrl = params.subgraphUrl ?? this.subgraphUrl;
        if (!subgraphUrl) {
            throw new Error('subgraphUrl is required — pass it in the Rain constructor config or in the method params');
        }
        return getTradeHistory({ ...params, subgraphUrl, subgraphApiKey: params.subgraphApiKey ?? this.subgraphApiKey });
    }
    async getPnL(params) {
        const subgraphUrl = params.subgraphUrl ?? this.subgraphUrl;
        const rpcUrl = params.rpcUrl ?? this.rpcUrl;
        if (!subgraphUrl) {
            throw new Error('subgraphUrl is required — pass it in the Rain constructor config or in the method params');
        }
        if (!rpcUrl) {
            throw new Error('rpcUrl is required — pass it in the Rain constructor config or in the method params');
        }
        return getPnL({
            ...params,
            subgraphUrl,
            subgraphApiKey: params.subgraphApiKey ?? this.subgraphApiKey,
            apiUrl: params.apiUrl ?? this.apiUrl,
            rpcUrl,
        });
    }
    async getLeaderboard(params) {
        const subgraphUrl = params.subgraphUrl ?? this.subgraphUrl;
        if (!subgraphUrl) {
            throw new Error('subgraphUrl is required — pass it in the Rain constructor config or in the method params');
        }
        return getLeaderboard({ ...params, subgraphUrl, subgraphApiKey: params.subgraphApiKey ?? this.subgraphApiKey });
    }
    getWsClient() {
        if (this.wsClient) {
            return this.wsClient;
        }
        if (!this.wsRpcUrl) {
            throw new Error('wsRpcUrl is required for WebSocket subscriptions — pass it in the Rain constructor config');
        }
        this.wsClient = createWsClient({
            wsRpcUrl: this.wsRpcUrl,
            reconnect: this.wsReconnect,
        });
        return this.wsClient;
    }
    subscribeToMarketEvents(params) {
        const client = this.getWsClient();
        return subscribeToMarketEvents(client, params);
    }
    subscribePriceUpdates(params) {
        const client = this.getWsClient();
        return subscribePriceUpdates(client, this.rpcUrl, params);
    }
    async destroyWebSocket() {
        if (this.wsClient) {
            const transport = this.wsClient.transport;
            if ('getSocket' in transport && typeof transport.getSocket === 'function') {
                try {
                    const socket = await transport.getSocket();
                    socket.close();
                }
                catch {
                    // Socket may already be closed
                }
            }
            this.wsClient = null;
        }
    }
    async login(params) {
        return loginUser({ ...params, apiUrl: this.apiUrl });
    }
}
