// @buidlrrr/rain-sdk/dist/RainAA.js
import { custom, createWalletClient } from 'viem';
export class RainAA {
    config;
    _client = null;
    _address = null;
    constructor(config) {
        if (!config.walletClient)
            throw new Error('walletClient is required');
        if (!config.alchemyApiKey)
            throw new Error('alchemyApiKey is required');
        if (!config.paymasterPolicyId)
            throw new Error('paymasterPolicyId is required');
        if (!config.chain)
            throw new Error('chain is required');
        this.config = config;
    }
    /**
     * Initializes the Smart Account
     */
    async connect() {
        if (this._address && this._client) {
            return this._address;
        }
        try {
            const [{ WalletClientSigner }, { alchemy }, { createSmartWalletClient }] = await Promise.all([
                import('@alchemy/aa-core'),
                import('@account-kit/infra'),
                import('@account-kit/wallet-client'),
            ]);
            const signer = new WalletClientSigner(createWalletClient({
                transport: custom(this.config.walletClient),
            }), 'wallet');
            const client = createSmartWalletClient({
                chain: this.config.chain,
                signer: signer,
                policyId: this.config.paymasterPolicyId,
                transport: alchemy({
                    apiKey: this.config.alchemyApiKey,
                    nodeRpcUrl: this.config.rpcUrl,
                }),
            });
            const account = await client.requestAccount();
            if (!account?.address) {
                throw new Error('Failed to create smart account');
            }
            this._client = client;
            this._address = account.address;
            return account.address;
        }
        catch (err) {
            console.error('[Rain SDK] connect failed:', err);
            throw err;
        }
    }
    /**
     * Returns smart account address
     */
    get address() {
        if (!this._address) {
            throw new Error('Rain not connected. Call rain.connect() first.');
        }
        return this._address;
    }
    /**
     * Returns smart account client
     */
    get client() {
        if (!this._client) {
            throw new Error('Rain not connected. Call rain.connect() first.');
        }
        return this._client;
    }
    /**
     * Sends a raw transaction from the smart account.
     */
    async sendTransaction(rawTx) {
        if (!this._client) {
            throw new Error('Rain not connected. Call rain.connect() first.');
        }
        const hash = await this._client.sendTransaction({
            to: rawTx.to,
            data: rawTx.data,
            value: rawTx.value,
        });
        return hash;
    }
    /**
     * Reset connection (optional)
     */
    disconnect() {
        this._client = null;
        this._address = null;
    }
}
