// @buidlrrr/rain-sdk/dist/transactions/getTransactionDetails.js
import { subgraphQuery } from '../utils/subgraph.js';
import { ENTITY_FIELDS, parseEntity } from './shared.js';
function buildSubgraphQuery(txHash) {
    const hash = txHash.toLowerCase();
    const parts = [];
    for (const entity of Object.keys(ENTITY_FIELDS)) {
        const fields = ENTITY_FIELDS[entity];
        parts.push(`  ${entity}(where: { transactionHash: "${hash}" }) { ${fields} }`);
    }
    return `{\n${parts.join('\n')}\n}`;
}
async function fetchReceipt(rpcUrl, txHash) {
    const res = await fetch(rpcUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            jsonrpc: '2.0',
            id: 1,
            method: 'eth_getTransactionReceipt',
            params: [txHash],
        }),
    });
    const json = await res.json();
    if (!json.result) {
        throw new Error(`Transaction receipt not found for ${txHash}`);
    }
    return json.result;
}
async function fetchBlockTimestamp(rpcUrl, blockNumber) {
    const res = await fetch(rpcUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            jsonrpc: '2.0',
            id: 2,
            method: 'eth_getBlockByNumber',
            params: [blockNumber, false],
        }),
    });
    const json = await res.json();
    if (!json.result) {
        throw new Error(`Block not found for ${blockNumber}`);
    }
    return json.result.timestamp;
}
export async function getTransactionDetails(params) {
    const { transactionHash, subgraphUrl, subgraphApiKey, rpcUrl } = params;
    const query = buildSubgraphQuery(transactionHash);
    // Fetch on-chain receipt and subgraph events in parallel
    const [receipt, subgraphData] = await Promise.all([
        fetchReceipt(rpcUrl, transactionHash),
        subgraphQuery(subgraphUrl, query, subgraphApiKey),
    ]);
    // Fetch block timestamp using the block number from the receipt
    const timestamp = await fetchBlockTimestamp(rpcUrl, receipt.blockNumber);
    // Parse all subgraph events
    const events = [];
    const seenIds = new Set();
    for (const entity of Object.keys(ENTITY_FIELDS)) {
        const items = subgraphData[entity] ?? [];
        for (const item of items) {
            if (!seenIds.has(item.id)) {
                seenIds.add(item.id);
                events.push(parseEntity(entity, item));
            }
        }
    }
    // Sort events by type for consistent ordering
    events.sort((a, b) => a.type.localeCompare(b.type));
    return {
        transactionHash: transactionHash,
        blockNumber: BigInt(receipt.blockNumber),
        timestamp: BigInt(timestamp),
        from: receipt.from,
        to: receipt.to,
        status: receipt.status === '0x1' ? 'success' : 'failed',
        gasUsed: BigInt(receipt.gasUsed),
        effectiveGasPrice: BigInt(receipt.effectiveGasPrice),
        events,
    };
}
