// @buidlrrr/rain-sdk/dist/utils/websocket.js
import { createPublicClient, webSocket } from 'viem';
import { arbitrum } from 'viem/chains';
import { TradePoolAbi } from '../abi/TradeMarketsAbi.js';
import { multicallRead } from './multicall.js';
const DEFAULT_RECONNECT = {
    attempts: 20,
    delay: 3_000,
};
export function createWsClient(config) {
    const { wsRpcUrl, reconnect = true, keepAlive = true } = config;
    let reconnectConfig;
    if (reconnect === false) {
        reconnectConfig = false;
    }
    else if (reconnect === true) {
        reconnectConfig = DEFAULT_RECONNECT;
    }
    else {
        reconnectConfig = {
            attempts: reconnect.attempts ?? DEFAULT_RECONNECT.attempts,
            delay: reconnect.delay ?? DEFAULT_RECONNECT.delay,
        };
    }
    let keepAliveConfig;
    if (keepAlive === false) {
        keepAliveConfig = false;
    }
    else if (keepAlive === true) {
        keepAliveConfig = { interval: 30_000 };
    }
    else {
        keepAliveConfig = { interval: keepAlive.interval ?? 30_000 };
    }
    const transportOptions = {};
    if (reconnectConfig !== false) {
        transportOptions.retryCount = reconnectConfig.attempts;
        transportOptions.retryDelay = reconnectConfig.delay;
        transportOptions.reconnect = true;
    }
    else {
        transportOptions.reconnect = false;
    }
    if (keepAliveConfig !== false) {
        transportOptions.keepAlive = keepAliveConfig.interval;
    }
    return createPublicClient({
        chain: arbitrum,
        transport: webSocket(wsRpcUrl, transportOptions),
    });
}
function parseLogToMarketTradeEvent(log) {
    const eventName = log.eventName;
    const args = log.args ?? {};
    return {
        eventName,
        marketAddress: log.address,
        blockNumber: log.blockNumber ?? 0n,
        transactionHash: log.transactionHash ?? '0x',
        logIndex: log.logIndex ?? 0,
        args,
    };
}
export function subscribeToMarketEvents(client, params) {
    const { marketAddress, eventNames, onEvent, onError } = params;
    const eventNameSet = eventNames ? new Set(eventNames) : null;
    const unwatch = client.watchContractEvent({
        address: marketAddress,
        abi: TradePoolAbi,
        onLogs: (logs) => {
            for (const log of logs) {
                const parsed = parseLogToMarketTradeEvent(log);
                if (eventNameSet && !eventNameSet.has(parsed.eventName)) {
                    continue;
                }
                onEvent(parsed);
            }
        },
        onError: (error) => {
            if (onError) {
                onError(error);
            }
        },
    });
    return unwatch;
}
const PRICE_AFFECTING_EVENTS = [
    'EnterOption',
    'ExecuteBuyOrder',
    'ExecuteSellOrder',
    'Sync',
];
export function subscribePriceUpdates(client, rpcUrl, params) {
    const { marketAddress, onPriceUpdate, onError } = params;
    let cachedNumOptions = null;
    let lastFetchedBlock = 0n;
    let fetchInProgress = false;
    async function fetchPrices(triggerEvent) {
        // Debounce: skip if we already fetched for this block
        if (triggerEvent.blockNumber <= lastFetchedBlock)
            return;
        // Skip if a fetch is already running
        if (fetchInProgress)
            return;
        fetchInProgress = true;
        try {
            // On first call, read numberOfOptions alongside prices
            if (cachedNumOptions === null) {
                const numResult = await multicallRead(rpcUrl, [{
                        address: marketAddress,
                        abi: TradePoolAbi,
                        functionName: 'numberOfOptions',
                        args: [],
                    }]);
                if (numResult[0]?.status !== 'success') {
                    throw new Error('Failed to read numberOfOptions from contract');
                }
                cachedNumOptions = Number(numResult[0].result);
            }
            const numOptions = cachedNumOptions;
            if (numOptions === 0) {
                lastFetchedBlock = triggerEvent.blockNumber;
                onPriceUpdate({ prices: [], triggeredBy: triggerEvent });
                return;
            }
            // Multicall getCurrentPrice for each option
            const priceContracts = Array.from({ length: numOptions }, (_, i) => ({
                address: marketAddress,
                abi: TradePoolAbi,
                functionName: 'getCurrentPrice',
                args: [BigInt(i)],
            }));
            const priceResults = await multicallRead(rpcUrl, priceContracts);
            const prices = Array.from({ length: numOptions }, (_, i) => {
                const result = priceResults[i];
                return {
                    choiceIndex: i,
                    optionName: `Option ${i}`,
                    currentPrice: result?.status === 'success' ? BigInt(result.result) : 0n,
                };
            });
            lastFetchedBlock = triggerEvent.blockNumber;
            onPriceUpdate({ prices, triggeredBy: triggerEvent });
        }
        catch (err) {
            if (onError) {
                onError(err instanceof Error ? err : new Error(String(err)));
            }
        }
        finally {
            fetchInProgress = false;
        }
    }
    return subscribeToMarketEvents(client, {
        marketAddress,
        eventNames: PRICE_AFFECTING_EVENTS,
        onEvent: (event) => {
            fetchPrices(event);
        },
        onError,
    });
}
