// @opinion-labs/opinion-clob-sdk/dist/chain/safe/multisend.js
import { concat, encodeFunctionData, encodePacked } from 'viem';
import { multiSendAbi } from '../contracts/safe.js';
/**
 * Encode a single MultiSend transaction
 */
function encodeMultiSendTx(tx) {
    const dataLength = (tx.data.length - 2) / 2; // Remove 0x prefix and count bytes
    return encodePacked(['uint8', 'address', 'uint256', 'uint256', 'bytes'], [tx.operation, tx.to, tx.value, BigInt(dataLength), tx.data]);
}
/**
 * Encode multiple transactions for MultiSend
 */
export function encodeMultiSendData(txs) {
    const encodedTxs = txs.map(encodeMultiSendTx);
    // Concatenate all encoded transactions
    let result = '0x';
    for (const encoded of encodedTxs) {
        result = concat([result, encoded]);
    }
    return result;
}
/**
 * Encode the multiSend function call
 */
export function encodeMultiSendCall(txs) {
    const data = encodeMultiSendData(txs);
    return encodeFunctionData({
        abi: multiSendAbi,
        functionName: 'multiSend',
        args: [data],
    });
}
