// @opinion-labs/opinion-clob-sdk/src/chain/order_utils/order_builder.ts
import { type Address, type Hex, hashTypedData } from 'viem';
import { Signer } from './signer.js';
import { generateSalt, normalizeAddress, prependZx } from './utils.js';
import { ValidationException } from '../../errors.js';
import { OrderSide, SignatureType } from '../../models/enums.js';
import { type Order, type OrderData, type SignedOrder } from '../../models/order.js';
import { EIP712_DOMAIN } from '../../config.js';

/**
 * EIP-712 Order type definition
 */
const ORDER_TYPES = {
  Order: [
    { name: 'salt', type: 'uint256' },
    { name: 'maker', type: 'address' },
    { name: 'signer', type: 'address' },
    { name: 'taker', type: 'address' },
    { name: 'tokenId', type: 'uint256' },
    { name: 'makerAmount', type: 'uint256' },
    { name: 'takerAmount', type: 'uint256' },
    { name: 'expiration', type: 'uint256' },
    { name: 'nonce', type: 'uint256' },
    { name: 'feeRateBps', type: 'uint256' },
    { name: 'side', type: 'uint8' },
    { name: 'signatureType', type: 'uint8' },
  ],
} as const;

/**
 * Order builder for creating and signing orders
 */
export class OrderBuilder {
  private exchangeAddress: Address;
  private chainId: number;
  private signer: Signer;
  private saltGenerator: () => bigint;

  constructor(
    exchangeAddress: string,
    chainId: number,
    signer: Signer,
    saltGenerator: () => bigint = generateSalt
  ) {
    this.exchangeAddress = normalizeAddress(exchangeAddress);
    this.chainId = chainId;
    this.signer = signer;
    this.saltGenerator = saltGenerator;
  }

  /**
   * Get EIP-712 domain
   */
  private getDomain() {
    return {
      name: EIP712_DOMAIN.name,
      version: EIP712_DOMAIN.version,
      chainId: BigInt(this.chainId),
      verifyingContract: this.exchangeAddress,
    };
  }

  /**
   * Build an order from order data
   */
  buildOrder(data: OrderData): Order {
    if (!this.validateInputs(data)) {
      throw new ValidationException('Invalid order inputs');
    }

    const signerAddr = data.signer || data.maker;

    if (signerAddr.toLowerCase() !== this.signer.address().toLowerCase()) {
      throw new ValidationException('Signer does not match');
    }

    const expiration = data.expiration || '0';
    const signatureType = data.signatureType ?? SignatureType.EOA;

    return {
      salt: this.saltGenerator(),
      maker: normalizeAddress(data.maker),
      signer: normalizeAddress(signerAddr),
      taker: normalizeAddress(data.taker),
      tokenId: BigInt(data.tokenId),
      makerAmount: data.makerAmount,
      takerAmount: data.takerAmount,
      expiration: BigInt(expiration),
      nonce: BigInt(data.nonce),
      feeRateBps: BigInt(data.feeRateBps),
      side: data.side,
      signatureType: signatureType,
    };
  }

  /**
   * Create struct hash for signing
   */
  private createStructHash(order: Order): Hex {
    return hashTypedData({
      domain: this.getDomain(),
      types: ORDER_TYPES,
      primaryType: 'Order',
      message: {
        salt: order.salt,
        maker: order.maker,
        signer: order.signer,
        taker: order.taker,
        tokenId: order.tokenId,
        makerAmount: order.makerAmount,
        takerAmount: order.takerAmount,
        expiration: order.expiration,
        nonce: order.nonce,
        feeRateBps: order.feeRateBps,
        side: order.side,
        signatureType: order.signatureType,
      },
    });
  }

  /**
   * Build order signature
   */
  async buildOrderSignature(order: Order): Promise<Hex> {
    const structHash = this.createStructHash(order);
    const signature = await this.signer.sign(structHash);
    return prependZx(signature.slice(2));
  }

  /**
   * Build and sign an order
   */
  async buildSignedOrder(data: OrderData): Promise<SignedOrder> {
    const order = this.buildOrder(data);
    const signature = await this.buildOrderSignature(order);
    return {
      order,
      signature,
    };
  }

  /**
   * Validate order inputs
   */
  private validateInputs(data: OrderData): boolean {
    if (!data.maker) return false;
    if (!data.tokenId) return false;
    if (data.makerAmount === undefined || data.makerAmount === null) return false;
    if (data.takerAmount === undefined || data.takerAmount === null) return false;
    if (data.side === undefined || data.side === null) return false;
    if (data.side !== OrderSide.BUY && data.side !== OrderSide.SELL) return false;

    const feeRateBps = parseInt(data.feeRateBps, 10);
    if (isNaN(feeRateBps) || feeRateBps < 0) return false;

    const nonce = parseInt(data.nonce, 10);
    if (isNaN(nonce) || nonce < 0) return false;

    const expiration = parseInt(data.expiration || '0', 10);
    if (isNaN(expiration) || expiration < 0) return false;

    if (
      data.signatureType !== undefined &&
      data.signatureType !== SignatureType.EOA &&
      data.signatureType !== SignatureType.POLY_GNOSIS_SAFE &&
      data.signatureType !== SignatureType.POLY_PROXY
    ) {
      return false;
    }

    return true;
  }
}
