// @opinion-labs/opinion-clob-sdk/src/chain/order_utils/signer.ts
import { type Address, type Hex, type PrivateKeyAccount, serializeSignature } from 'viem';
import { privateKeyToAccount, sign } from 'viem/accounts';

/**
 * Signer class for signing orders and messages
 */
export class Signer {
  private account: PrivateKeyAccount;
  private privateKey: Hex;

  constructor(privateKey: Hex) {
    this.privateKey = privateKey;
    this.account = privateKeyToAccount(privateKey);
  }

  /**
   * Get the signer's address
   */
  address(): Address {
    return this.account.address;
  }

  /**
   * Sign a hash (message digest) directly without Ethereum message prefix
   */
  async sign(hash: Hex): Promise<Hex> {
    // Sign the raw hash directly (no message prefix)
    const signResult = await sign({
      hash,
      privateKey: this.privateKey,
    });
    // Serialize the signature components to a single hex string
    return serializeSignature(signResult);
  }

  /**
   * Sign typed data (EIP-712)
   */
  async signTypedData(typedData: {
    domain: {
      name: string;
      version: string;
      chainId: bigint;
      verifyingContract: Address;
    };
    types: Record<string, Array<{ name: string; type: string }>>;
    primaryType: string;
    message: Record<string, unknown>;
  }): Promise<Hex> {
    const signature = await this.account.signTypedData({
      domain: typedData.domain,
      types: typedData.types,
      primaryType: typedData.primaryType,
      message: typedData.message,
    });
    return signature;
  }

  /**
   * Get the underlying account
   */
  getAccount(): PrivateKeyAccount {
    return this.account;
  }
}

/**
 * Create a random signer for testing
 */
export function createRandomSigner(): Signer {
  const randomBytes = new Uint8Array(32);
  crypto.getRandomValues(randomBytes);
  const hexString = Array.from(randomBytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
  const privateKey: Hex = `0x${hexString}`;
  return new Signer(privateKey);
}
