// @opinion-labs/opinion-clob-sdk/src/chain/order_utils/utils.ts
import { type Address, getAddress, keccak256, toHex } from 'viem';
import { OrderSide } from '../../models/enums.js';
import { InvalidParamError } from '../../errors.js';

/**
 * Generate a random salt for order uniqueness
 */
export function generateSalt(): bigint {
  const randomBytes = new Uint8Array(32);
  crypto.getRandomValues(randomBytes);
  return BigInt(keccak256(toHex(randomBytes)));
}

/**
 * Normalize address to checksum format
 */
export function normalizeAddress(address: string): Address {
  return getAddress(address);
}

/**
 * Prepend 0x to hex string if not present
 */
export function prependZx(value: string): `0x${string}` {
  if (value.startsWith('0x')) {
    return value as `0x${string}`;
  }
  return `0x${value}`;
}

/**
 * Parse price string to numerator and denominator
 * Handles prices with up to 6 decimal places (e.g., "0.123456")
 */
export function parsePrice(priceStr: string): { numerator: bigint; denominator: bigint } {
  const price = parseFloat(priceStr);

  if (price <= 0 || price >= 1) {
    throw new InvalidParamError(`Price must be between 0 and 1 (exclusive), got: ${priceStr}`);
  }

  // Find decimal places (max 6)
  const parts = priceStr.split('.');
  const decimalPlaces = parts[1] ? Math.min(parts[1].length, 6) : 0;

  const multiplier = BigInt(10 ** decimalPlaces);
  const numerator = BigInt(Math.round(price * Number(multiplier)));

  return {
    numerator,
    denominator: multiplier,
  };
}

/**
 * Calculate order amounts based on price and maker amount
 * Ensures exact price matching using fraction arithmetic
 */
export function calculateOrderAmounts(
  price: number,
  makerAmount: bigint,
  side: OrderSide,
  _decimals?: number
): { recalculatedMakerAmount: bigint; takerAmount: bigint } {
  if (price <= 0 || price >= 1) {
    throw new InvalidParamError(`Price must be between 0 and 1 (exclusive), got: ${price}`);
  }

  // Convert price to fraction with 6 decimal precision
  const priceStr = price.toFixed(6);
  const { numerator: priceNum, denominator: priceDenom } = parsePrice(priceStr);

  if (side === OrderSide.BUY) {
    // For BUY: price = makerAmount / takerAmount
    // Find k such that: recalculated_maker = k * price_num, taker = k * price_denom
    // k = makerAmount / price_num (rounded down for safety)
    const k = makerAmount / priceNum;

    if (k === 0n) {
      throw new InvalidParamError('Maker amount too small for given price');
    }

    const recalculatedMakerAmount = k * priceNum;
    const takerAmount = k * priceDenom;

    return { recalculatedMakerAmount, takerAmount };
  } else {
    // For SELL: price = takerAmount / makerAmount
    // Find k such that: taker = k * price_num, recalculated_maker = k * price_denom
    const k = makerAmount / priceDenom;

    if (k === 0n) {
      throw new InvalidParamError('Maker amount too small for given price');
    }

    const recalculatedMakerAmount = k * priceDenom;
    const takerAmount = k * priceNum;

    return { recalculatedMakerAmount, takerAmount };
  }
}

/**
 * Safely convert human-readable amount to wei units
 */
export function safeAmountToWei(amount: number, decimals: number): bigint {
  if (amount <= 0) {
    throw new InvalidParamError(`Amount must be positive, got: ${amount}`);
  }

  if (decimals < 0 || decimals > 18) {
    throw new InvalidParamError(`Decimals must be between 0 and 18, got: ${decimals}`);
  }

  // Use string manipulation to avoid floating point precision loss
  const [intPart, fracPart = ''] = amount.toString().split('.');
  const paddedFrac = fracPart.padEnd(decimals, '0').slice(0, decimals);
  const weiStr = intPart + paddedFrac;

  const result = BigInt(weiStr);

  if (result <= 0n) {
    throw new InvalidParamError(`Calculated amount is zero or negative`);
  }

  return result;
}
