// @opinion-labs/opinion-clob-sdk/src/chain/safe/multisend.ts
import { type Address, type Hex, concat, encodeFunctionData, encodePacked } from 'viem';
import { MultiSendOperation } from './constants.js';
import { multiSendAbi } from '../contracts/safe.js';

/**
 * MultiSend transaction structure
 */
export interface MultiSendTx {
  operation: MultiSendOperation;
  to: Address;
  value: bigint;
  data: Hex;
}

/**
 * Encode a single MultiSend transaction
 */
function encodeMultiSendTx(tx: MultiSendTx): Hex {
  const dataLength = (tx.data.length - 2) / 2; // Remove 0x prefix and count bytes

  return encodePacked(
    ['uint8', 'address', 'uint256', 'uint256', 'bytes'],
    [tx.operation, tx.to, tx.value, BigInt(dataLength), tx.data]
  );
}

/**
 * Encode multiple transactions for MultiSend
 */
export function encodeMultiSendData(txs: MultiSendTx[]): Hex {
  const encodedTxs = txs.map(encodeMultiSendTx);

  // Concatenate all encoded transactions
  let result: Hex = '0x';
  for (const encoded of encodedTxs) {
    result = concat([result, encoded]);
  }

  return result;
}

/**
 * Encode the multiSend function call
 */
export function encodeMultiSendCall(txs: MultiSendTx[]): Hex {
  const data = encodeMultiSendData(txs);

  return encodeFunctionData({
    abi: multiSendAbi,
    functionName: 'multiSend',
    args: [data],
  });
}
