// @opinion-labs/opinion-clob-sdk/src/chain/safe/safe.ts
import {
  type Address,
  type Hex,
  type Hash,
  type PublicClient,
  type WalletClient,
  createPublicClient,
  createWalletClient,
  http,
  getContract,
  hashTypedData,
} from 'viem';
import { bsc } from 'viem/chains';
import { privateKeyToAccount } from 'viem/accounts';
import { ProxyAgent, type Dispatcher } from 'undici';
import { safeAbi } from '../contracts/safe.js';
import { SafeOperation, MultiSendOperation, ZERO_ADDRESS } from './constants.js';
import { type MultiSendTx, encodeMultiSendCall } from './multisend.js';
import { Signer } from '../order_utils/signer.js';

/**
 * Safe transaction parameters
 */
export interface SafeTxParams {
  to: Address;
  value: bigint;
  data: Hex;
  operation: SafeOperation;
  safeTxGas: bigint;
  baseGas: bigint;
  gasPrice: bigint;
  gasToken: Address;
  refundReceiver: Address;
  nonce: bigint;
}

/**
 * Safe EIP-712 types
 */
const SAFE_TX_TYPES = {
  SafeTx: [
    { name: 'to', type: 'address' },
    { name: 'value', type: 'uint256' },
    { name: 'data', type: 'bytes' },
    { name: 'operation', type: 'uint8' },
    { name: 'safeTxGas', type: 'uint256' },
    { name: 'baseGas', type: 'uint256' },
    { name: 'gasPrice', type: 'uint256' },
    { name: 'gasToken', type: 'address' },
    { name: 'refundReceiver', type: 'address' },
    { name: 'nonce', type: 'uint256' },
  ],
} as const;

/**
 * Safe wrapper class for multi-sig operations
 */
export class Safe {
  private publicClient: PublicClient;
  private walletClient: WalletClient;
  private signer: Signer;
  private safeAddress: Address;
  private multiSendAddress: Address;
  private chainId: number;

  constructor(
    rpcUrl: string,
    privateKey: Hex,
    safeAddress: Address,
    multiSendAddress: Address,
    chainId: number = 56,
    proxyUrl?: string
  ) {
    this.chainId = chainId;
    const chain = chainId === 56 ? bsc : bsc; // Add more chains as needed

    // Create custom fetch with proxy support if proxyUrl is provided
    let fetchOptions: Parameters<typeof http>[1] | undefined;
    if (proxyUrl) {
      const proxyDispatcher: Dispatcher = new ProxyAgent(proxyUrl);
      fetchOptions = {
        fetchOptions: {
          dispatcher: proxyDispatcher,
        },
      };
    }

    this.publicClient = createPublicClient({
      chain,
      transport: http(rpcUrl, fetchOptions),
    });

    const account = privateKeyToAccount(privateKey);
    this.walletClient = createWalletClient({
      account,
      chain,
      transport: http(rpcUrl, fetchOptions),
    });

    this.signer = new Signer(privateKey);
    this.safeAddress = safeAddress;
    this.multiSendAddress = multiSendAddress;
  }

  /**
   * Get the Safe contract instance
   */
  private getSafeContract() {
    return getContract({
      address: this.safeAddress,
      abi: safeAbi,
      client: this.publicClient,
    });
  }

  /**
   * Get the current Safe nonce
   */
  async getNonce(): Promise<bigint> {
    const contract = this.getSafeContract();
    return await contract.read.nonce();
  }

  /**
   * Get EIP-712 domain for Safe
   */
  private getDomain() {
    return {
      chainId: BigInt(this.chainId),
      verifyingContract: this.safeAddress,
    };
  }

  /**
   * Build Safe transaction parameters
   */
  async buildSafeTxParams(
    to: Address,
    value: bigint,
    data: Hex,
    operation: SafeOperation = SafeOperation.CALL,
    nonce?: bigint
  ): Promise<SafeTxParams> {
    const safeNonce = nonce ?? (await this.getNonce());

    return {
      to,
      value,
      data,
      operation,
      safeTxGas: 0n,
      baseGas: 0n,
      gasPrice: 0n,
      gasToken: ZERO_ADDRESS,
      refundReceiver: ZERO_ADDRESS,
      nonce: safeNonce,
    };
  }

  /**
   * Hash Safe transaction for signing
   */
  hashSafeTx(params: SafeTxParams): Hex {
    return hashTypedData({
      domain: this.getDomain(),
      types: SAFE_TX_TYPES,
      primaryType: 'SafeTx',
      message: {
        to: params.to,
        value: params.value,
        data: params.data,
        operation: params.operation,
        safeTxGas: params.safeTxGas,
        baseGas: params.baseGas,
        gasPrice: params.gasPrice,
        gasToken: params.gasToken,
        refundReceiver: params.refundReceiver,
        nonce: params.nonce,
      },
    });
  }

  /**
   * Sign Safe transaction
   */
  async signSafeTx(params: SafeTxParams): Promise<Hex> {
    const txHash = this.hashSafeTx(params);
    return await this.signer.sign(txHash);
  }

  /**
   * Build signature in Safe format (r, s, v concatenated)
   */
  buildSignature(signature: Hex): Hex {
    // For a single signer, the signature is already in the correct format
    // Safe expects signatures sorted by signer address
    return signature;
  }

  /**
   * Execute a Safe transaction
   */
  async executeTransaction(
    params: SafeTxParams,
    signature: Hex
  ): Promise<{ txHash: Hash; safeTxHash: Hex }> {
    const safeTxHash = this.hashSafeTx(params);

    const { request } = await this.publicClient.simulateContract({
      address: this.safeAddress,
      abi: safeAbi,
      functionName: 'execTransaction',
      args: [
        params.to,
        params.value,
        params.data,
        params.operation,
        params.safeTxGas,
        params.baseGas,
        params.gasPrice,
        params.gasToken,
        params.refundReceiver,
        signature,
      ],
      account: this.walletClient.account,
    });

    const txHash = await this.walletClient.writeContract(request);

    return { txHash, safeTxHash };
  }

  /**
   * Estimate gas for MultiSend inner transactions.
   * Simulates each CALL transaction from the Safe address to get gas estimates,
   * then applies a 1.5x safety margin. This matches the Python SDK's behavior
   * and helps surface real revert reasons during gas estimation rather than
   * getting an opaque GS013 error at execution time.
   */
  async estimateMultiSendGas(txs: MultiSendTx[]): Promise<bigint> {
    let totalGas = 0n;
    for (const tx of txs) {
      if (tx.operation === MultiSendOperation.DELEGATE_CALL) {
        continue;
      }
      const gas = await this.publicClient.estimateGas({
        account: this.safeAddress,
        to: tx.to,
        value: tx.value,
        data: tx.data,
      });
      totalGas += gas;
    }
    return (totalGas * 3n) / 2n;
  }

  /**
   * Build MultiSend transaction
   */
  async buildMultiSendTx(txs: MultiSendTx[], nonce?: bigint): Promise<SafeTxParams> {
    const multiSendData = encodeMultiSendCall(txs);
    const safeTxGas = await this.estimateMultiSendGas(txs);
    const safeNonce = nonce ?? (await this.getNonce());

    return {
      to: this.multiSendAddress,
      value: 0n,
      data: multiSendData,
      operation: SafeOperation.DELEGATE_CALL,
      safeTxGas,
      baseGas: 0n,
      gasPrice: 0n,
      gasToken: ZERO_ADDRESS,
      refundReceiver: ZERO_ADDRESS,
      nonce: safeNonce,
    };
  }

  /**
   * Execute multiple transactions via MultiSend
   */
  async executeMultiSend(
    txs: MultiSendTx[]
  ): Promise<{ txHash: Hash; safeTxHash: Hex; returnValue: Hex | null }> {
    const params = await this.buildMultiSendTx(txs);
    const signature = await this.signSafeTx(params);
    const { txHash, safeTxHash } = await this.executeTransaction(params, signature);

    return { txHash, safeTxHash, returnValue: null };
  }

  /**
   * Get signer address
   */
  getSignerAddress(): Address {
    return this.signer.address();
  }

  /**
   * Get Safe address
   */
  getSafeAddress(): Address {
    return this.safeAddress;
  }

  /**
   * Get public client for contract reads
   */
  getPublicClient(): PublicClient {
    return this.publicClient;
  }
}
