// @opinion-labs/opinion-api/src/sdk.gen.ts
// This file is auto-generated by @hey-api/openapi-ts

import type { Client, Options as Options2, TDataShape } from './client';
import { client } from './client.gen';
import type { GetMarketByMarketIdData, GetMarketByMarketIdErrors, GetMarketByMarketIdResponses, GetMarketCategoricalByMarketIdData, GetMarketCategoricalByMarketIdErrors, GetMarketCategoricalByMarketIdResponses, GetMarketData, GetMarketErrors, GetMarketResponses, GetMarketSlugBySlugData, GetMarketSlugBySlugErrors, GetMarketSlugBySlugResponses, GetOrderByOrderIdData, GetOrderByOrderIdErrors, GetOrderByOrderIdResponses, GetOrderData, GetOrderErrors, GetOrderResponses, GetPositionsData, GetPositionsErrors, GetPositionsResponses, GetPositionsUserByWalletAddressData, GetPositionsUserByWalletAddressErrors, GetPositionsUserByWalletAddressResponses, GetQuoteTokenData, GetQuoteTokenErrors, GetQuoteTokenResponses, GetTokenFeeRatesData, GetTokenFeeRatesErrors, GetTokenFeeRatesResponses, GetTokenLatestPriceData, GetTokenLatestPriceErrors, GetTokenLatestPriceResponses, GetTokenOrderbookData, GetTokenOrderbookErrors, GetTokenOrderbookResponses, GetTokenPriceHistoryData, GetTokenPriceHistoryErrors, GetTokenPriceHistoryResponses, GetTopicByTopicIdData, GetTopicByTopicIdErrors, GetTopicByTopicIdResponses, GetTopicData, GetTopicErrors, GetTopicMultiByTopicIdData, GetTopicMultiByTopicIdErrors, GetTopicMultiByTopicIdResponses, GetTopicResponses, GetTradeData, GetTradeErrors, GetTradeResponses, GetTradeUserByWalletAddressData, GetTradeUserByWalletAddressErrors, GetTradeUserByWalletAddressResponses, GetUserAuthData, GetUserAuthErrors, GetUserAuthResponses, GetUserBalanceData, GetUserBalanceErrors, GetUserBalanceResponses, PostOrderCancelData, PostOrderCancelErrors, PostOrderCancelResponses, PostOrderData, PostOrderErrors, PostOrderResponses } from './types.gen';

export type Options<TData extends TDataShape = TDataShape, ThrowOnError extends boolean = boolean> = Options2<TData, ThrowOnError> & {
    /**
     * You can provide a client instance returned by `createClient()` instead of
     * individual options. This might be also useful if you want to implement a
     * custom client.
     */
    client?: Client;
    /**
     * You can pass arbitrary values through the `meta` object. This can be
     * used to access values that aren't defined as part of the SDK function.
     */
    meta?: Record<string, unknown>;
};

/**
 * Get Market List
 *
 * Get Market List(All Markets)
 */
export const getMarket = <ThrowOnError extends boolean = false>(options: Options<GetMarketData, ThrowOnError>) => {
    return (options.client ?? client).get<GetMarketResponses, GetMarketErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/market',
        ...options
    });
};

/**
 * Get Detail of a Categorical Market
 *
 * Get Detail of a Categorical Market
 */
export const getMarketCategoricalByMarketId = <ThrowOnError extends boolean = false>(options: Options<GetMarketCategoricalByMarketIdData, ThrowOnError>) => {
    return (options.client ?? client).get<GetMarketCategoricalByMarketIdResponses, GetMarketCategoricalByMarketIdErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/market/categorical/{marketId}',
        ...options
    });
};

/**
 * Get Detail of a Binary Market
 *
 * Get Detail of a Binary Market
 */
export const getMarketByMarketId = <ThrowOnError extends boolean = false>(options: Options<GetMarketByMarketIdData, ThrowOnError>) => {
    return (options.client ?? client).get<GetMarketByMarketIdResponses, GetMarketByMarketIdErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/market/{marketId}',
        ...options
    });
};

/**
 * Get Market Detail by Slug
 *
 * Get market detail by slug, supports both binary and categorical markets
 */
export const getMarketSlugBySlug = <ThrowOnError extends boolean = false>(options: Options<GetMarketSlugBySlugData, ThrowOnError>) => {
    return (options.client ?? client).get<GetMarketSlugBySlugResponses, GetMarketSlugBySlugErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/market/slug/{slug}',
        ...options
    });
};

/**
 * Get orders
 *
 * Get user's orders with simplified query parameters
 */
export const getOrder = <ThrowOnError extends boolean = false>(options: Options<GetOrderData, ThrowOnError>) => {
    return (options.client ?? client).get<GetOrderResponses, GetOrderErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/order',
        ...options
    });
};

/**
 * Create an order
 *
 * Create an Order
 */
export const postOrder = <ThrowOnError extends boolean = false>(options: Options<PostOrderData, ThrowOnError>) => {
    return (options.client ?? client).post<PostOrderResponses, PostOrderErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/order',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};

/**
 * Cancel an order
 *
 * Cancel an Order
 */
export const postOrderCancel = <ThrowOnError extends boolean = false>(options: Options<PostOrderCancelData, ThrowOnError>) => {
    return (options.client ?? client).post<PostOrderCancelResponses, PostOrderCancelErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/order/cancel',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};

/**
 * Get order detail
 *
 * Get order detail with related trades
 */
export const getOrderByOrderId = <ThrowOnError extends boolean = false>(options: Options<GetOrderByOrderIdData, ThrowOnError>) => {
    return (options.client ?? client).get<GetOrderByOrderIdResponses, GetOrderByOrderIdErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/order/{orderId}',
        ...options
    });
};

/**
 * Get user positions
 *
 * Get user's current positions with simplified query parameters
 */
export const getPositions = <ThrowOnError extends boolean = false>(options: Options<GetPositionsData, ThrowOnError>) => {
    return (options.client ?? client).get<GetPositionsResponses, GetPositionsErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/positions',
        ...options
    });
};

/**
 * Get another user's positions
 *
 * Get positions of a specific user by wallet address
 */
export const getPositionsUserByWalletAddress = <ThrowOnError extends boolean = false>(options: Options<GetPositionsUserByWalletAddressData, ThrowOnError>) => {
    return (options.client ?? client).get<GetPositionsUserByWalletAddressResponses, GetPositionsUserByWalletAddressErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/positions/user/{walletAddress}',
        ...options
    });
};

/**
 * Get quote token list
 *
 * Retrieve a list of quote tokens (trading pair quote currencies)
 */
export const getQuoteToken = <ThrowOnError extends boolean = false>(options: Options<GetQuoteTokenData, ThrowOnError>) => {
    return (options.client ?? client).get<GetQuoteTokenResponses, GetQuoteTokenErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/quoteToken',
        ...options
    });
};

/**
 * Get fee rates for a token
 *
 * Get the taker and maker fee rates for a specific token
 */
export const getTokenFeeRates = <ThrowOnError extends boolean = false>(options: Options<GetTokenFeeRatesData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTokenFeeRatesResponses, GetTokenFeeRatesErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/token/fee-rates',
        ...options
    });
};

/**
 * Get latest price of a token
 *
 * Get the latest trade price and details for a specific token
 */
export const getTokenLatestPrice = <ThrowOnError extends boolean = false>(options: Options<GetTokenLatestPriceData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTokenLatestPriceResponses, GetTokenLatestPriceErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/token/latest-price',
        ...options
    });
};

/**
 * Get orderbook of a token
 *
 * Get orderbook (market depth) of a specific token
 */
export const getTokenOrderbook = <ThrowOnError extends boolean = false>(options: Options<GetTokenOrderbookData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTokenOrderbookResponses, GetTokenOrderbookErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/token/orderbook',
        ...options
    });
};

/**
 * Get price history of a token
 *
 * Get historical price data for a specific token (Polymarket-compatible format)
 */
export const getTokenPriceHistory = <ThrowOnError extends boolean = false>(options: Options<GetTokenPriceHistoryData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTokenPriceHistoryResponses, GetTokenPriceHistoryErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/token/price-history',
        ...options
    });
};

/**
 * Get Topic List
 *
 * Get Topic List(All Topics)
 */
export const getTopic = <ThrowOnError extends boolean = false>(options: Options<GetTopicData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTopicResponses, GetTopicErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/topic',
        ...options
    });
};

/**
 * Get Detail of a Categorical Topic
 *
 * Get Detail of a Categorical Topic
 */
export const getTopicMultiByTopicId = <ThrowOnError extends boolean = false>(options: Options<GetTopicMultiByTopicIdData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTopicMultiByTopicIdResponses, GetTopicMultiByTopicIdErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/topic/multi/{topicId}',
        ...options
    });
};

/**
 * Get Detail of a Binary Topic
 *
 * Get Detail of a Binary Topic
 */
export const getTopicByTopicId = <ThrowOnError extends boolean = false>(options: Options<GetTopicByTopicIdData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTopicByTopicIdResponses, GetTopicByTopicIdErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/topic/{topicId}',
        ...options
    });
};

/**
 * Get user's trades
 *
 * Get user's trades with simplified query parameters
 */
export const getTrade = <ThrowOnError extends boolean = false>(options: Options<GetTradeData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTradeResponses, GetTradeErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/trade',
        ...options
    });
};

/**
 * Get another user's trades
 *
 * Get trades of a specific user by wallet address. Only returns filled (successful) trades.
 */
export const getTradeUserByWalletAddress = <ThrowOnError extends boolean = false>(options: Options<GetTradeUserByWalletAddressData, ThrowOnError>) => {
    return (options.client ?? client).get<GetTradeUserByWalletAddressResponses, GetTradeUserByWalletAddressErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/trade/user/{walletAddress}',
        ...options
    });
};

/**
 * Auth
 *
 * Authentication of apikey and returns user info
 */
export const getUserAuth = <ThrowOnError extends boolean = false>(options?: Options<GetUserAuthData, ThrowOnError>) => {
    return (options?.client ?? client).get<GetUserAuthResponses, GetUserAuthErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/user/auth',
        ...options
    });
};

/**
 * Get user's quote token balances
 *
 * Get user's quote token (USDC, etc.) balances including available and frozen amounts
 */
export const getUserBalance = <ThrowOnError extends boolean = false>(options: Options<GetUserBalanceData, ThrowOnError>) => {
    return (options.client ?? client).get<GetUserBalanceResponses, GetUserBalanceErrors, ThrowOnError>({
        security: [
            {
                name: 'apikey',
                type: 'apiKey'
            }
        ],
        url: '/user/balance',
        ...options
    });
};
