// @opinion-labs/opinion-api/src/types.gen.ts
// This file is auto-generated by @hey-api/openapi-ts

export type ClientOptions = {
    baseUrl: 'https://openapi.opinion.trade/openapi' | (string & {});
};

export type ModelsCollateralCurrency = {
    chainId?: string;
    contract?: string;
};

export type ModelsKlineData = {
    /**
     * 成交量
     */
    amount?: string;
    /**
     * 收盘价,当K线为最晚的一根时，是最新成交价
     */
    close?: string;
    /**
     * 成交笔数
     */
    count?: number;
    /**
     * 最高价
     */
    high?: string;
    /**
     * k线id
     */
    id?: number;
    /**
     * 最低价
     */
    low?: string;
    /**
     * 开盘价
     */
    open?: string;
    /**
     * 成交额, 即 sum(每一笔成交价 * 该笔的成交量)
     */
    vol?: string;
};

export type ModelsTokenBalance = {
    /**
     * decimals with 6 precision
     */
    amount?: string;
    chainId?: string;
    contract?: string;
    /**
     * decimals with 6 precision
     */
    frozenAmount?: string;
    id?: number;
    /**
     * uint256/BigNumber
     */
    tokenId?: string;
    updateTimestamp?: number;
    walletAddress?: string;
    walletUser?: string;
};

export type ModelsTopicCreationStatus = 1 | 2 | 3 | 4;

export type ModelsTopicStatus = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export type ModelsTradeMetrics = {
    avgVolume?: string;
    avgVolumeShares?: string;
    selfTrades?: string;
    totalFees?: string;
    trades?: string;
    volume?: string;
    volumeShares?: string;
};

export type OpenapiAddOrderRespOpenApi = {
    orderData?: OpenapiOrderDataOpenApi;
};

export type OpenapiBalanceRespOpenApi = {
    /**
     * List of quote token balances
     */
    balances?: Array<OpenapiQuoteTokenBalance>;
    /**
     * Chain ID
     */
    chainId?: string;
    /**
     * Multi-sig wallet address (empty if not exists)
     */
    multiSignAddress?: string;
    /**
     * User's wallet address
     */
    walletAddress?: string;
};

export type OpenapiCancelOrderRequestOpenApi = {
    orderId?: string;
};

export type OpenapiCancelOrderRespOpenApi = {
    result?: boolean;
};

export type OpenapiChildMarketDataOpenApi = {
    chainId?: string;
    /**
     * Resolution info
     */
    conditionId?: string;
    /**
     * Time fields (moved to bottom)
     */
    createdAt?: number;
    cutoffAt?: number;
    /**
     * Incentive factor masked as empty object when present
     */
    incentiveFactor?: OpenapiIncentiveFactorOpenApi;
    /**
     * Basic market info
     */
    marketId?: number;
    marketTitle?: string;
    noLabel?: string;
    noTokenId?: string;
    /**
     * Additional data
     */
    questionId?: string;
    /**
     * Chain info
     */
    quoteToken?: string;
    resolvedAt?: number;
    resultTokenId?: string;
    rules?: string;
    /**
     * URL-friendly slug identifier for the market
     */
    slug?: string;
    status?: ModelsTopicStatus;
    /**
     * Human-readable status
     */
    statusEnum?: string;
    /**
     * Market metrics
     */
    volume?: string;
    /**
     * Labels and rules
     */
    yesLabel?: string;
    /**
     * Token info
     */
    yesTokenId?: string;
};

export type OpenapiFeeRatesRespOpenApi = {
    /**
     * Maker fee rate in basis points
     */
    makerFeeBps?: string;
    /**
     * Taker fee rate in basis points
     */
    takerFeeBps?: string;
    /**
     * Token ID
     */
    tokenId?: string;
};

export type OpenapiIncentiveFactorOpenApi = {
    [key: string]: unknown;
};

export type OpenapiLatestPriceRespOpenApi = {
    /**
     * Latest trade price
     */
    price?: string;
    /**
     * Last trade side: BUY or SELL
     */
    side?: string;
    /**
     * Last trade size
     */
    size?: string;
    /**
     * Time fields (moved to bottom)
     */
    timestamp?: number;
    /**
     * Token ID
     */
    tokenId?: string;
};

export type OpenapiMarketDataOpenApi = {
    chainId?: string;
    /**
     * Child markets
     */
    childMarkets?: Array<OpenapiChildMarketDataOpenApi>;
    /**
     * Resolution info
     */
    conditionId?: string;
    /**
     * Time fields (moved to bottom)
     */
    createdAt?: number;
    cutoffAt?: number;
    /**
     * Incentive factor masked as empty object when present
     */
    incentiveFactor?: OpenapiIncentiveFactorOpenApi;
    /**
     * Basic market info
     */
    marketId?: number;
    marketTitle?: string;
    /**
     * 1: Categorical Market, 0: Binary Market
     */
    marketType?: number;
    noLabel?: string;
    noTokenId?: string;
    /**
     * Additional data
     */
    questionId?: string;
    /**
     * Chain info
     */
    quoteToken?: string;
    resolvedAt?: number;
    resultTokenId?: string;
    rules?: string;
    /**
     * URL-friendly slug identifier for the market
     */
    slug?: string;
    status?: ModelsTopicStatus;
    /**
     * Human-readable status
     */
    statusEnum?: string;
    /**
     * Market metrics
     */
    volume?: string;
    volume24h?: string;
    volume7d?: string;
    /**
     * Labels and rules
     */
    yesLabel?: string;
    /**
     * Token info
     */
    yesTokenId?: string;
};

export type OpenapiMarketDetailRespOpenApi = {
    data?: OpenapiMarketDataOpenApi;
};

export type OpenapiMarketListRespOpenApi = {
    list?: Array<OpenapiMarketDataOpenApi>;
    total?: number;
};

export type OpenapiOrderDataOpenApi = {
    /**
     * Time fields (moved to bottom)
     */
    createdAt?: number;
    expiresAt?: number;
    /**
     * Filled amount (quote token)
     */
    filledAmount?: string;
    /**
     * Execution info (filled status)
     */
    filledShares?: string;
    /**
     * Market info
     */
    marketId?: number;
    marketTitle?: string;
    /**
     * Total amount in order (quote token)
     */
    orderAmount?: string;
    /**
     * Order identification
     */
    orderId?: string;
    /**
     * Order size (original order submission)
     * For limit orders: orderShares is specified, orderAmount = orderShares * price
     * For market orders: orderAmount is specified (buy) or orderShares is specified (sell)
     */
    orderShares?: string;
    outcome?: string;
    /**
     * 1-yes, 2-no
     */
    outcomeSide?: number;
    /**
     * Human-readable: Yes, No
     */
    outcomeSideEnum?: string;
    /**
     * Price per share (for limit orders)
     */
    price?: string;
    profit?: string;
    /**
     * Token info
     */
    quoteToken?: string;
    rootMarketId?: number;
    rootMarketTitle?: string;
    /**
     * Order details
     */
    side?: number;
    /**
     * Human-readable: Buy, Sell
     */
    sideEnum?: string;
    /**
     * 1-pending, 2-finished, 3-canceled, 4-expired, 5-failed
     */
    status?: number;
    /**
     * Human-readable: Pending, Finished, Canceled, Expired, Failed
     */
    statusEnum?: string;
    /**
     * Related trades (only populated in order detail endpoint)
     */
    trades?: Array<OpenapiTradeDataOpenApi>;
    /**
     * 1-market, 2-limit
     */
    tradingMethod?: number;
    /**
     * Human-readable: Market, Limit
     */
    tradingMethodEnum?: string;
    /**
     * Deprecated: use orderId instead (kept for backward compatibility)
     */
    transNo?: string;
};

export type OpenapiOrderDetailRespOpenApi = {
    orderData?: OpenapiOrderDataOpenApi;
};

export type OpenapiOrderListRespOpenApi = {
    list?: Array<OpenapiOrderDataOpenApi>;
    total?: number;
};

export type OpenapiOrderbookLevel = {
    /**
     * Price level
     */
    price?: string;
    /**
     * Total size at this price level
     */
    size?: string;
};

export type OpenapiOrderbookRespOpenApi = {
    /**
     * Sell orders, sorted by price ascending
     */
    asks?: Array<OpenapiOrderbookLevel>;
    /**
     * Buy orders, sorted by price descending
     */
    bids?: Array<OpenapiOrderbookLevel>;
    /**
     * Condition ID
     */
    market?: string;
    /**
     * Unix timestamp in milliseconds
     */
    timestamp?: number;
    /**
     * Asset/Token ID
     */
    tokenId?: string;
};

export type OpenapiPositionDataOpenApi = {
    avgEntryPrice?: string;
    claimStatus?: number;
    /**
     * Human-readable: CanNotClaim, WaitClaim, Claiming, ClaimFailed, Claimed
     */
    claimStatusEnum?: string;
    /**
     * Other fields
     */
    conditionId?: string;
    /**
     * Current value in quote token (shares × current price)
     */
    currentValueInQuoteToken?: string;
    dailyPnlChange?: string;
    dailyPnlChangePercent?: string;
    marketCutoffAt?: number;
    /**
     * Market info
     */
    marketId?: number;
    marketStatus?: number;
    /**
     * Human-readable market status
     */
    marketStatusEnum?: string;
    marketTitle?: string;
    /**
     * Outcome info
     */
    outcome?: string;
    /**
     * 1-yes, 2-no
     */
    outcomeSide?: number;
    /**
     * Human-readable: Yes, No
     */
    outcomeSideEnum?: string;
    quoteToken?: string;
    /**
     * Root market info
     */
    rootMarketId?: number;
    rootMarketTitle?: string;
    sharesFrozen?: string;
    /**
     * Shares info
     */
    sharesOwned?: string;
    tokenId?: string;
    /**
     * PnL info
     */
    unrealizedPnl?: string;
    unrealizedPnlPercent?: string;
};

export type OpenapiPositionsRespOpenApi = {
    list?: Array<OpenapiPositionDataOpenApi>;
    total?: number;
};

export type OpenapiPriceHistoryRespOpenApi = {
    /**
     * Historical price data points
     */
    history?: Array<OpenapiPricePoint>;
};

export type OpenapiPricePoint = {
    /**
     * Price
     */
    p?: string;
    /**
     * UTC timestamp in seconds
     */
    t?: number;
};

export type OpenapiQuoteTokenBalance = {
    /**
     * Available balance (formatted with decimals)
     */
    availableBalance?: string;
    /**
     * Frozen balance (formatted with decimals)
     */
    frozenBalance?: string;
    /**
     * Quote token address
     */
    quoteToken?: string;
    /**
     * Token decimals
     */
    tokenDecimals?: number;
    /**
     * Total balance (formatted with decimals)
     */
    totalBalance?: string;
};

export type OpenapiQuoteTokenDataOpenApi = {
    chainId?: string;
    /**
     * Time fields (moved to bottom)
     */
    createdAt?: number;
    ctfExchangeAddress?: string;
    decimal?: number;
    id?: number;
    quoteTokenAddress?: string;
    quoteTokenName?: string;
    symbol?: string;
};

export type OpenapiQuoteTokenListRespOpenApi = {
    list?: Array<OpenapiQuoteTokenDataOpenApi>;
    total?: number;
};

export type OpenapiTradeDataOpenApi = {
    /**
     * Amount in quote token (e.g., USDC)
     */
    amount?: string;
    chainId?: string;
    /**
     * Time fields (moved to bottom)
     */
    createdAt?: number;
    /**
     * Financial info
     */
    fee?: number;
    /**
     * Human-readable fee value (e.g., "0.5")
     */
    feeFormatted?: string;
    /**
     * Market info
     */
    marketId?: number;
    marketTitle?: string;
    /**
     * Trade identification
     */
    orderNo?: string;
    outcome?: string;
    /**
     * 1-yes, 2-no
     */
    outcomeSide?: number;
    /**
     * Human-readable: Yes, No
     */
    outcomeSideEnum?: string;
    price?: string;
    profit?: string;
    quoteToken?: string;
    /**
     * USD price of quote token
     */
    quoteTokenUsdPrice?: string;
    rootMarketId?: number;
    rootMarketTitle?: string;
    shares?: string;
    /**
     * Trade details
     */
    side?: string;
    /**
     * Status and metadata
     */
    status?: number;
    /**
     * Human-readable: Pending, Finished, Canceled, Expired, Failed
     */
    statusEnum?: string;
    tradeNo?: string;
    txHash?: string;
    /**
     * Total USD value of this trade
     */
    usdAmount?: string;
};

export type OpenapiTradeListRespOpenApi = {
    list?: Array<OpenapiTradeDataOpenApi>;
    total?: number;
};

export type V2ActivationResponse = {
    /**
     * 激活交易对象
     */
    tx_object?: {
        [key: string]: unknown;
    };
};

export type V2ActivityItem = {
    /**
     * Amount
     */
    amount?: string;
    chainId?: string;
    /**
     * tradeTime
     */
    createTime?: number;
    currencyAddress?: string;
    /**
     * Price
     */
    price?: string;
    /**
     * profit - only sold order has profit
     */
    profit?: string;
    profitPercentage?: string;
    /**
     * show type : 1, 2
     */
    showType?: number;
    thumbnail_url?: string;
    /**
     * TokenId
     */
    tokenId?: string;
    /**
     * topicId
     */
    topicId?: number;
    /**
     * topicTitle
     */
    topicTitle?: string;
    /**
     * TotalPrice
     */
    totalPrice?: string;
    /**
     * transaction object: Yes or No
     */
    txObject?: string;
    /**
     * transaction type : 1-buy, 2-sell
     */
    txType?: string;
    /**
     * userAvatar
     */
    userAvatar?: string;
    /**
     * userName
     */
    userName?: string;
    /**
     * walletAddress
     */
    walletAddress?: string;
};

export type V2ActivityListResp = {
    /**
     * list
     */
    list?: Array<V2ActivityItem>;
    /**
     * total
     */
    total?: number;
};

export type V2AddEmailToWaitlistReq = {
    email: string;
};

export type V2AddEmailToWaitlistResp = {
    alreadyExists?: boolean;
    email?: string;
    joinedAt?: number;
    message?: string;
    success?: boolean;
};

export type V2AddEventReq = {
    cover_url?: string;
    datail_url?: string;
    desc?: string;
    end_time?: number;
    event_name?: string;
    image_mobile_url?: string;
    long_image_url?: string;
    sort_weight?: number;
    start_time?: number;
    thumbnail_url?: string;
};

export type V2AddEventResp = {
    event_id?: number;
};

export type V2AddOrderReq = {
    /**
     * conditionToken
     */
    contractAddress?: string;
    currencyAddress?: string;
    /**
     * Order expiration time stamp (no signature) for build order salt
     */
    expiration?: string;
    /**
     * A processing rate expressed in basis points, collected from the earnings of the order creator
     */
    feeRateBps?: string;
    /**
     * The creator of the order, that is, the source of the order funds
     */
    maker?: string;
    /**
     * The number of creators, i.e. the maximum number of tokens to be sold
     */
    makerAmount?: string;
    /**
     * nonce for on-chain order cancellation
     */
    nonce?: string;
    /**
     * Order expiration time stamp (for automatic cancellation)
     */
    orderExpTime?: string;
    origin?: string;
    /**
     * price
     */
    price?: string;
    /**
     * Market protection rate, the final transaction price deviates from the current spread, 0 means no protection, eg: 0.05
     */
    safeRate?: string;
    /**
     * A unique salt value that ensures entropy randomness
     */
    salt?: string;
    /**
     * Direction of the order: SELL or SELL
     */
    side?: string;
    /**
     * order unique sign
     */
    sign?: string;
    /**
     * Order signature
     */
    signature?: string;
    /**
     * The type of signature used for the order: EOA, POLY_PROXY or POLY_GNOSIS_SAFE
     */
    signatureType?: string;
    /**
     * The signer of the order
     */
    signer?: string;
    /**
     * The address of the recipient of the order. Use a zero address to indicate a public order
     */
    taker?: string;
    /**
     * The number of recipients is the minimum number of tokens to be received
     */
    takerAmount?: string;
    /**
     * sign timestamp
     */
    timestamp?: number;
    /**
     * To buy or sell the token of the c t f e r c1155 asset id
     */
    tokenId?: string;
    /**
     * topicId
     */
    topicId?: number;
    /**
     * Trading method :1- market price, 2- limit price
     */
    tradingMethod?: number;
};

export type V2AddOrderResp = {
    orderData?: V2OrderData;
};

export type V2AddRoleReq = {
    permissions?: Array<string>;
};

export type V2AddRoleResp = {
    role_id?: number;
};

export type V2AddUserRoleReq = {
    role_ids?: Array<string>;
    wallet_address?: string;
};

export type V2AddUserRoleResp = {
    user_role_id?: number;
};

export type V2ApplyReferralCodeRequest = {
    /**
     * 链ID
     */
    chainId: string;
};

export type V2BindToReferrerRequest = {
    /**
     * 链ID
     */
    chainId: string;
    /**
     * 推荐码
     */
    referralCode: string;
};

export type V2BindTraderRequest = {
    /**
     * 推荐人地址
     */
    referrerAddress: string;
    /**
     * 签名有效期(小时)
     */
    signatureValidHours: number;
    /**
     * 交易者地址
     */
    traderAddress: string;
};

export type V2BotActivateReq = {
    addresses?: Array<string>;
};

export type V2BotActivateResp = {
    count?: number;
};

export type V2CallbackResponse = {
    message?: string;
    success?: boolean;
};

export type V2CheckRestrictAreaReq = {
    city?: string;
    continent?: string;
    country?: string;
    isEUCountry?: boolean;
    latitude?: string;
    longitude?: string;
    postalCode?: string;
    region?: string;
    regionCode?: string;
    timezone?: string;
};

export type V2CheckRestrictAreaResp = {
    isRestricted?: boolean;
};

export type V2CheckWhitelistResponse = {
    result?: boolean;
    type?: string;
    walletAddress?: string;
};

export type V2ChildTopicData = {
    /**
     * topic abstract
     */
    abstract?: string;
    chainId?: string;
    /**
     * Number of comments
     */
    commentCount?: number;
    conditionId?: string;
    /**
     * topic's content
     */
    content?: string;
    /**
     * cover picture of this topic
     */
    coverUrl?: string;
    /**
     * created time
     */
    createTime?: number;
    /**
     * creator's wallet address
     */
    creatorAddress?: string;
    /**
     * creator's avatar
     */
    creatorAvatar?: string;
    /**
     * management user id
     */
    creatorId?: number;
    /**
     * creator's name
     */
    creatorName?: string;
    /**
     * currency address
     */
    currencyAddress?: string;
    cutoffTime?: number;
    feeRateBps?: string;
    /**
     * hold no count
     */
    holdNoCount?: string;
    /**
     * hold no price
     */
    holdNoPrice?: string;
    /**
     * hold yes count
     */
    holdYesCount?: string;
    /**
     * hold price
     */
    holdYesPrice?: string;
    /**
     * Percentage change
     */
    incRate?: string;
    incentiveFactor?: V2IncentiveFactor;
    /**
     * AI created flag
     */
    isCreatedByAI?: boolean;
    /**
     * mock topic flag
     */
    isMock?: boolean;
    /**
     * AI resolvable flag
     */
    isResolvableByAI?: boolean;
    isShow?: number;
    /**
     * kline thumbnail
     */
    klineThumbnail?: Array<ModelsKlineData>;
    /**
     * labelId array of this topic
     */
    labelId?: Array<number>;
    /**
     * labelName array of this topic
     */
    labelName?: Array<string>;
    noBuyPrice?: string;
    /**
     * display text for no
     */
    noLabel?: string;
    noMarketPrice?: string;
    /**
     * No-Position
     */
    noPos?: string;
    noRemainToken?: string;
    noSellPrice?: string;
    questionId?: string;
    resolveExpression?: string;
    resolvedTime?: number;
    /**
     * win pos
     */
    resultPos?: string;
    rules?: string;
    sortWeight?: number;
    status?: ModelsTopicStatus;
    /**
     * thumbnail picture of this topic
     */
    thumbnailUrl?: string;
    /**
     * topic title
     */
    title?: string;
    /**
     * topic title short
     */
    titleShort?: string;
    /**
     * topId
     */
    topicId?: number;
    /**
     * total bet
     */
    totalPrice?: string;
    volume?: string;
    volume24h?: string;
    volume7d?: string;
    yesBuyPrice?: string;
    /**
     * display text for yes
     */
    yesLabel?: string;
    yesMarketPrice?: string;
    /**
     * Yes-Position
     */
    yesPos?: string;
    yesRemainToken?: string;
    yesSellPrice?: string;
};

export type V2ClaimResp = {
    /**
     * claim status:0 - can''t claim, 1 - wait claim, 2 - claiming, 3 - claim failed, 4 - claimed, while claim failed, user can claim again
     */
    claimStatus?: number;
};

export type V2CommentItem = {
    /**
     * commentator's avatar
     */
    avatar?: string;
    /**
     * child comment
     */
    childComment?: Array<V2CommentItem>;
    /**
     * comment content
     */
    content?: string;
    /**
     * create time
     */
    createTime?: number;
    /**
     * hold no count
     */
    holdNoCount?: string;
    /**
     * hold no price
     */
    holdNoPrice?: string;
    /**
     * hold yes count
     */
    holdYesCount?: string;
    /**
     * hold price
     */
    holdYesPrice?: string;
    /**
     * ID
     */
    id?: number;
    /**
     * is liked
     */
    isLiked?: boolean;
    /**
     * liked count
     */
    likedCount?: number;
    /**
     * parent comment id
     */
    parentId?: number;
    /**
     * top parent id
     */
    topParentId?: number;
    /**
     * total ChildComment
     */
    totalChildComment?: number;
    /**
     * commentator's nickname
     */
    userName?: string;
    /**
     * commentator's wallet address
     */
    walletAddress?: string;
};

export type V2CommentListResp = {
    /**
     * comment list
     */
    list?: Array<V2CommentItem>;
    /**
     * total count
     */
    total?: number;
};

export type V2CommentReq = {
    /**
     * comment content
     */
    content: string;
    /**
     * parentId
     */
    parentId?: number;
};

export type V2CommentResp = {
    /**
     * data
     */
    data?: V2CommentItem;
};

export type V2CompletionRecord = {
    complete_time?: string;
    reward_claim_time?: string;
    reward_claimed?: boolean;
};

export type V2CreateMockTopicReq = {
    abstract?: string;
    /**
     * Optional: category of this prediction market
     */
    category?: string;
    chain_id?: string;
    content?: string;
    cover_url?: string;
    currency_address?: string;
    cutoff_time: number;
    is_resolvable_by_ai?: boolean;
    label_ids?: Array<number>;
    no_label: string;
    oracle_address?: string;
    /**
     * Optional: preferred source to resolve this market
     */
    preferredSource?: string;
    rules?: string;
    thumbnail_url?: string;
    title: string;
    title_short?: string;
    yes_label: string;
};

export type V2CreateMockTopicResp = {
    topic_id?: number;
};

export type V2CreateMultiTopicReq = {
    /**
     * cover picture of this topic
     */
    cover_url?: string;
    cutoff_time?: number;
    label_ids?: Array<number>;
    liquidity_factor?: number;
    position_factor?: number;
    rules?: string;
    sort_child_by?: number;
    /**
     * thumbnail picture of this topic
     */
    thumbnail_url?: string;
    title?: string;
    title_short?: string;
    topic_list?: Array<V2CreateTopicReq>;
    volume_factor?: number;
};

export type V2CreateMultiTopicResp = {
    mutil_topic_id?: number;
    topic_id_list?: Array<number>;
};

export type V2CreateResolutionProposalReq = {
    /**
     * Expire time in hours, default 72
     */
    expire_hours?: number;
    /**
     * Result data
     */
    result_data?: string;
    /**
     * Result URL
     */
    result_url?: string;
    /**
     * 0=No, 1=Yes
     */
    win_slot?: number;
};

export type V2CreateResolutionProposalResp = {
    proposal_id?: number;
};

export type V2CreateTopicLabelReq = {
    image_mobile_url?: string;
    image_url?: string;
    is_show?: number;
    label_name?: string;
    weight?: number;
};

export type V2CreateTopicLabelResp = {
    label_id?: number;
};

export type V2CreateTopicPinReq = {
    chain_id?: string;
    sort_weight?: number;
    topic_id?: number;
    topic_type?: number;
};

export type V2CreateTopicPinResp = {
    topic_pin_id?: number;
};

export type V2CreateTopicReq = {
    /**
     * topic abstract
     */
    abstract?: string;
    /**
     * Optional: category of this prediction market
     */
    category?: string;
    chain_id?: string;
    /**
     * topic's content
     */
    content?: string;
    /**
     * cover picture of this topic
     */
    cover_url?: string;
    currency_address?: string;
    cutoff_time?: number;
    isJudgeUser?: boolean;
    is_show?: number;
    label_ids?: Array<number>;
    liquidity_factor?: number;
    /**
     * display text for no
     */
    no_label?: string;
    oracle_address?: string;
    position_factor?: number;
    /**
     * Optional: preferred source to resolve this market
     */
    preferredSource?: string;
    resolve_expression?: string;
    rules?: string;
    sort_weight?: number;
    /**
     * thumbnail picture of this topic
     */
    thumbnail_url?: string;
    title?: string;
    title_short?: string;
    volume_factor?: number;
    /**
     * display text for yes
     */
    yes_label?: string;
};

export type V2CreateTopicResp = {
    topic_id?: number;
};

export type V2CreatorLeaderboardItem = {
    /**
     * user's avatar
     */
    avatar?: string;
    id?: number;
    rankingChange?: number;
    rankingType?: number;
    /**
     * ranking value : value of points|volume|profit
     */
    rankingValue?: string;
    /**
     * user's nickname
     */
    userName?: string;
    /**
     * user's wallet address
     */
    walletAddress?: string;
    x_user_id?: string;
    x_username?: string;
};

export type V2CreatorLeaderboardResp = {
    list?: Array<V2CreatorLeaderboardItem>;
    total?: number;
};

export type V2CreatorTopMoversItem = {
    /**
     * user's avatar
     */
    avatar?: string;
    id?: number;
    /**
     * ranking value : value of points|volume|profit
     */
    rankingChange?: number;
    rankingType?: number;
    /**
     * user's nickname
     */
    userName?: string;
    /**
     * user's wallet address
     */
    walletAddress?: string;
};

export type V2CreatorTopMoversResp = {
    date?: string;
    list?: Array<V2CreatorTopMoversItem>;
    total?: number;
};

export type V2CurrencyBalance = {
    available?: string;
    contract?: string;
    frozen?: string;
    total?: string;
};

export type V2CustomAirdropResp = {
    result?: string;
};

export type V2DeleteCommentResp = {
    result?: boolean;
};

export type V2DeleteEventResp = {
    result?: boolean;
};

export type V2DeleteTopicLabelResp = {
    result?: boolean;
};

export type V2DirectResolveTopicReq = {
    /**
     * "yes" or "no" or token symbol
     */
    winningSide: string;
};

export type V2DirectResolveTopicResp = {
    message?: string;
    success?: boolean;
};

export type V2DisconnectResponse = {
    message?: string;
    success?: boolean;
};

export type V2EventData = {
    cover_url?: string;
    create_time?: number;
    datail_url?: string;
    desc?: string;
    end_time?: number;
    event_name?: string;
    id?: number;
    image_mobile_url?: string;
    long_image_url?: string;
    sort_weight?: number;
    start_time?: number;
    thumbnail_url?: string;
};

export type V2EventListResp = {
    list?: Array<V2EventData>;
    total?: number;
};

export type V2FollowItem = {
    /**
     * user avatar
     */
    avatar?: string;
    /**
     * be followed user's wallet address
     */
    followedUserWalletAddress?: string;
    /**
     * Number of followers
     */
    followerCount?: number;
    /**
     * id
     */
    id?: number;
    /**
     * user nickname
     */
    userName?: string;
};

export type V2FollowListResp = {
    /**
     * list
     */
    list?: Array<V2FollowItem>;
    /**
     * total
     */
    total?: number;
};

export type V2FollowResp = {
    /**
     * result : true for success, false for failure
     */
    result?: boolean;
};

export type V2GetApiKeyResp = {
    apiKey?: string;
    walletAddress?: string;
    walletUsers?: {
        [key: string]: string;
    };
};

export type V2GetBalanceResp = {
    balance?: string;
    chainId?: string;
    currencyAddress?: string;
    netWorth?: string;
    /**
     * AvatarUrl                string `json:"avatarUrl"`
     */
    totalBalance?: string;
};

export type V2GetConnectUrlResponse = {
    url?: string;
};

export type V2GetCumulativeTradingVolumeResp = {
    chainId?: string;
    cumulativeVolume?: string;
};

export type V2GetEmailVerifyCodeReq = {
    email: string;
};

export type V2GetEmailVerifyCodeResp = {
    result?: boolean;
};

export type V2GetMultiSignedWalletAddressResp = {
    multiSignedWalletAddress?: {
        [key: string]: string;
    };
};

export type V2GetNewsListReq = {
    countryCode?: string;
    indicatorId?: number;
    limit?: number;
    page?: number;
};

export type V2GetNewsListResp = {
    list?: Array<V2IndicatorNewsData>;
    total?: number;
};

export type V2GetNewsSourceListReq = {
    countryCode?: string;
    indicatorId?: number;
    limit?: number;
    page?: number;
};

export type V2GetNewsSourceListResp = {
    list?: Array<V2IndicatorNewsSourceData>;
    total?: number;
};

export type V2GetReleaseListReq = {
    chainId?: string;
    countryCode?: string;
    indicatorId?: number;
    limit?: number;
    page?: number;
};

export type V2GetReleaseListResp = {
    list?: Array<V2IndicatorReleaseData>;
    total?: number;
};

export type V2GetSnapshotsResp = {
    chainId?: string;
    snapshots?: Array<V2UserSnapshot>;
};

export type V2GetStatusResponse = {
    connected?: boolean;
    username?: string;
};

export type V2GetTokenBalancesOfBotsByTopicIdResp = {
    balances?: Array<ModelsTokenBalance>;
    totalBalance?: string;
};

export type V2GetTradesMetricsResp = {
    metrics?: ModelsTradeMetrics;
};

export type V2GetUserCampaignProgressReq = {
    chainId?: string;
};

export type V2GetUserCampaignProgressResp = {
    orders?: number;
    shares?: number;
    volume?: number;
};

export type V2GetVolumeDistributionResp = {
    volumeDistribution?: Array<V2VolumeDistribution>;
};

export type V2GetWatchListResp = {
    /**
     * total count
     */
    total?: number;
    /**
     * watch list
     */
    watchList?: Array<V2TopicData>;
};

export type V2GnosisSafeTxSubmissionReq = {
    base_gas?: string;
    data?: string;
    /**
     * may result in extra rpc calls and time out if true
     */
    debug?: boolean;
    gas_price?: string;
    gas_token?: string;
    nonce?: string;
    operation?: number;
    refund_receiver?: string;
    safe_tx_gas?: string;
    signatures?: string;
    to_address?: string;
    value?: string;
};

export type V2GnosisSafeTxSubmissionResp = {
    safe_tx_hash?: string;
    tx_hash?: string;
};

export type V2GnosisSafeWalletCreationResp = {
    calldata?: string;
    gnosis_safe_proxy_factory?: string;
    is_created?: boolean;
    success?: boolean;
};

export type V2HolderItem = {
    /**
     * avatar
     */
    avatar?: string;
    /**
     * profit
     */
    profit?: string;
    /**
     * sharesAmount
     */
    sharesAmount?: string;
    /**
     * userName
     */
    userName?: string;
    /**
     * walletAddress
     */
    walletAddress?: string;
};

export type V2HolderListResp = {
    /**
     * list
     */
    list?: Array<V2HolderItem>;
    /**
     * total
     */
    total?: number;
};

export type V2IncentiveFactor = {
    liquidity?: number;
    position?: number;
    volume?: number;
};

export type V2IndicatorCategoryData = {
    id?: number;
    name?: string;
    /**
     * 1: active, 2: disabled
     */
    status?: number;
    weight?: number;
};

export type V2IndicatorCountResp = {
    total?: number;
};

export type V2IndicatorCountryData = {
    boundingBox?: Array<number>;
    capitalCity?: string;
    capitalCoordinates?: Array<number>;
    countryCode?: string;
    countryName?: string;
    id?: number;
    /**
     * 1: active, 2: disabled
     */
    status?: number;
    weight?: number;
};

export type V2IndicatorData = {
    categoryIds?: Array<number>;
    categoryIdsStr?: string;
    chainId?: string;
    countryCode?: string;
    description?: string;
    forecastValue?: string;
    frequency?: string;
    id?: number;
    importance?: number;
    oracleId?: string;
    period?: string;
    previousValue?: string;
    releaseId?: number;
    releaseTime?: number;
    releases?: Array<V2IndicatorReleaseData>;
    /**
     * 1: active, 2: disabled
     */
    status?: number;
    title?: string;
    titleShort?: string;
    topic?: V2TopicData;
    topicId?: number;
    /**
     * 0: single topic, 1:mutil topic
     */
    topicType?: number;
    unit?: string;
};

export type V2IndicatorListReq = {
    categoryId?: number;
    chainId?: string;
    countryCode?: string;
    limit?: number;
    page?: number;
    /**
     * default 0: release time, 1: importance, 2: country, 3: title, 4: previous, 5: forecast, 6: topic volume
     */
    sortBy?: V2IndicatorSortBy;
    /**
     * default 0: asc, 1: desc
     */
    sortDirection?: V2IndicatorSortDirection;
};

export type V2IndicatorListResp = {
    list?: Array<V2IndicatorData>;
    total?: number;
};

export type V2IndicatorNewsData = {
    author?: string;
    countryCode?: string;
    description?: string;
    fetchSource?: string;
    id?: number;
    imageUrl?: string;
    indicatorId?: number;
    publishTime?: number;
    source?: string;
    /**
     * 1: active, 2: disabled
     */
    status?: number;
    title?: string;
    url?: string;
};

export type V2IndicatorNewsSourceData = {
    countryCode?: string;
    id?: number;
    indicatorId?: number;
    parameter?: string;
    source?: string;
    /**
     * 1: active, 2: disabled
     */
    status?: number;
};

export type V2IndicatorNewsSourceTypeData = {
    paramPattern?: string;
    reference?: string;
    sourceName?: string;
};

export type V2IndicatorReleaseData = {
    chainId?: string;
    forecastValue?: string;
    id?: number;
    indicatorId?: number;
    oracleId?: string;
    period?: string;
    previousValue?: string;
    releaseTime?: number;
    status?: number;
    topicId?: number;
    topicType?: number;
    value?: string;
};

export type V2IndicatorSimpleData = {
    chainId?: string;
    countryCode?: string;
    id?: number;
    title?: string;
};

export type V2IndicatorSortBy = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export type V2IndicatorSortDirection = 0 | 1;

export type V2InvitationRecord = {
    code?: string;
    createTime?: string;
    walletAddress?: string;
    walletUser?: string;
};

export type V2LeaderboardItem = {
    /**
     * user's avatar
     */
    avatar?: string;
    id?: number;
    rankingChange?: number;
    rankingType?: number;
    /**
     * ranking value : value of points|volume|profit
     */
    rankingValue?: string;
    /**
     * user's nickname
     */
    userName?: string;
    /**
     * user's wallet address
     */
    walletAddress?: string;
    /**
     * user's x account
     */
    xUserId?: string;
    xUsername?: string;
};

export type V2LeaderboardResp = {
    list?: Array<V2LeaderboardItem>;
    total?: number;
};

export type V2LikeCommentResp = {
    /**
     * Result
     */
    result?: boolean;
};

export type V2ListResolutionProposalsResp = {
    limit?: number;
    list?: Array<V2ResolutionProposalResp>;
    page?: number;
    total?: number;
};

export type V2MarketEvaluateReq = {
    expectedEndTime?: number;
    marketTitle: string;
};

export type V2MarketEvaluateResp = {
    expectedEndTime?: number;
    isResolvableByAI?: boolean;
    message?: string;
    options?: Array<string>;
    resolutionCriteria?: string;
};

export type V2ModelAssessment = {
    confidenceScore?: number;
    modelName?: string;
    reasoning?: string;
    resolvable?: boolean;
    result?: string;
    timestamp?: number;
};

export type V2ModifyUserReq = {
    /**
     * avatar
     */
    avatarUrl?: string;
    /**
     * email
     */
    email?: string;
    /**
     * This parameter is required when modifying email addresses
     */
    emailVerifyCode?: string;
    /**
     * introduction
     */
    introduction?: string;
    inviteCode?: string;
    sources?: string;
    userName?: string;
};

export type V2ModifyUserResp = {
    /**
     * avatar
     */
    avatarUrl?: string;
    /**
     * email
     */
    email?: string;
    /**
     * introduction
     */
    introduction?: string;
    userName?: string;
};

export type V2NoticeItem = {
    createTime?: number;
    id?: number;
    /**
     * notice content with json string, tx finished: {"topicTitle":"xxx","txCount":"xxx","txAmount":"88.999","txAvgPrice":"xxx"}, price fluctuation: {"topicTitle":"xxx","fluctuation":"0.01","fluctuationTime":"1719539958","rangeOfFluctuation":"0.1","valueOfFluctuation":"88.8"}
     */
    noticeContent?: string;
    readTime?: number;
    status?: number;
    /**
     * notice type: 1-transaction finished, 2-price fluctuation
     */
    type?: number;
    walletAddress?: string;
};

export type V2NoticeListResp = {
    list?: Array<V2NoticeItem>;
    total?: number;
};

export type V2NoticeUnreadCountsResp = {
    /**
     * 未读消息数
     */
    count?: number;
};

export type V2OrderData = {
    /**
     * Amount field represents different things based on Side:
     * - For BUY orders (Side=1): Amount = USDC (quote token) - the total USDC to spend
     * - For SELL orders (Side=2): Amount = Shares (base token) - the total shares to sell
     * This is derived from MakerAmount in CTF Exchange standard where:
     * BUY: MakerAmount=USDC, TakerAmount=Shares
     * SELL: MakerAmount=Shares, TakerAmount=USDC
     */
    amount?: string;
    chainId?: string;
    createdAt?: number;
    currencyAddress?: string;
    expiration?: number;
    filled?: string;
    finishAmount?: string;
    finishShare?: string;
    mutilTitle?: string;
    mutilTopicId?: number;
    orderId?: number;
    outcome?: string;
    /**
     * 1 - yes, 2 - no
     */
    outcomeSide?: number;
    price?: string;
    profit?: string;
    /**
     * 1-for buy, 2-for sell
     */
    side?: number;
    /**
     * 1-pending, 2-finished, 3-canceled, 4-expired, 5-failed
     */
    status?: number;
    topicId?: number;
    topicTitle?: string;
    totalPrice?: string;
    /**
     * 买卖方式:1-市价,2-现价
     */
    tradingMethod?: number;
    transNo?: string;
};

export type V2OrderListResp = {
    list?: Array<V2OrderData>;
    total?: number;
};

export type V2PointAirdrop = {
    amount?: number;
    chain_id?: string;
    wallet_address?: string;
};

export type V2Points = {
    avatar?: string;
    points?: string;
    userName?: string;
    walletAddress?: string;
    walletUser?: string;
};

export type V2PollResultItem = {
    /**
     * chain id
     */
    chainId?: string;
    /**
     * claim status:0 - can''t claim, 1 - wait claim, 2 - claiming, 3 - claim failed, 4 - claimed, while claim failed, user can claim again
     */
    claimStatus?: number;
    conditionId?: string;
    /**
     * currency address
     */
    currencyAddress?: string;
    /**
     * topicTitle
     */
    mutilTitle?: string;
    /**
     * id of topic
     */
    mutilTopicId?: number;
    noAmount?: string;
    noLabel?: string;
    noPositionId?: string;
    profit?: string;
    profitRate?: string;
    questionId?: string;
    recordId?: number;
    /**
     * result position: yes or no
     */
    resultPosition?: string;
    /**
     * show type : 1, 2
     */
    showType?: number;
    /**
     * timestamp
     */
    timestamp?: number;
    /**
     * id of topic
     */
    topicId?: number;
    /**
     * topicPicture
     */
    topicPicture?: string;
    /**
     * topicTitle
     */
    topicTitle?: string;
    /**
     * Total price  : same as the price of the claim
     */
    totalPrice?: string;
    /**
     * Total shares
     */
    totalShares?: string;
    /**
     * "Yes or No"
     */
    txObject?: string;
    yesAmount?: string;
    yesLabel?: string;
    yesPositionId?: string;
};

export type V2PollResultResp = {
    list?: Array<V2PollResultItem>;
    total?: number;
};

export type V2PortalCancelOrderByWalletAddressesReq = {
    walletAddresses?: Array<string>;
};

export type V2PortalCancelOrderByWalletAddressesResp = {
    [key: string]: Array<string>;
};

export type V2PortalFixBalanceByWalletAddressesReq = {
    chainId?: string;
    walletAddresses?: Array<string>;
};

export type V2PortalFixBalanceByWalletAddressesResp = {
    failed?: {
        [key: string]: V2PortalFixBalanceResp;
    };
    success?: {
        [key: string]: V2PortalFixBalanceResp;
    };
};

export type V2PortalFixBalanceReq = {
    currency?: Array<string>;
    token?: Array<string>;
};

export type V2PortalFixBalanceResp = {
    currency?: Array<V2CurrencyBalance>;
    err_msg?: string;
    token?: Array<V2TokenBalance>;
    transNoToCancel?: Array<string>;
    wallet_address?: string;
    wallet_user?: string;
};

export type V2PortalGetBalanceResp = {
    currency?: Array<V2CurrencyBalance>;
    err_msg?: string;
    token?: Array<V2TokenBalance>;
    wallet_address?: string;
    wallet_user?: string;
};

export type V2PortalIncentiveRewardConfig = {
    cap?: number;
    makerFactor?: number;
    takerFactor?: number;
    type?: V2PortalIncentiveType;
};

export type V2PortalIncentiveRewardReq = {
    description?: string;
    totalReward?: number;
    type?: V2PortalIncentiveType;
    weekId?: string;
};

export type V2PortalIncentiveScoreResp = {
    list?: Array<unknown>;
    total?: number;
    totalScore?: number;
};

export type V2PortalIncentiveType = 'all' | 'liquidity' | 'volume' | 'position' | 'custom';

export type V2PortalLoginReq = {
    nonce?: string;
    sign?: string;
    siwe_message?: string;
    timestamp?: number;
};

export type V2PortalLoginResp = {
    address?: string;
    expire?: number;
    token?: string;
};

export type V2PortfolioData = {
    claimStatus?: number;
    conditionId?: string;
    currency_address?: string;
    dayProfit?: string;
    dayProfitRate?: string;
    lastPrice?: string;
    mutilTitle?: string;
    mutilTopicId?: number;
    outcome?: string;
    /**
     * 1 - yes, 2 - no
     */
    outcomeSide?: number;
    positionAvgPrice?: string;
    profit?: string;
    profitRate?: string;
    tokenAmount?: string;
    tokenFrozenAmount?: string;
    tokenId?: string;
    topicCutoffTime?: number;
    topicId?: number;
    topicStatus?: number;
    topicTitle?: string;
    /**
     * FirstOrderTime    uint64 `json:"firstCreateTime"`
     * LastOrderTime     uint64 `json:"lastCreateTime"`
     */
    value?: string;
};

export type V2PortfolioResp = {
    list?: Array<V2PortfolioData>;
    total?: number;
};

export type V2ProgressOfTopicCreatedByUserResp = {
    create_time?: number;
    creation_fee_time?: number;
    /**
     * creation fee tx hash
     */
    creation_fee_tx_hash?: string;
    /**
     * creator's avatar url
     */
    creator_avatar_url?: string;
    /**
     * creator's name
     */
    creator_name?: string;
    resolve_time?: number;
    review_time?: number;
    /**
     * topic status
     */
    status?: ModelsTopicStatus;
    topic_creation_status?: ModelsTopicCreationStatus;
    topic_id?: number;
};

export type V2QueryInvitationCodeResp = {
    code?: string;
};

export type V2QueryInvitationRecordResp = {
    list?: Array<V2InvitationRecord>;
    total?: number;
    totalTrade?: number;
};

export type V2QueryPointsLeaderBoardResp = {
    list?: Array<V2Points>;
    rank?: number;
};

export type V2QueryPointsRecord = {
    description?: string;
    points?: number;
    recordId?: string;
    type?: string;
    updateTime?: string;
    walletAddress?: string;
    walletUser?: string;
};

export type V2QueryPointsRecordResp = {
    list?: Array<V2QueryPointsRecord>;
    total?: number;
};

export type V2QueryPointsResp = {
    todayPoints?: number;
    totalPoints?: number;
    walletAddress?: string;
    walletUser?: string;
};

export type V2QuestData = {
    chain_id?: string;
    /**
     * Complete count for current period (daily, weekly, etc.). For regular quests, it equals to TotalComplete.
     */
    complete_count?: number;
    end_time?: string;
    expire_time?: string;
    is_active?: boolean;
    last_complete_time?: string;
    max_complete?: number;
    /**
     * for nft quest
     */
    nft_contract?: string;
    nft_token_ids?: Array<string>;
    quest_key?: string;
    quest_type?: string;
    reward_amount?: number;
    reward_type?: string;
    /**
     * for social quest
     */
    social_uri?: string;
    start_time?: string;
    /**
     * available, completed, expired
     */
    status?: string;
    title?: string;
    /**
     * Total complete count
     */
    total_complete?: number;
};

export type V2QuestListResp = {
    quest_list?: Array<V2QuestData>;
};

export type V2QuestResp = {
    completion_record?: V2CompletionRecord;
    quest?: V2QuestData;
};

export type V2ReadAllResp = {
    result?: boolean;
};

export type V2ReferralCodeResponse = {
    /**
     * 链ID
     */
    chainId?: string;
    /**
     * 创建时间
     */
    createTime?: number;
    /**
     * 最大邀请人数
     */
    maxReferrals?: number;
    /**
     * 推荐码
     */
    referralCode?: string;
    /**
     * 激活签名
     */
    signature?: string;
};

export type V2ReferralCommission = {
    /**
     * 返佣金额 (wei)
     */
    amount?: string;
    /**
     * 区块号
     */
    block_number?: number;
    chain_id?: number;
    /**
     * 抵押品代币地址
     */
    collateral_token?: string;
    /**
     * 创建时间
     */
    create_time?: string;
    id?: number;
    /**
     * 推荐人地址
     */
    referrer_address?: string;
    /**
     * 事件时间戳
     */
    timestamp?: number;
    /**
     * 交易者地址
     */
    trader_address?: string;
    /**
     * 交易哈希
     */
    tx_hash?: string;
};

export type V2ReferralCommissionsResponse = {
    commissions?: Array<V2ReferralCommission>;
    limit?: number;
    page?: number;
    total?: number;
    total_pages?: number;
};

export type V2ReferredTrader = {
    bind_time?: number;
    last_trade_time?: number;
    referral_code?: string;
    total_rebate_earned?: number;
    total_trades?: number;
    total_volume?: number;
    trader_address?: string;
};

export type V2ReferredTradersResponse = {
    limit?: number;
    page?: number;
    total?: number;
    total_pages?: number;
    traders?: Array<V2ReferredTrader>;
};

export type V2ResolutionCommitteeListResp = {
    members?: Array<V2ResolutionCommitteeMember>;
    total?: number;
};

export type V2ResolutionCommitteeMember = {
    has_voted?: boolean;
    name?: string;
    /**
     * 0=not voted, 1=approve, 2=reject
     */
    vote?: number;
    wallet_address?: string;
};

export type V2ResolutionEvaluateReq = {
    topicId: number;
};

export type V2ResolutionEvaluateResp = {
    confidenceScore?: number;
    message?: string;
    modelsAssessment?: Array<V2ModelAssessment>;
    resolvableOverall?: boolean;
    result?: string;
    timestamp?: number;
};

export type V2ResolutionProposalResp = {
    committee_members?: Array<V2ResolutionCommitteeMember>;
    create_time?: number;
    current_approvals?: number;
    current_rejections?: number;
    executed_at?: number;
    executed_tx_hash?: string;
    expire_time?: number;
    id?: number;
    proposer_address?: string;
    proposer_name?: string;
    required_approvals?: number;
    result_data?: string;
    result_url?: string;
    status?: number;
    status_text?: string;
    topic_id?: number;
    topic_title?: string;
    votes?: Array<V2ResolutionVoteRecordResp>;
    /**
     * "Yes" or "No"
     */
    win_label?: string;
    win_slot?: number;
};

export type V2ResolutionVoteRecordResp = {
    comment?: string;
    create_time?: number;
    vote?: number;
    /**
     * "approve" or "reject"
     */
    vote_text?: string;
    voter_address?: string;
    voter_name?: string;
};

export type V2ResolveTopicReq = {
    result_data?: string;
    result_url?: string;
    /**
     * 0 for no, 1 for yes
     */
    win_slot?: number;
};

export type V2SignatureResponse = {
    /**
     * EIP-712 签名
     */
    signature?: string;
};

export type V2StatisticsReq = {
    chainId?: string;
};

export type V2StatisticsResp = {
    activeTraders?: number;
    activeUsers?: number;
    openInterest?: number;
    totalDeposit?: number;
    tradingVolume?: number;
    tradingVolume7d?: number;
};

export type V2StatisticsVolumeReq = {
    chainId?: string;
    startOfDay?: number;
};

export type V2StatisticsVolumeResp = {
    dailyVolume?: number;
    totalVolume?: number;
};

export type V2SubmitTopicCreationFeeTxHashReq = {
    creation_fee_tx_hash?: string;
    token_type?: V2TokenType;
};

export type V2SubmitTopicCreationFeeTxHashResp = {
    result?: boolean;
};

export type V2SyncOracleIndicatorsReq = {
    chainId?: string;
};

export type V2SyncOracleReleasesReq = {
    chainId?: string;
    indicatorId?: number;
    oracleIndicatorId?: string;
};

export type V2TokenBalance = {
    available?: string;
    contract?: string;
    frozen?: string;
    token_id?: string;
    total?: string;
};

export type V2TokenType = 'erc20' | 'native';

export type V2TopicData = {
    /**
     * topic abstract
     */
    abstract?: string;
    chainId?: string;
    childList?: Array<V2ChildTopicData>;
    /**
     * Number of comments
     */
    commentCount?: number;
    conditionId?: string;
    /**
     * topic's content
     */
    content?: string;
    /**
     * cover picture of this topic
     */
    coverUrl?: string;
    /**
     * created time
     */
    createTime?: number;
    /**
     * creator's wallet address
     */
    creatorAddress?: string;
    /**
     * creator's avatar
     */
    creatorAvatar?: string;
    /**
     * management user id
     */
    creatorId?: number;
    /**
     * creator's name
     */
    creatorName?: string;
    currencyAddress?: string;
    cutoffTime?: number;
    feeRateBps?: string;
    /**
     * hold no count
     */
    holdNoCount?: string;
    /**
     * hold no price
     */
    holdNoPrice?: string;
    /**
     * hold yes count
     */
    holdYesCount?: string;
    /**
     * hold price
     */
    holdYesPrice?: string;
    /**
     * Percentage change
     */
    incRate?: string;
    incentiveFactor?: V2IncentiveFactor;
    /**
     * macro indicator
     */
    indicator?: V2IndicatorData;
    /**
     * AI created flag
     */
    isCreatedByAI?: boolean;
    /**
     * mock topic flag
     */
    isMock?: boolean;
    /**
     * AI resolvable flag
     */
    isResolvableByAI?: boolean;
    isShow?: number;
    /**
     * kline thumbnail
     */
    klineThumbnail?: Array<ModelsKlineData>;
    /**
     * labelId array of this topic
     */
    labelId?: Array<number>;
    labelIdStr?: string;
    /**
     * labelName array of this topic
     */
    labelName?: Array<string>;
    /**
     * Lifecycle stages
     */
    lifecycle?: Array<V2TopicLifecycleStageData>;
    noBuyPrice?: string;
    /**
     * display text for no
     */
    noLabel?: string;
    noMarketPrice?: string;
    /**
     * No-Position
     */
    noPos?: string;
    noRemainToken?: string;
    noSellPrice?: string;
    questionId?: string;
    resolveExpression?: string;
    resolvedTime?: number;
    /**
     * win pos
     */
    resultPos?: string;
    rules?: string;
    sortChildBy?: number;
    sortWeight?: number;
    status?: ModelsTopicStatus;
    /**
     * thumbnail picture of this topic
     */
    thumbnailUrl?: string;
    /**
     * topic title
     */
    title?: string;
    /**
     * topic title short
     */
    titleShort?: string;
    /**
     * topId
     */
    topicId?: number;
    /**
     * total bet
     */
    totalPrice?: string;
    volume?: string;
    volume24h?: string;
    volume7d?: string;
    yesBuyPrice?: string;
    /**
     * display text for yes
     */
    yesLabel?: string;
    yesMarketPrice?: string;
    /**
     * Yes-Position
     */
    yesPos?: string;
    yesRemainToken?: string;
    yesSellPrice?: string;
};

export type V2TopicDetailResp = {
    data?: V2TopicData;
};

export type V2TopicLifecycleStageData = {
    create_time?: number;
    metadata?: {
        [key: string]: unknown;
    };
    stage?: string;
    update_time?: number;
};

export type V2TopicListResp = {
    /**
     * list data
     */
    list?: Array<V2TopicData>;
    /**
     * total count
     */
    total?: number;
};

export type V2TopicPinData = {
    chain_id?: string;
    create_time?: number;
    id?: number;
    sort_weight?: number;
    topic_id?: number;
    topic_type?: number;
};

export type V2TopicPinListResp = {
    list?: Array<V2TopicPinData>;
    total?: number;
};

export type V2TradeBindingResponse = {
    /**
     * 链ID
     */
    chainId?: number;
    /**
     * 推荐码
     */
    referralCode?: string;
    /**
     * 推荐人地址
     */
    referrerAddress?: string;
    /**
     * 绑定签名
     */
    signature?: string;
    /**
     * 绑定时间戳
     */
    timestamp?: number;
    /**
     * 交易者地址
     */
    traderAddress?: string;
    /**
     * 交易对象
     */
    tx_object?: {
        [key: string]: unknown;
    };
};

export type V2TradeData = {
    amount?: string;
    avgPrice?: string;
    chainId?: string;
    createdAt?: number;
    currencyAddress?: string;
    currencyAmount?: string;
    currencyPrice?: string;
    /**
     * Human-readable fee value (e.g., "0.5")
     */
    fee?: string;
    isSelfTrade?: number;
    lastPrice?: string;
    mutilTitle?: string;
    mutilTopicId?: number;
    outcome?: string;
    /**
     * 1 - yes, 2 - no
     */
    outcomeSide?: number;
    price?: string;
    profit?: string;
    shares?: string;
    side?: string;
    /**
     * 1-pending, 2-finished, 3-canceled, 4-expired, 5-failed
     */
    status?: number;
    title?: string;
    topicId?: number;
    tradeNo?: string;
    tradeType?: number;
    transNo?: string;
    txHash?: string;
};

export type V2TradeListResp = {
    list?: Array<V2TradeData>;
    total?: number;
};

export type V2UnFollowResp = {
    /**
     * result : true for success, false for failure
     */
    result?: boolean;
};

export type V2UnWatchTopicResp = {
    /**
     * unwatch result : true for success, false for failure
     */
    result?: boolean;
};

export type V2UnlikeCommentResp = {
    /**
     * Result
     */
    result?: boolean;
};

export type V2UpdateEventReq = {
    cover_url?: string;
    datail_url?: string;
    desc?: string;
    end_time?: number;
    event_name?: string;
    image_mobile_url?: string;
    long_image_url?: string;
    sort_weight?: number;
    start_time?: number;
    thumbnail_url?: string;
};

export type V2UpdateEventResp = {
    result?: boolean;
};

export type V2UpdateMultiTopicReq = {
    /**
     * cover picture of this topic
     */
    cover_url?: string;
    cutoff_time?: number;
    label_ids?: Array<number>;
    liquidity_factor?: number;
    position_factor?: number;
    rules?: string;
    sort_child_by?: number;
    /**
     * thumbnail picture of this topic
     */
    thumbnail_url?: string;
    title?: string;
    title_short?: string;
    topic_list?: Array<V2UpdateTopicReq>;
    volume_factor?: number;
};

export type V2UpdateMultiTopicSortChildByReq = {
    sort_child_by?: number;
};

export type V2UpdateRoleReq = {
    id?: number;
    permissions?: Array<string>;
};

export type V2UpdateRoleResp = {
    result?: boolean;
};

export type V2UpdateTopicLabelResp = {
    result?: boolean;
};

export type V2UpdateTopicPinReq = {
    chain_id?: string;
    sort_weight?: number;
    topic_id?: number;
    topic_pin_id?: number;
    topic_type?: number;
};

export type V2UpdateTopicPinResp = {
    result?: boolean;
};

export type V2UpdateTopicReq = {
    /**
     * topic abstract
     */
    abstract?: string;
    /**
     * topic's content
     */
    content?: string;
    /**
     * cover picture of this topic
     */
    cover_url?: string;
    cutoff_time?: number;
    is_show?: number;
    label_ids?: Array<number>;
    liquidity_factor?: number;
    /**
     * display text for no
     */
    no_label?: string;
    position_factor?: number;
    resolve_expression?: string;
    rules?: string;
    sort_weight?: number;
    /**
     * thumbnail picture of this topic
     */
    thumbnail_url?: string;
    title?: string;
    title_short?: string;
    topic_id?: number;
    volume_factor?: number;
    /**
     * display text for yes
     */
    yes_label?: string;
};

export type V2UpdateTopicResp = {
    result?: boolean;
};

export type V2UpdateUserRoleReq = {
    role_ids?: Array<string>;
    wallet_address?: string;
};

export type V2UpdateUserRoleResp = {
    result?: boolean;
};

export type V2UserProfileResp = {
    alert?: boolean;
    avatarUrl?: string;
    balance?: Array<V2GetBalanceResp>;
    email?: string;
    followed?: boolean;
    follower?: number;
    following?: number;
    introduction?: string;
    location?: string;
    multiSignedWalletAddress?: {
        [key: string]: string;
    };
    netWorth?: string;
    portfolio?: string;
    /**
     * DayProfit                string            `json:"dayProfit"`
     */
    profitIncRate?: string;
    rankTheWeek?: number;
    score?: string;
    totalProfit?: string;
    userName?: string;
    volume?: string;
    volumeIncRate?: string;
    walletAddress?: string;
    xUserId?: string;
    xUsername?: string;
};

export type V2UserSnapshot = {
    date?: string;
    netCharge?: number;
    netWorth?: number;
    profit?: number;
    volume?: number;
};

export type V2ValidityEvaluateReq = {
    marketTitle: string;
};

export type V2ValidityEvaluateResp = {
    isValid?: boolean;
    message?: string;
    refinement?: V2ValidityRefinement;
    rewrittenTitle?: string;
};

export type V2ValidityRefinement = {
    /**
     * Optional: category of this prediction market
     */
    category?: string;
    expectedEndTime?: number;
    marketTitle?: string;
    options?: Array<string>;
    /**
     * Optional: preferred source to resolve this market
     */
    preferredSource?: string;
    resolutionCriteria?: string;
};

export type V2VolumeDistribution = {
    chainId?: string;
    label?: string;
    labelId?: number;
    volume?: number;
};

export type V2VoteResolutionProposalReq = {
    /**
     * Optional comment
     */
    comment?: string;
    /**
     * 1=approve, 2=reject
     */
    vote?: number;
};

export type V2VoteResolutionProposalResp = {
    current_approvals?: number;
    executed?: boolean;
    message?: string;
    proposal_status?: number;
    required_approvals?: number;
    success?: boolean;
};

export type V2WatchTopicResp = {
    /**
     * watch result : true for success, false for failure
     */
    result?: boolean;
};

export type ViewApiBaseResponse = {
    errmsg?: string;
    /**
     * 0:成功
     */
    errno?: number;
    result?: unknown;
};

export type ViewCreateTopicResp = {
    topic_id?: number;
};

export type ViewCurrencyData = {
    chain_id?: string;
    create_time?: number;
    ctfexchange_address?: string;
    currency_address?: string;
    currency_name?: string;
    decimal?: number;
    display_decimal?: number;
    id?: number;
    is_stable_coin?: number;
    order_min_amount?: number;
    status?: number;
    symbol?: string;
    withdraw_min_amount?: number;
};

export type ViewCurrencyListResp = {
    list?: Array<ViewCurrencyData>;
    total?: number;
};

export type ViewKlineBatchQuestion = {
    /**
     * 问题ID
     */
    question_id?: string;
    /**
     * Symbol数据数组，包含两种类型
     */
    symbols?: Array<ViewKlineBatchSymbol>;
};

export type ViewKlineBatchResp = {
    /**
     * K线周期
     */
    period?: string;
    /**
     * 问题数据列表
     */
    questions?: Array<ViewKlineBatchQuestion>;
    /**
     * 响应时间戳
     */
    ts?: number;
};

export type ViewKlineBatchSymbol = {
    /**
     * K线数据数组，按时间倒序排列
     */
    data?: Array<ViewKlineData>;
    /**
     * Symbol类型：0=原始价格, 1=镜像价格
     */
    symbol_type?: number;
    /**
     * 交易数据数组，按时间倒序排列
     */
    trades?: Array<ViewKlineTrade>;
};

export type ViewKlineData = {
    /**
     * 成交量
     */
    amount?: string;
    /**
     * 收盘价,当K线为最晚的一根时，是最新成交价
     */
    close?: string;
    /**
     * 成交笔数
     */
    count?: number;
    /**
     * 最高价
     */
    high?: string;
    /**
     * k线id
     */
    id?: number;
    /**
     * 最低价
     */
    low?: string;
    /**
     * 开盘价
     */
    open?: string;
    /**
     * 成交额, 即 sum(每一笔成交价 * 该笔的成交量)
     */
    vol?: string;
};

export type ViewKlineResp = {
    data?: Array<ViewKlineData>;
    /**
     * K线类型
     */
    period?: string;
    /**
     * 备注
     */
    question_id?: string;
    /**
     * 交易对 positionId
     */
    symbol?: string;
    /**
     * symbolTypes=0 yes symbol , symbolTypes=1 no symbol
     */
    symbol_types?: number;
    trades?: Array<ViewKlineTrade>;
    /**
     * 数据获取时间
     */
    ts?: number;
};

export type ViewKlineTrade = {
    amount?: string;
    price?: string;
    shares?: string;
    side?: string;
    time?: number;
};

export type ViewLabelData = {
    /**
     * icon
     */
    icon_url?: string;
    /**
     * 标签ID
     */
    label_id?: number;
    /**
     * 标签名称
     */
    label_name?: string;
};

export type ViewMarketDepthResp = {
    /**
     * 卖盘,[price(成交价), amount(成交量), user_amount(用户挂单数量)], 按price升序
     */
    asks?: Array<Array<string>>;
    /**
     * 买盘,[price(成交价), amount(成交量), user_amount(用户挂单数量)], 按price降序
     */
    bids?: Array<Array<string>>;
    /**
     * 最近成交价格
     */
    last_price?: string;
    /**
     * 备注
     */
    question_id?: string;
    /**
     * positionId
     */
    symbol?: string;
    /**
     * 数据获取时间
     */
    ts?: number;
};

export type ViewTokenFeeRateOnChainData = {
    /**
     * whether fee is enabled
     */
    enabled?: boolean;
    /**
     * error message if failed to read
     */
    error?: string;
    /**
     * maker fee in basis points
     */
    maker_fee_rate_bps?: string;
    /**
     * minimum fee amount
     */
    min_fee_amount?: string;
    /**
     * taker fee in basis points
     */
    taker_fee_rate_bps?: string;
    token_id?: string;
};

export type ViewTokenFeeRatesListReq = {
    chain_id?: string;
    limit?: number;
    page?: number;
    tokens?: Array<string>;
};

export type ViewTokenFeeRatesOnChainResp = {
    list?: Array<ViewTokenFeeRateOnChainData>;
};

export type ViewTopicLabelListResp = {
    label_list?: Array<ViewLabelData>;
};

export type ViewUpdateCurrencyReq = {
    status?: number;
};

export type ViewUpdateTokenFeeRatesReq = {
    /**
     * 代币地址
     */
    ctf_exchange?: string;
    /**
     * maker费率
     */
    maker_fee?: string;
    /**
     * 使用状态：0为禁用，1为启用
     */
    status?: number;
    /**
     * taker费率
     */
    taker_fee?: string;
    /**
     * token地址
     */
    token_id?: string;
};

export type ViewUpdateTokenFeeRatesResp = {
    result?: boolean;
};

export type ViewUserData = {
    avatar_url?: string;
    balance?: string;
    day_in?: string;
    day_out?: string;
    email?: string;
    introduction?: string;
    multi_signed_wallet_address?: string;
    user_id?: number;
    user_name?: string;
    wallet_address?: string;
    x_user_id?: string;
    x_username?: string;
};

export type GetMarketData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query?: {
        /**
         * Page number
         */
        page?: number;
        /**
         * Number of items per page
         */
        limit?: number;
        /**
         * Market status: activated, resolved
         */
        status?: string;
        /**
         * 0: All Markets, 1: Categorical Market, 2: Binary Market
         */
        marketType?: number;
        /**
         * Chain ID
         */
        chainId?: string;
        /**
         * Sort type: 0=NoSort, 1=ByTimeDesc, 2=ByCutoffTimeAsc, 3=ByVolumeDesc, 4=ByVolumeAsc, 5=ByVolume24hDesc, 6=ByVolume24hAsc, 7=ByVolume7dDesc, 8=ByVolume7dAsc
         */
        sortBy?: number;
        /**
         * Collection type: 0=Currently active only (default), 1=All but only collection markets, 2=No collection market
         */
        collectionType?: number;
    };
    url: '/market';
};

export type GetMarketErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetMarketError = GetMarketErrors[keyof GetMarketErrors];

export type GetMarketResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiMarketListRespOpenApi;
    };
};

export type GetMarketResponse = GetMarketResponses[keyof GetMarketResponses];

export type GetMarketCategoricalByMarketIdData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path: {
        /**
         * Market ID
         */
        marketId: number;
    };
    query?: never;
    url: '/market/categorical/{marketId}';
};

export type GetMarketCategoricalByMarketIdErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetMarketCategoricalByMarketIdError = GetMarketCategoricalByMarketIdErrors[keyof GetMarketCategoricalByMarketIdErrors];

export type GetMarketCategoricalByMarketIdResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiMarketDetailRespOpenApi;
    };
};

export type GetMarketCategoricalByMarketIdResponse = GetMarketCategoricalByMarketIdResponses[keyof GetMarketCategoricalByMarketIdResponses];

export type GetMarketByMarketIdData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path: {
        /**
         * Market ID
         */
        marketId: number;
    };
    query?: never;
    url: '/market/{marketId}';
};

export type GetMarketByMarketIdErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetMarketByMarketIdError = GetMarketByMarketIdErrors[keyof GetMarketByMarketIdErrors];

export type GetMarketByMarketIdResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiMarketDetailRespOpenApi;
    };
};

export type GetMarketByMarketIdResponse = GetMarketByMarketIdResponses[keyof GetMarketByMarketIdResponses];

export type GetMarketSlugBySlugData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path: {
        /**
         * URL-friendly slug identifier for the market
         */
        slug: string;
    };
    query?: never;
    url: '/market/slug/{slug}';
};

export type GetMarketSlugBySlugErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetMarketSlugBySlugError = GetMarketSlugBySlugErrors[keyof GetMarketSlugBySlugErrors];

export type GetMarketSlugBySlugResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiMarketDetailRespOpenApi;
    };
};

export type GetMarketSlugBySlugResponse = GetMarketSlugBySlugResponses[keyof GetMarketSlugBySlugResponses];

export type GetOrderData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query?: {
        /**
         * Page number
         */
        page?: number;
        /**
         * Number of items per page (max 20)
         */
        limit?: number;
        /**
         * Market ID filter
         */
        marketId?: number;
        /**
         * Chain ID
         */
        chainId?: string;
        /**
         * Order status filter: numeric values (1-7) or comma-separated (e.g., '1,2,3'). 1=pending, 2=filled, 3=canceled, 4=expired, 5=failed
         */
        status?: string;
    };
    url: '/order';
};

export type GetOrderErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetOrderError = GetOrderErrors[keyof GetOrderErrors];

export type GetOrderResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiOrderListRespOpenApi;
    };
};

export type GetOrderResponse = GetOrderResponses[keyof GetOrderResponses];

export type PostOrderData = {
    body: V2AddOrderReq;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query?: never;
    url: '/order';
};

export type PostOrderErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type PostOrderError = PostOrderErrors[keyof PostOrderErrors];

export type PostOrderResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiAddOrderRespOpenApi;
    };
};

export type PostOrderResponse = PostOrderResponses[keyof PostOrderResponses];

export type PostOrderCancelData = {
    body: OpenapiCancelOrderRequestOpenApi;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query?: never;
    url: '/order/cancel';
};

export type PostOrderCancelErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type PostOrderCancelError = PostOrderCancelErrors[keyof PostOrderCancelErrors];

export type PostOrderCancelResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiCancelOrderRespOpenApi;
    };
};

export type PostOrderCancelResponse = PostOrderCancelResponses[keyof PostOrderCancelResponses];

export type GetOrderByOrderIdData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path: {
        /**
         * Order ID
         */
        orderId: string;
    };
    query?: never;
    url: '/order/{orderId}';
};

export type GetOrderByOrderIdErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetOrderByOrderIdError = GetOrderByOrderIdErrors[keyof GetOrderByOrderIdErrors];

export type GetOrderByOrderIdResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiOrderDetailRespOpenApi;
    };
};

export type GetOrderByOrderIdResponse = GetOrderByOrderIdResponses[keyof GetOrderByOrderIdResponses];

export type GetPositionsData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query?: {
        /**
         * Page number
         */
        page?: number;
        /**
         * Number of items per page (max 20)
         */
        limit?: number;
        /**
         * Market ID filter
         */
        marketId?: number;
        /**
         * Chain ID
         */
        chainId?: string;
    };
    url: '/positions';
};

export type GetPositionsErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetPositionsError = GetPositionsErrors[keyof GetPositionsErrors];

export type GetPositionsResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiPositionsRespOpenApi;
    };
};

export type GetPositionsResponse = GetPositionsResponses[keyof GetPositionsResponses];

export type GetPositionsUserByWalletAddressData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path: {
        /**
         * Target user's wallet address
         */
        walletAddress: string;
    };
    query?: {
        /**
         * Page number
         */
        page?: number;
        /**
         * Number of items per page (max 20)
         */
        limit?: number;
        /**
         * Market ID filter
         */
        marketId?: number;
        /**
         * Chain ID
         */
        chainId?: string;
    };
    url: '/positions/user/{walletAddress}';
};

export type GetPositionsUserByWalletAddressErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetPositionsUserByWalletAddressError = GetPositionsUserByWalletAddressErrors[keyof GetPositionsUserByWalletAddressErrors];

export type GetPositionsUserByWalletAddressResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiPositionsRespOpenApi;
    };
};

export type GetPositionsUserByWalletAddressResponse = GetPositionsUserByWalletAddressResponses[keyof GetPositionsUserByWalletAddressResponses];

export type GetQuoteTokenData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query?: {
        /**
         * Page number
         */
        page?: number;
        /**
         * Number of items per page
         */
        limit?: number;
        /**
         * Quote token name
         */
        quoteTokenName?: string;
        /**
         * Chain ID
         */
        chainId?: string;
    };
    url: '/quoteToken';
};

export type GetQuoteTokenErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetQuoteTokenError = GetQuoteTokenErrors[keyof GetQuoteTokenErrors];

export type GetQuoteTokenResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiQuoteTokenListRespOpenApi;
    };
};

export type GetQuoteTokenResponse = GetQuoteTokenResponses[keyof GetQuoteTokenResponses];

export type GetTokenFeeRatesData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query: {
        /**
         * Token ID
         */
        token_id: string;
        /**
         * Chain ID
         */
        chain_id: string;
    };
    url: '/token/fee-rates';
};

export type GetTokenFeeRatesErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTokenFeeRatesError = GetTokenFeeRatesErrors[keyof GetTokenFeeRatesErrors];

export type GetTokenFeeRatesResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiFeeRatesRespOpenApi;
    };
};

export type GetTokenFeeRatesResponse = GetTokenFeeRatesResponses[keyof GetTokenFeeRatesResponses];

export type GetTokenLatestPriceData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query: {
        /**
         * Token ID
         */
        token_id: string;
    };
    url: '/token/latest-price';
};

export type GetTokenLatestPriceErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTokenLatestPriceError = GetTokenLatestPriceErrors[keyof GetTokenLatestPriceErrors];

export type GetTokenLatestPriceResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiLatestPriceRespOpenApi;
    };
};

export type GetTokenLatestPriceResponse = GetTokenLatestPriceResponses[keyof GetTokenLatestPriceResponses];

export type GetTokenOrderbookData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query: {
        /**
         * Token ID
         */
        token_id: string;
    };
    url: '/token/orderbook';
};

export type GetTokenOrderbookErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTokenOrderbookError = GetTokenOrderbookErrors[keyof GetTokenOrderbookErrors];

export type GetTokenOrderbookResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiOrderbookRespOpenApi;
    };
};

export type GetTokenOrderbookResponse = GetTokenOrderbookResponses[keyof GetTokenOrderbookResponses];

export type GetTokenPriceHistoryData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query: {
        /**
         * Token ID
         */
        token_id: string;
        /**
         * Price data interval: 1m, 1h, 1d, 1w, max
         */
        interval?: string;
        /**
         * Start timestamp in Unix seconds
         */
        start_at?: number;
        /**
         * End timestamp in Unix seconds
         */
        end_at?: number;
    };
    url: '/token/price-history';
};

export type GetTokenPriceHistoryErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTokenPriceHistoryError = GetTokenPriceHistoryErrors[keyof GetTokenPriceHistoryErrors];

export type GetTokenPriceHistoryResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiPriceHistoryRespOpenApi;
    };
};

export type GetTokenPriceHistoryResponse = GetTokenPriceHistoryResponses[keyof GetTokenPriceHistoryResponses];

export type GetTopicData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query?: {
        /**
         * Page number
         */
        page?: number;
        /**
         * Number of items per page
         */
        limit?: number;
        /**
         * Keywords for search (title or content)
         */
        keywords?: string;
        /**
         * 0: All Topics, 1: Categorical Topic, 2: Binary Topic
         */
        topicType?: number;
    };
    url: '/topic';
};

export type GetTopicErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTopicError = GetTopicErrors[keyof GetTopicErrors];

export type GetTopicResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: V2TopicListResp;
    };
};

export type GetTopicResponse = GetTopicResponses[keyof GetTopicResponses];

export type GetTopicMultiByTopicIdData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path: {
        /**
         * Topic ID
         */
        topicId: number;
    };
    query?: never;
    url: '/topic/multi/{topicId}';
};

export type GetTopicMultiByTopicIdErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTopicMultiByTopicIdError = GetTopicMultiByTopicIdErrors[keyof GetTopicMultiByTopicIdErrors];

export type GetTopicMultiByTopicIdResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: V2TopicDetailResp;
    };
};

export type GetTopicMultiByTopicIdResponse = GetTopicMultiByTopicIdResponses[keyof GetTopicMultiByTopicIdResponses];

export type GetTopicByTopicIdData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path: {
        /**
         * Topic ID
         */
        topicId: number;
    };
    query?: never;
    url: '/topic/{topicId}';
};

export type GetTopicByTopicIdErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTopicByTopicIdError = GetTopicByTopicIdErrors[keyof GetTopicByTopicIdErrors];

export type GetTopicByTopicIdResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: V2TopicDetailResp;
    };
};

export type GetTopicByTopicIdResponse = GetTopicByTopicIdResponses[keyof GetTopicByTopicIdResponses];

export type GetTradeData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query?: {
        /**
         * Page number
         */
        page?: number;
        /**
         * Number of items per page (max 20)
         */
        limit?: number;
        /**
         * Market ID filter
         */
        marketId?: number;
        /**
         * Chain ID
         */
        chainId?: string;
    };
    url: '/trade';
};

export type GetTradeErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTradeError = GetTradeErrors[keyof GetTradeErrors];

export type GetTradeResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiTradeListRespOpenApi;
    };
};

export type GetTradeResponse = GetTradeResponses[keyof GetTradeResponses];

export type GetTradeUserByWalletAddressData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path: {
        /**
         * Target user's wallet address
         */
        walletAddress: string;
    };
    query?: {
        /**
         * Page number
         */
        page?: number;
        /**
         * Number of items per page (max 20)
         */
        limit?: number;
        /**
         * Market ID filter
         */
        marketId?: number;
        /**
         * Chain ID
         */
        chainId?: string;
    };
    url: '/trade/user/{walletAddress}';
};

export type GetTradeUserByWalletAddressErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetTradeUserByWalletAddressError = GetTradeUserByWalletAddressErrors[keyof GetTradeUserByWalletAddressErrors];

export type GetTradeUserByWalletAddressResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiTradeListRespOpenApi;
    };
};

export type GetTradeUserByWalletAddressResponse = GetTradeUserByWalletAddressResponses[keyof GetTradeUserByWalletAddressResponses];

export type GetUserAuthData = {
    body?: never;
    path?: never;
    query?: never;
    url: '/user/auth';
};

export type GetUserAuthErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetUserAuthError = GetUserAuthErrors[keyof GetUserAuthErrors];

export type GetUserAuthResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: V2GetApiKeyResp;
    };
};

export type GetUserAuthResponse = GetUserAuthResponses[keyof GetUserAuthResponses];

export type GetUserBalanceData = {
    body?: never;
    headers: {
        /**
         * apikey
         */
        apikey: string;
    };
    path?: never;
    query: {
        /**
         * Chain ID
         */
        chain_id: string;
    };
    url: '/user/balance';
};

export type GetUserBalanceErrors = {
    /**
     * Bad Request
     */
    400: ViewApiBaseResponse;
};

export type GetUserBalanceError = GetUserBalanceErrors[keyof GetUserBalanceErrors];

export type GetUserBalanceResponses = {
    /**
     * OK
     */
    200: ViewApiBaseResponse & {
        result?: OpenapiBalanceRespOpenApi;
    };
};

export type GetUserBalanceResponse = GetUserBalanceResponses[keyof GetUserBalanceResponses];
