// @opinion-labs/opinion-clob-sdk/src/builder/user_client.ts
/**
 * UserClient - Simulates a real user's wallet client
 *
 * Used for testing Builder mode. UserClient holds the user's private key and can sign orders.
 * In production, signing is done by the user's wallet (e.g., MetaMask).
 */

import { type Address, type Hex, type PrivateKeyAccount, serializeSignature } from 'viem';
import { privateKeyToAccount, generatePrivateKey, sign } from 'viem/accounts';
import * as fs from 'fs';

/**
 * User data structure for file storage
 */
interface UserData {
  address: string;
  privateKey: string;
}

/**
 * EIP-712 typed data structure
 */
export interface TypedData {
  domain: {
    name?: string;
    version?: string;
    chainId?: number | bigint;
    verifyingContract?: Address;
  };
  types: Record<string, Array<{ name: string; type: string }>>;
  primaryType: string;
  message: Record<string, unknown>;
}

/**
 * UserClient simulates a real user's wallet client.
 *
 * Holds user's private key and can sign EIP-712 typed data.
 * Used for testing the complete Builder order flow.
 *
 * @example
 * ```typescript
 * // Create a random user
 * const user = UserClient.createRandom();
 *
 * // Or load from file
 * const user = UserClient.loadFromFile('.test_user.json');
 *
 * // Sign an order
 * const signature = await user.signTypedData(typedData);
 * ```
 */
export class UserClient {
  private account: PrivateKeyAccount;
  private _privateKey: Hex;

  /**
   * The user's wallet address
   */
  readonly address: Address;

  /**
   * Initialize UserClient.
   *
   * @param privateKey - User's private key (hex string, with or without 0x prefix)
   */
  constructor(privateKey: string) {
    let pk = privateKey;
    if (!pk.startsWith('0x')) {
      pk = '0x' + pk;
    }

    this._privateKey = pk as Hex;
    this.account = privateKeyToAccount(this._privateKey);
    this.address = this.account.address;
  }

  /**
   * Generate a random test user.
   *
   * @returns New UserClient instance
   */
  static createRandom(): UserClient {
    const privateKey = generatePrivateKey();
    return new UserClient(privateKey);
  }

  /**
   * Load user from JSON file.
   *
   * @param path - JSON file path
   * @returns UserClient instance
   * @throws Error if file does not exist or has invalid format
   */
  static loadFromFile(path: string): UserClient {
    const content = fs.readFileSync(path, 'utf-8');
    const data = JSON.parse(content) as UserData;

    if (!data.privateKey) {
      throw new Error("JSON file must contain 'privateKey' field");
    }

    return new UserClient(data.privateKey);
  }

  /**
   * Load user from file, or create new user and save if file doesn't exist.
   *
   * @param path - JSON file path
   * @returns UserClient instance
   */
  static loadOrCreate(path: string): UserClient {
    if (fs.existsSync(path)) {
      return UserClient.loadFromFile(path);
    } else {
      const user = UserClient.createRandom();
      user.saveToFile(path);
      return user;
    }
  }

  /**
   * Save user to JSON file.
   *
   * @param path - JSON file path
   */
  saveToFile(path: string): void {
    const data: UserData = {
      address: this.address,
      privateKey: this._privateKey,
    };

    fs.writeFileSync(path, JSON.stringify(data, null, 2));

    console.log(`User saved to ${path}`);
    console.log(`   Address: ${this.address}`);
  }

  /**
   * Sign EIP-712 typed data.
   *
   * This is the core method for user signing Safe transactions in Builder mode.
   * In production, this signing is done by the user's wallet.
   *
   * @param typedData - EIP-712 typed data containing domain, types, message
   * @returns Signature (hex string with 0x prefix)
   */
  async signTypedData(typedData: TypedData): Promise<Hex> {
    const signature = await this.account.signTypedData({
      domain: typedData.domain,
      types: typedData.types,
      primaryType: typedData.primaryType,
      message: typedData.message,
    });
    return signature;
  }

  /**
   * Sign a hash directly (for order signing).
   *
   * This is the core method for user signing orders in Builder mode.
   * Uses the same signing method as the original SDK.
   *
   * @param hashToSign - Hash to sign (hex string, with or without 0x prefix)
   * @returns Signature (hex string with 0x prefix)
   */
  async signHash(hashToSign: string): Promise<Hex> {
    let hash = hashToSign;
    if (!hash.startsWith('0x')) {
      hash = '0x' + hash;
    }

    const signResult = await sign({
      hash: hash as Hex,
      privateKey: this._privateKey,
    });

    return serializeSignature(signResult);
  }

  /**
   * Get the underlying account
   */
  getAccount(): PrivateKeyAccount {
    return this.account;
  }

  toString(): string {
    return `UserClient(address=${this.address})`;
  }
}
