// @prob/types/src/event.ts
import type { NormalMarket, SportsMarket } from './market'

export type RelatedEvent = {
  label: string
  slug: string
  recurrence?: string
  seriesType?: string
  startDate?: string
  active?: boolean
}

export type EventSeriesReference = {
  id: string
  title: string
  recurrence: string
  seriesType: string
}

export enum EventStatus {
  Active = 'active',
  Closed = 'closed',
  Archived = 'archived',
  Restricted = 'restricted',
  Live = 'live',
  Ended = 'ended',
  Unknown = 'unknown',
}

export enum EventType {
  Normal = 'normal',
  Sport = 'sport',
  Series = 'series',
}

export interface BaseEvent {
  id: string
  slug: string
  title: string
  /* important fields to distinguish a event from others */
  type: EventType // 'normal' | 'sport' | 'series'
  marketStructure: 'single' | 'grouped'
  status: EventStatus // 'active' | 'closed' | 'archived' | 'restricted' | 'live' | 'ended' | 'unknown'
  /* other optional fields */
  description: string
  startTime: number
  icon: string
  // competitive?: number
  volume: string
  // liquidity?: string
  markets: Array<NormalMarket | SportsMarket>
  // comment_count?: number
  // market_count?: number
  tags: string[]
  // subcategory?: string
  // enable_order_book?: boolean
  // cyom?: boolean
  // neg_risk?: boolean
  // spread?: number
  endTime: number
}

export interface NormalEvent extends BaseEvent {
  type: EventType.Normal
  markets: NormalMarket[]
}

export interface SeriesEvent extends BaseEvent {
  type: EventType.Series | EventType.Sport
  seriesSlug: string
  series?: EventSeriesReference[]
  // categories?: any[]
  // collections?: any[]
  seriesSuggestions?: RelatedEvent[]
}

export interface SportEvent extends SeriesEvent {
  type: EventType.Sport
  markets: SportsMarket[]
  score: string
  elapsed: string
  period: string
  eventDate: string
  // gameStatus?: string
  // eventWeek?: number
}

export type Event = NormalEvent | SeriesEvent

export interface EventsParams {
  limit?: number
  offset?: number
  closed?: boolean
  archived?: boolean
  active?: boolean
  related_tags?: boolean
  tag?: string // Tag slug (for backwards compatibility)
  tag_id?: number // Tag ID (preferred by API)
  recurrence?: string // Recurrence filter: 'daily', 'weekly', 'monthly'
  order?: string
  ascending?: boolean
  start_date_min?: string // Date-time format (ISO 8601)
  start_date_max?: string // Date-time format (ISO 8601)
  end_date_min?: string // Date-time format (ISO 8601)
  end_date_max?: string // Date-time format (ISO 8601)
}

export interface EventParams {
  slug: string
  with_markets?: boolean
}

export interface EventsResponse {
  data?: Event[]
  count?: number
  next_cursor?: string
}
