// @prob/types/src/history.ts
export interface UserTransactionsQueryParams {
  /**
   * Pagination limit (default: 100, range: 0–500)
   */
  limit?: number

  /**
   * Pagination offset (default: 0, range: 0–10000)
   */
  offset?: number

  /**
   * User Profile Address (0x-prefixed, 40 hex chars)
   * Example: "0x56687bf447db6ffa42ffe2204a05edaa20f55839"
   */
  user: string

  /**
   * Comma-separated list of condition IDs.
   * Mutually exclusive with eventId.
   * Example: ["0xabc123..."]
   */
  market?: string[]

  /**
   * Comma-separated list of event IDs.
   * Mutually exclusive with market.
   * Example: [23656, 23657]
   */
  eventId?: number[]

  /**
   * Filter by transaction type(s)
   * Example: ["TRADE", "REDEEM"]
   * Available options: "TRADE", "SPLIT", "MERGE", "REDEEM", "REWARD", "CONVERSION"
   */
  type?: Array<'TRADE' | 'SPLIT' | 'MERGE' | 'REDEEM' | 'REWARD' | 'CONVERSION'>

  /**
   * Start timestamp (Unix time, milliseconds)
   * Must be >= 0
   */
  start?: number

  /**
   * End timestamp (Unix time, milliseconds)
   * Must be >= 0
   */
  end?: number

  /**
   * Field to sort by (default: "TIMESTAMP")
   * Available options: "TIMESTAMP", "TOKENS", "CASH"
   */
  sortBy?: 'TIMESTAMP' | 'TOKENS' | 'CASH'

  /**
   * Sort direction (default: "DESC")
   * Available options: "ASC", "DESC"
   */
  sortDirection?: 'ASC' | 'DESC'

  /**
   * Trade side
   * Available options: "BUY", "SELL"
   */
  side?: 'BUY' | 'SELL'
}

export interface GetActivityParams {
  user: string
  limit: number
  offset: number
}

export interface UserActivity {
  proxyWallet: string
  timestamp: number
  conditionId: string
  type: string
  size: number
  usdcSize: number
  transactionHash: string
  price: number
  asset: string
  side: string
  outcomeIndex: number
  title: string
  slug: string
  icon: string
  eventSlug: string
  outcome: string
  name: string
  pseudonym: string
  bio: string
  profileImage: string
  profileImageOptimized: string
}
