// @prob/types/src/market.ts
import type { Token } from './order'

export type MarketOutcome = {
  name: string
  address: string
}

export type MarketResolveData = {
  resolved: boolean
  resolver: string
  outcome: string
  resolvedTx?: string
  resolvedTimestamp?: number
  disputed?: boolean
  livenessSeconds?: number
  disputeTx?: string
  disputeTimestamp?: number
}

export type BasicMarket = {
  id: string
  condition_id: `0x${string}`
  slug: string
  question: string
  question_id: `0x${string}`
  description?: string
  outcomes: MarketOutcome[]
  groupItemTitle?: string

  tokens: Array<{
    token_id: string
    outcome: string
  }>

  icon?: string
  image?: string
  category?: string
  tags?: string[]

  startAt: string
  endAt: string
  createdAt: string
  closedAt?: string

  status: MarketStatus
  resolve: MarketResolveData
}

export type MarketStatistics = {
  volume?: string
  volume24hr?: number
  liquidity?: string
  acceptingOrders?: boolean
  acceptingOrdersAt?: string
  rewardsMinSize?: number
  rewardsMaxSpread?: number
}

export type NormalMarket = BasicMarket & MarketStatistics

export type SportsMarket = NormalMarket & {
  sportsMarketType: string
  gameStartTime: string
}

export enum MarketStatus {
  Active = 'active',
  Resolving = 'resolving',
  Resolved = 'resolved',
  Archived = 'archived',
  Unknown = 'unknown',
}

export type Market = NormalMarket | SportsMarket
