// @prob/types/src/positions.ts
export interface UserPositionsQueryParams {
  /** User wallet address (required) */
  user: string

  /**
   * Comma-separated list of condition IDs.
   * Mutually exclusive with eventId.
   * Example: ["0xabc123..."]
   */
  market?: string[]

  /**
   * Comma-separated list of event IDs.
   * Mutually exclusive with market.
   * Example: [23656, 23657]
   */
  eventId?: number[]

  /**
   * Minimum position size threshold (default: 1)
   * Must be >= 0
   */
  sizeThreshold?: number

  /** Only include redeemable positions (default: false) */
  redeemable?: boolean

  /** Only include mergeable positions (default: false) */
  mergeable?: boolean

  /** Pagination limit (default: 100, range: 0–500) */
  limit?: number

  /** Pagination offset (default: 0, range: 0–10000) */
  offset?: number

  /**
   * Sort field
   * @default "TOKENS"
   * Available options: "CURRENT", "INITIAL", "TOKENS", "CASHPNL", "PERCENTPNL", "TITLE", "RESOLVING", "PRICE", "AVGPRICE"
   */
  sortBy?:
    | 'CURRENT'
    | 'INITIAL'
    | 'TOKENS'
    | 'CASHPNL'
    | 'PERCENTPNL'
    | 'TITLE'
    | 'RESOLVING'
    | 'PRICE'
    | 'AVGPRICE'

  /**
   * Sort direction
   * @default "DESC"
   * Available options: "ASC", "DESC"
   */
  sortDirection?: 'ASC' | 'DESC'

  /** Optional title search filter (max length: 100 chars) */
  title?: string
}

export interface UserPositionMarket {
  /** Token symbol or contract address of the asset */
  asset: string

  /** 0x-prefixed 64-hex condition ID */
  conditionId: string

  /** Market question title */
  title: string

  /** Market slug (used for routing/links) */
  slug: string

  /** Market or event icon URL */
  icon: string

  /** Parent event slug */
  eventSlug: string

  /** Outcome label (e.g. "Yes", "No", "Over", "Under") */
  outcome: string

  /** Outcome index (0-based) */
  outcomeIndex: number

  /** Opposite outcome label */
  oppositeOutcome: string

  /** Opposite asset token address */
  oppositeAsset: string

  /** Market end date (ISO string) */
  endDate: string

  /** Whether this market is part of a negative risk set */
  negativeRisk: boolean
}


export interface PositionValue {
  user: string
  value: number
}

export interface GetCurrentPositionsParams {
  /** User wallet address (required) */
  user: string

  /** Pagination limit */
  limit?: number

  /** Pagination offset */
  offset?: number

  /** Sort field */
  sortBy?: string

  /** Sort direction */
  sortDirection?: string

  /** Only include redeemable positions */
  redeemable?: boolean

  /** Event ID */
  eventId?: string

  /** Optional title search filter */
  title?: string
}

export interface UserPosition {
  /** User Profile Address (0x-prefixed, 40 hex chars) */
  proxyWallet: string

  /** Token symbol or contract address of the asset */
  asset: string

  /** 0x-prefixed 64-hex condition ID */
  conditionId: string

  /** Size of the user's position */
  size: number

  /** Average buy price */
  avgPrice: number

  /** Initial position value (based on purchase price) */
  initialValue: number

  /** Current market value */
  currentValue: number

  /** Cash profit or loss */
  cashPnl: number

  /** Profit or loss as a percentage */
  percentPnl: number

  /** Total amount of tokens bought */
  totalBought: number

  /** Realized profit or loss (closed positions) */
  realizedPnl: number

  /** Percentage of realized profit or loss */
  percentRealizedPnl: number

  /** Current market price for the outcome token */
  curPrice: number

  /** Whether this position can be redeemed (after market resolves) */
  redeemable: boolean

  /** Whether multiple positions can be merged */
  mergeable: boolean

  /** Market question title */
  title: string

  /** Market slug (used for routing/links) */
  slug: string

  /** Market or event icon URL */
  icon: string

  /** Parent event slug */
  eventSlug: string

  /** Outcome label (e.g. "Yes", "No", "Over", "Under") */
  outcome: string

  /** Outcome index (0-based) */
  outcomeIndex: number

  /** Opposite outcome label */
  oppositeOutcome: string

  /** Opposite asset token address */
  oppositeAsset: string

  /** Market end date (ISO string) */
  endDate: string

  /** Whether this market is part of a negative risk set */
  negativeRisk: boolean
}
